{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c341e94",
   "metadata": {},
   "source": [
    "# Chaining Search Terms\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8387f59",
   "metadata": {},
   "source": [
    "Before we do any work, we need to import several functions from cdapython:\n",
    "\n",
    "\n",
    "- `Q` and `query` which power the search\n",
    "- `columns` which lets us view entity field names\n",
    "- `unique_terms` which lets view entity field contents\n",
    "    \n",
    "We're also asking cdapython to report it's version so we can be sure we're using the one we mean to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28a10e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.21\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "print(cdapython.__version__)\n",
    "Q.set_host_url(\"http://35.192.60.10:8080/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0810c61c",
   "metadata": {},
   "source": [
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "\n",
    "- **[Q.run()](../../../Documentation/usage/#qrun)** returns data for the specified search \n",
    "- **[Q.count()](../../../Documentation/usage/#qcounts)** returns summary information (counts) data that fit the specified search\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8084b3",
   "metadata": {},
   "source": [
    "## Endpoint Chaining\n",
    "\n",
    "We're going to build on our [previous search](../BasicSearch) to see what information exists about cancers that were first diagnosed in the brain. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d804f764",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "736ad485",
   "metadata": {},
   "source": [
    "Previously we looked at subject, research_subject, specimen and file results seperately, but we can also combine these. \n",
    "\n",
    "Let's say what we're really interested in is finding analysis done on specimens, so we're looking for files that belong to specimens that match our search. If we search the files endpoint directly, we'll get back all files that meet our search criteria, regardless of whether they are specimens specficially. Instead, we can chain our query to the specimen endpoint and then to the files endpoint and get the combined result:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7284df72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3715 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 6f1e0228-5c96-4e0f-a061-0dd2b86d31e9\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 405031\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles =  myquery.specimen.file.run()\n",
    "myqueryspecimenfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d9154b",
   "metadata": {},
   "source": [
    "We get back 50495 files that belong to specimens that meet our search criteria. As before, we can preview the results by using the `.to_dataframe()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "410b0e2b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11b33eec-a909-492d-a6ab-24f422eafd7f</td>\n",
       "      <td>[{'system': 'GDC', 'value': '11b33eec-a909-492...</td>\n",
       "      <td>TCGA-S9-A7R8-01A-21-A44C-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:11b33eec-a909-492d-a6ab-24f422ea...</td>\n",
       "      <td>22001</td>\n",
       "      <td>f02f1a7a67103b440c140fdb46d1b177</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A7R8</td>\n",
       "      <td>7883e417-a793-4614-9528-6c2e9e44379f</td>\n",
       "      <td>8d5144f0-21cf-41b8-950d-478d8c6a09a5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1c962277-712d-4f3f-b897-bf095007834f</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1c962277-712d-4f3...</td>\n",
       "      <td>5da533e1-7ae6-4ada-b546-1302c4f848ee.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>drs://dg.4DFC:1c962277-712d-4f3f-b897-bf095007...</td>\n",
       "      <td>4200</td>\n",
       "      <td>43c5d345f83c82a7b1d7cff2b0cb6914</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-DFCI-008524</td>\n",
       "      <td>9838df48-80d1-459c-8bd3-27b89bbc07a2</td>\n",
       "      <td>f99ec994-498e-4a06-b48c-b1c2b604b995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>215266c2-bd28-4be1-abc4-822c17e5cf39</td>\n",
       "      <td>[{'system': 'GDC', 'value': '215266c2-bd28-4be...</td>\n",
       "      <td>3fb08188-6d48-4ba4-a60d-1aa2481f6acc.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:215266c2-bd28-4be1-abc4-822c17e5...</td>\n",
       "      <td>22932</td>\n",
       "      <td>b11afd90588929a135011ad6ce488b2a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0005276</td>\n",
       "      <td>33c52fe3-b1ea-439e-a2c1-40d22611c5fc</td>\n",
       "      <td>9a98756b-c890-5249-8fda-d09387e5227c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2bb9b55d-7ba3-470e-be48-80bb3a34ca1d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2bb9b55d-7ba3-470...</td>\n",
       "      <td>TCGA-FG-A60L-01A-21-A44D-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:2bb9b55d-7ba3-470e-be48-80bb3a34...</td>\n",
       "      <td>22003</td>\n",
       "      <td>d7f9da1c349215b3162ed8c1ca36a65e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-FG-A60L</td>\n",
       "      <td>738f5c96-ee44-4fd7-bb8c-e2b86f8053c8</td>\n",
       "      <td>b82e8a8b-1274-4c00-a264-962ee6e5dd21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2d3968a5-5ae4-4f7b-8888-7d9dd6c1bb07</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2d3968a5-5ae4-4f7...</td>\n",
       "      <td>219b6c16-a27d-4faf-bb26-c648227d34a4.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MDA</td>\n",
       "      <td>drs://dg.4DFC:2d3968a5-5ae4-4f7b-8888-7d9dd6c1...</td>\n",
       "      <td>1053</td>\n",
       "      <td>849544142f98fb56cb8cb25fb0c7ceb5</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MDA-10974</td>\n",
       "      <td>ba433488-0dce-4b73-b514-ee3063a539da</td>\n",
       "      <td>a971368a-fc80-4a01-b4d0-e4d545f0d009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>4f8d926f-0092-4af7-b052-58b8ca3f3af6</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4f8d926f-0092-4af...</td>\n",
       "      <td>3177e8cd-1a42-44d1-b5ae-f43d39104218.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>drs://dg.4DFC:4f8d926f-0092-4af7-b052-58b8ca3f...</td>\n",
       "      <td>4131</td>\n",
       "      <td>af9ced3a578019b6d0d10804ea327bd6</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-DFCI-008015</td>\n",
       "      <td>5456dedb-70ef-47af-b3ff-769795aaa089</td>\n",
       "      <td>e3c2fb74-a76f-521a-adea-2b0f2522f839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>59f2933c-8e71-48e3-a10e-a82de3909a5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '59f2933c-8e71-48e...</td>\n",
       "      <td>04ddc9f5-d77b-46b6-bd9c-6aa61f733c10.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:59f2933c-8e71-48e3-a10e-a82de390...</td>\n",
       "      <td>22932</td>\n",
       "      <td>a1b7d05c86f982a308515257875aa9ce</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0002636</td>\n",
       "      <td>ec8a314e-8edb-4703-91f3-da9ceb85e357</td>\n",
       "      <td>5ade49c4-a932-57d8-aa3a-d03118e174f6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>61ebaf46-6b86-4209-8b30-e5c3cb098d7f</td>\n",
       "      <td>[{'system': 'GDC', 'value': '61ebaf46-6b86-420...</td>\n",
       "      <td>2f0970e6-a701-4372-b183-7e03b53c880f.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>drs://dg.4DFC:61ebaf46-6b86-4209-8b30-e5c3cb09...</td>\n",
       "      <td>1050</td>\n",
       "      <td>4b9800e4aef4553953e03f70393af32a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-DFCI-000904</td>\n",
       "      <td>2299b2db-1b34-4833-bb1d-04fb529aed64</td>\n",
       "      <td>354c9414-62f4-559c-8ffe-eb8c8c09a971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>6ae411a6-1215-4b50-afc0-7bdb777e25d9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6ae411a6-1215-4b5...</td>\n",
       "      <td>TCGA-TQ-A7RF-01A-21-A44C-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:6ae411a6-1215-4b50-afc0-7bdb777e...</td>\n",
       "      <td>22003</td>\n",
       "      <td>6468eef92eea945ca0acebf1a95b3345</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-TQ-A7RF</td>\n",
       "      <td>f5268956-789d-4b22-8f19-a3ac9353d33d</td>\n",
       "      <td>8e6663ee-8c6a-412e-b5d1-6d855916367a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>6bc8330b-d971-4bdb-b14f-f9f1cdc096a6</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6bc8330b-d971-4bd...</td>\n",
       "      <td>8f2acbdd-9516-47fd-9748-d419a1239428.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:6bc8330b-d971-4bdb-b14f-f9f1cdc0...</td>\n",
       "      <td>22932</td>\n",
       "      <td>4a17cf06f587337efeabe8fcd3ebedc3</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0008309</td>\n",
       "      <td>ec00e6fc-bb1a-4a72-bb1e-4d8bef926548</td>\n",
       "      <td>01e3e01b-dc70-474d-b184-c79b225bcbc0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   11b33eec-a909-492d-a6ab-24f422eafd7f   \n",
       "1   1c962277-712d-4f3f-b897-bf095007834f   \n",
       "2   215266c2-bd28-4be1-abc4-822c17e5cf39   \n",
       "3   2bb9b55d-7ba3-470e-be48-80bb3a34ca1d   \n",
       "4   2d3968a5-5ae4-4f7b-8888-7d9dd6c1bb07   \n",
       "..                                   ...   \n",
       "95  4f8d926f-0092-4af7-b052-58b8ca3f3af6   \n",
       "96  59f2933c-8e71-48e3-a10e-a82de3909a5c   \n",
       "97  61ebaf46-6b86-4209-8b30-e5c3cb098d7f   \n",
       "98  6ae411a6-1215-4b50-afc0-7bdb777e25d9   \n",
       "99  6bc8330b-d971-4bdb-b14f-f9f1cdc096a6   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '11b33eec-a909-492...   \n",
       "1   [{'system': 'GDC', 'value': '1c962277-712d-4f3...   \n",
       "2   [{'system': 'GDC', 'value': '215266c2-bd28-4be...   \n",
       "3   [{'system': 'GDC', 'value': '2bb9b55d-7ba3-470...   \n",
       "4   [{'system': 'GDC', 'value': '2d3968a5-5ae4-4f7...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '4f8d926f-0092-4af...   \n",
       "96  [{'system': 'GDC', 'value': '59f2933c-8e71-48e...   \n",
       "97  [{'system': 'GDC', 'value': '61ebaf46-6b86-420...   \n",
       "98  [{'system': 'GDC', 'value': '6ae411a6-1215-4b5...   \n",
       "99  [{'system': 'GDC', 'value': '6bc8330b-d971-4bd...   \n",
       "\n",
       "                                                label  \\\n",
       "0           TCGA-S9-A7R8-01A-21-A44C-20_RPPA_data.tsv   \n",
       "1   5da533e1-7ae6-4ada-b546-1302c4f848ee.genie.ali...   \n",
       "2   3fb08188-6d48-4ba4-a60d-1aa2481f6acc.genie.ali...   \n",
       "3           TCGA-FG-A60L-01A-21-A44D-20_RPPA_data.tsv   \n",
       "4   219b6c16-a27d-4faf-bb26-c648227d34a4.genie.ali...   \n",
       "..                                                ...   \n",
       "95  3177e8cd-1a42-44d1-b5ae-f43d39104218.genie.ali...   \n",
       "96  04ddc9f5-d77b-46b6-bd9c-6aa61f733c10.genie.ali...   \n",
       "97  2f0970e6-a701-4372-b183-7e03b53c880f.genie.ali...   \n",
       "98          TCGA-TQ-A7RF-01A-21-A44C-20_RPPA_data.tsv   \n",
       "99  8f2acbdd-9516-47fd-9748-d419a1239428.genie.ali...   \n",
       "\n",
       "                  data_category                          data_type  \\\n",
       "0            Proteome Profiling  Protein Expression Quantification   \n",
       "1   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "2         Copy Number Variation      Gene Level Copy Number Scores   \n",
       "3            Proteome Profiling  Protein Expression Quantification   \n",
       "4   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "..                          ...                                ...   \n",
       "95  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "96        Copy Number Variation      Gene Level Copy Number Scores   \n",
       "97  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "98           Proteome Profiling  Protein Expression Quantification   \n",
       "99        Copy Number Variation      Gene Level Copy Number Scores   \n",
       "\n",
       "   file_format associated_project  \\\n",
       "0          TSV           TCGA-LGG   \n",
       "1          MAF         GENIE-DFCI   \n",
       "2          TSV          GENIE-MSK   \n",
       "3          TSV           TCGA-LGG   \n",
       "4          MAF          GENIE-MDA   \n",
       "..         ...                ...   \n",
       "95         MAF         GENIE-DFCI   \n",
       "96         TSV          GENIE-MSK   \n",
       "97         MAF         GENIE-DFCI   \n",
       "98         TSV           TCGA-LGG   \n",
       "99         TSV          GENIE-MSK   \n",
       "\n",
       "                                              drs_uri  byte_size  \\\n",
       "0   drs://dg.4DFC:11b33eec-a909-492d-a6ab-24f422ea...      22001   \n",
       "1   drs://dg.4DFC:1c962277-712d-4f3f-b897-bf095007...       4200   \n",
       "2   drs://dg.4DFC:215266c2-bd28-4be1-abc4-822c17e5...      22932   \n",
       "3   drs://dg.4DFC:2bb9b55d-7ba3-470e-be48-80bb3a34...      22003   \n",
       "4   drs://dg.4DFC:2d3968a5-5ae4-4f7b-8888-7d9dd6c1...       1053   \n",
       "..                                                ...        ...   \n",
       "95  drs://dg.4DFC:4f8d926f-0092-4af7-b052-58b8ca3f...       4131   \n",
       "96  drs://dg.4DFC:59f2933c-8e71-48e3-a10e-a82de390...      22932   \n",
       "97  drs://dg.4DFC:61ebaf46-6b86-4209-8b30-e5c3cb09...       1050   \n",
       "98  drs://dg.4DFC:6ae411a6-1215-4b50-afc0-7bdb777e...      22003   \n",
       "99  drs://dg.4DFC:6bc8330b-d971-4bdb-b14f-f9f1cdc0...      22932   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   f02f1a7a67103b440c140fdb46d1b177       Genomic             None   \n",
       "1   43c5d345f83c82a7b1d7cff2b0cb6914       Genomic             None   \n",
       "2   b11afd90588929a135011ad6ce488b2a       Genomic             None   \n",
       "3   d7f9da1c349215b3162ed8c1ca36a65e       Genomic             None   \n",
       "4   849544142f98fb56cb8cb25fb0c7ceb5       Genomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  af9ced3a578019b6d0d10804ea327bd6       Genomic             None   \n",
       "96  a1b7d05c86f982a308515257875aa9ce       Genomic             None   \n",
       "97  4b9800e4aef4553953e03f70393af32a       Genomic             None   \n",
       "98  6468eef92eea945ca0acebf1a95b3345       Genomic             None   \n",
       "99  4a17cf06f587337efeabe8fcd3ebedc3       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number           subject_id  \\\n",
       "0                    None         TCGA-S9-A7R8   \n",
       "1                    None    GENIE-DFCI-008524   \n",
       "2                    None  GENIE-MSK-P-0005276   \n",
       "3                    None         TCGA-FG-A60L   \n",
       "4                    None      GENIE-MDA-10974   \n",
       "..                    ...                  ...   \n",
       "95                   None    GENIE-DFCI-008015   \n",
       "96                   None  GENIE-MSK-P-0002636   \n",
       "97                   None    GENIE-DFCI-000904   \n",
       "98                   None         TCGA-TQ-A7RF   \n",
       "99                   None  GENIE-MSK-P-0008309   \n",
       "\n",
       "                      researchsubject_id           researchsubject_specimen_id  \n",
       "0   7883e417-a793-4614-9528-6c2e9e44379f  8d5144f0-21cf-41b8-950d-478d8c6a09a5  \n",
       "1   9838df48-80d1-459c-8bd3-27b89bbc07a2  f99ec994-498e-4a06-b48c-b1c2b604b995  \n",
       "2   33c52fe3-b1ea-439e-a2c1-40d22611c5fc  9a98756b-c890-5249-8fda-d09387e5227c  \n",
       "3   738f5c96-ee44-4fd7-bb8c-e2b86f8053c8  b82e8a8b-1274-4c00-a264-962ee6e5dd21  \n",
       "4   ba433488-0dce-4b73-b514-ee3063a539da  a971368a-fc80-4a01-b4d0-e4d545f0d009  \n",
       "..                                   ...                                   ...  \n",
       "95  5456dedb-70ef-47af-b3ff-769795aaa089  e3c2fb74-a76f-521a-adea-2b0f2522f839  \n",
       "96  ec8a314e-8edb-4703-91f3-da9ceb85e357  5ade49c4-a932-57d8-aa3a-d03118e174f6  \n",
       "97  2299b2db-1b34-4833-bb1d-04fb529aed64  354c9414-62f4-559c-8ffe-eb8c8c09a971  \n",
       "98  f5268956-789d-4b22-8f19-a3ac9353d33d  8e6663ee-8c6a-412e-b5d1-6d855916367a  \n",
       "99  ec00e6fc-bb1a-4a72-bb1e-4d8bef926548  01e3e01b-dc70-474d-b184-c79b225bcbc0  \n",
       "\n",
       "[100 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fda2e6",
   "metadata": {},
   "source": [
    "<div style=\"background-color:#6ce6b9;color:black;padding:20px;\">\n",
    "<h3>Valid Endpoint Chains</h3>\n",
    "\n",
    "Not all endpoints can be chained together. This is a restriction caused by the data itself. `diagnosis` and `treatment` information does not have files directly attached to it, instead these files are associated with the `researchsubject`, as such both \"myquery.treatment.files.run()\" and \"myquery.diagnosis.files.run()\" will fail, as there are no files to retrieve. Valid chains are:\n",
    "    \n",
    "<ul>\n",
    "<li><b>myquery.subject.file.run:</b> This will return all the files that meet the query and that are directly tied to subject</li>\n",
    "<li><b>myquery.researchsubject.file.run:</b>This will return all the files that meet the query and that are directly tied to researchsubject</li>\n",
    "<li><b>myquery.specimen.file.run:</b> This will return all the files that meet the query and that are directly tied to specimen</li>\n",
    "<li><b>myquery.subject.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to subject</li>\n",
    "<li><b>myquery.researchsubject.file.count.run:</b>This will return the count of files that meet the query and that are directly tied to researchsubject</li>\n",
    "<li><b>myquery.specimen.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to specimen</li>\n",
    "</ul>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
