{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a72d1a4b",
   "metadata": {},
   "source": [
    "# Basic Search"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "143f2e20",
   "metadata": {},
   "source": [
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "                                                                    \n",
    "Before we do any work, we needs to import these functions cdapython.\n",
    "We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5245b7b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "print(Q.get_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "02c50256",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()\n",
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d538da92",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n",
    "\n",
    "\n",
    "If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specfic format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.\n",
    "\n",
    "In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.\n",
    "\n",
    "In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76026a3",
   "metadata": {},
   "source": [
    "## Basic search with endpoints\n",
    "\n",
    "Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c92a98ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e855573",
   "metadata": {
    "tags": [
     "Help!"
    ]
   },
   "source": [
    "\n",
    "<div class=\"cdawarn\" style=\"background-color:#f9cfbf;color:black;padding:20px;\">\n",
    "    \n",
    "<h3>Where did those terms come from?</h3>\n",
    "    \n",
    "If you aren't sure how we knew what terms to put in our search, please refer back to the <a href=\"../SearchTerms\">What search terms are available?</a> notebook. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d483e944",
   "metadata": {},
   "source": [
    "### subject\n",
    "Now we can use that query to search any of information types. Let's start by looking at what Subjects meet our criteria. To do that, we will send our query to the subject endpoint, then ask for it to run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d9137aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3786 ms\n"
     ]
    }
   ],
   "source": [
    "subjectresults = myquery.subject.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf7fb60",
   "metadata": {},
   "source": [
    "We saved the output in a variable `subjectresults`, so we don't get much visible output. To see what our results are, we need to look into the variable. The simplest way is to call `subjectresults` directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cff572d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 60d928ef-0741-4f38-8f2f-941c02d84fec\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 2384\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5186036c",
   "metadata": {},
   "source": [
    "This output tells us our QueryID, which we don't really need, but the computer does to track our questions. Then it tells us four parameters that describe our results:\n",
    "\n",
    "---\n",
    "\n",
    "- **Offset:** This is how many rows of information we've told the query to skip in the data, here we didn't tell it to skip anything, so the offset is zero\n",
    "- **Count:** This is how many rows the current page of our results table has. To keep searches fast, we default to pages with 100 rows.\n",
    "- **Total Row Count:** This is how many rows are in the full results table\n",
    "- **More pages:** This is always a True or False. False means that our current page has all the available results. True means that we will see only the first 100 results in this table, and will need to page through for more.\n",
    "\n",
    "---\n",
    "    \n",
    "We have 2384 subject matches, let's look at the actual table. The easiest way to do this is by using the python function `.to_dataframe()` on our `subjectresults` variable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e97e19f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACRIN-DSC-MR-Brain-048</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'ACRIN-DSC-MR-Brai...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[acrin_dsc_mr_brain]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACRIN-FMISO-Brain-050</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'ACRIN-FMISO-Brain...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[acrin_fmiso_brain]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-04795</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-04795'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_gbm]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3N-02186</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'C3N-02186'}, {'sy...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-19526.0</td>\n",
       "      <td>[CPTAC-3, cptac_gbm, CPTAC3-Discovery]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3N-03186</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'C3N-03186'}, {'sy...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-20009.0</td>\n",
       "      <td>[CPTAC-3, cptac_gbm, CPTAC3-Discovery]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>646.0</td>\n",
       "      <td>Cancer Related</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>QIN-GBM-TR-44</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'QIN-GBM-TR-44'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[qin_gbm_treatment_response]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-02-0027</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-02-0027'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-12369.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>370.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-06-0747</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-0747'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-19657.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>82.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-06-0877</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-0877'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-28760.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-12-0707</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-12-0707'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20136.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>864.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        id                                         identifier  \\\n",
       "0   ACRIN-DSC-MR-Brain-048  [{'system': 'IDC', 'value': 'ACRIN-DSC-MR-Brai...   \n",
       "1    ACRIN-FMISO-Brain-050  [{'system': 'IDC', 'value': 'ACRIN-FMISO-Brain...   \n",
       "2                C3L-04795          [{'system': 'IDC', 'value': 'C3L-04795'}]   \n",
       "3                C3N-02186  [{'system': 'GDC', 'value': 'C3N-02186'}, {'sy...   \n",
       "4                C3N-03186  [{'system': 'GDC', 'value': 'C3N-03186'}, {'sy...   \n",
       "..                     ...                                                ...   \n",
       "95           QIN-GBM-TR-44      [{'system': 'IDC', 'value': 'QIN-GBM-TR-44'}]   \n",
       "96            TCGA-02-0027  [{'system': 'GDC', 'value': 'TCGA-02-0027'}, {...   \n",
       "97            TCGA-06-0747  [{'system': 'GDC', 'value': 'TCGA-06-0747'}, {...   \n",
       "98            TCGA-06-0877  [{'system': 'GDC', 'value': 'TCGA-06-0877'}, {...   \n",
       "99            TCGA-12-0707  [{'system': 'GDC', 'value': 'TCGA-12-0707'}, {...   \n",
       "\n",
       "         species     sex          race               ethnicity  days_to_birth  \\\n",
       "0   Homo sapiens    None          None                    None            NaN   \n",
       "1   Homo sapiens    None          None                    None            NaN   \n",
       "2   Homo sapiens    None          None                    None            NaN   \n",
       "3   homo sapiens    male  not reported            not reported       -19526.0   \n",
       "4   homo sapiens  female  not reported            not reported       -20009.0   \n",
       "..           ...     ...           ...                     ...            ...   \n",
       "95  Homo sapiens    None          None                    None            NaN   \n",
       "96  homo sapiens  female         white  not hispanic or latino       -12369.0   \n",
       "97  homo sapiens    male         white  not hispanic or latino       -19657.0   \n",
       "98  homo sapiens    male         white  not hispanic or latino       -28760.0   \n",
       "99  homo sapiens    male         white  not hispanic or latino       -20136.0   \n",
       "\n",
       "                subject_associated_project vital_status  days_to_death  \\\n",
       "0                     [acrin_dsc_mr_brain]         None            NaN   \n",
       "1                      [acrin_fmiso_brain]         None            NaN   \n",
       "2                              [cptac_gbm]         None            NaN   \n",
       "3   [CPTAC-3, cptac_gbm, CPTAC3-Discovery]        Alive            NaN   \n",
       "4   [CPTAC-3, cptac_gbm, CPTAC3-Discovery]         Dead          646.0   \n",
       "..                                     ...          ...            ...   \n",
       "95            [qin_gbm_treatment_response]         None            NaN   \n",
       "96                    [tcga_gbm, TCGA-GBM]         Dead          370.0   \n",
       "97                    [tcga_gbm, TCGA-GBM]         Dead           82.0   \n",
       "98                    [tcga_gbm, TCGA-GBM]         Dead            NaN   \n",
       "99                    [tcga_gbm, TCGA-GBM]         Dead          864.0   \n",
       "\n",
       "    cause_of_death  \n",
       "0             None  \n",
       "1             None  \n",
       "2             None  \n",
       "3             None  \n",
       "4   Cancer Related  \n",
       "..             ...  \n",
       "95            None  \n",
       "96            None  \n",
       "97            None  \n",
       "98            None  \n",
       "99            None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44a65df",
   "metadata": {},
   "source": [
    "By default `to_dataframe()` shows us the first and last five rows for the first page of our results, so we can easily preview our data.\n",
    "\n",
    "Since we queried the Subject endpoint, our default results tell us Subject level information, that is, information about unique individuals: their sex, race, age, species, etc. The `id` column tells us the unique identifier for each individual. The identifier column has nested information about what study or studies a Subject participated in, and will list all of their researchsubject identifiers. \n",
    "\n",
    "The `to_dataframe()` function converts the results to a pandas dataframe. So if we save the dataframe to a variable, we can use any pandas functions to explore it. For example, lets see whether any of the Subjects in our first 100 results are black or african american. First we'll save the results to a dataframe, then subset that dataframe to only show rows where the word \"black\" appears in the \"race\" column. \"NAs\" which are shown as \"None\" in these tables, so for our filter to work, we'll need to specifically tell it to ignore NAs:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4975c1e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>TCGA-08-0521</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-08-0521'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>-6377.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>146.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                         identifier  \\\n",
       "69  TCGA-08-0521  [{'system': 'GDC', 'value': 'TCGA-08-0521'}, {...   \n",
       "\n",
       "         species   sex                       race           ethnicity  \\\n",
       "69  homo sapiens  male  black or african american  hispanic or latino   \n",
       "\n",
       "    days_to_birth subject_associated_project vital_status  days_to_death  \\\n",
       "69        -6377.0       [tcga_gbm, TCGA-GBM]         Dead          146.0   \n",
       "\n",
       "   cause_of_death  \n",
       "69           None  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectdata = subjectresults.to_dataframe()\n",
    "subjectdata[subjectdata['race'].str.contains(\"black\", na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8750617",
   "metadata": {},
   "source": [
    "There are subjects in our first hundred results that meet the criteria. If we just want to be sure that the data contains some value, this might be good enough. But often we want to search the entire set of results and not just the first page. \n",
    "\n",
    "We'll cover how to work with large results dataframes in the <a href=\"../Pagination\">Pagination</a> notebook. Or, learn how to get summary information from search results in the <a href=\"../DataSummaries\">Data Summaries</a> notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05e52f3f",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the subject. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73745aa6",
   "metadata": {},
   "source": [
    "### researchsubject\n",
    "\n",
    "If we're interested in what researchsubjects meet our critera, we can also run our query against the researchsubject endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2d8c874",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3968 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: bd1c1b69-827e-4b7a-a879-06627ce21d88\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 3716\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults = myquery.researchsubject.run()\n",
    "researchsubjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "977f36fe",
   "metadata": {},
   "source": [
    "Now we see that our 2384 subjects have 3716 researchsubjects between them, that means that some, but not all, of our subjects were participants in more than one study. Let's peek at the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a20f7e62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9c32418f-ce23-4259-86f8-0b3b46fcfe33</td>\n",
       "      <td>[{'system': 'GDC', 'value': '9c32418f-ce23-425...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-03295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b6e6624f-5776-4d7e-a2c1-595b2b539f46</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b6e6624f-5776-4d7...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Malignant Lymphomas, NOS or Diffuse</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-02253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51343e16-05b6-4314-8169-59ff765bda0a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '51343e16-05b6-431...</td>\n",
       "      <td>EXCEPTIONAL_RESPONDERS-ER</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>ER-B0DO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>54ada9c7-e8f7-4ecc-b829-903be7e3ad82</td>\n",
       "      <td>[{'system': 'GDC', 'value': '54ada9c7-e8f7-4ec...</td>\n",
       "      <td>GENIE-NKI</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-NKI-8IN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>be68cfb0-c47f-49a1-9e39-b0b3e57372fc</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'be68cfb0-c47f-49a...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-02615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>5838a973-c162-47cd-a93d-e28fd92f1ca5</td>\n",
       "      <td>[{'system': 'GDC', 'value': '5838a973-c162-47c...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-01046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ba037761-7166-40e1-8f27-1c1561a18b9c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ba037761-7166-40e...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-02970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>c5d83fdb-5a16-44c2-b392-8deb1a3c5b57</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c5d83fdb-5a16-44c...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-01327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>5718a50d-332c-4882-8736-2ca8989946dd</td>\n",
       "      <td>[{'system': 'GDC', 'value': '5718a50d-332c-488...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-02984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1451b12b-b8b7-47eb-a919-8043b8bafae9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1451b12b-b8b7-47e...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-NPJ7-0011-R10A-SM-HAKXW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   9c32418f-ce23-4259-86f8-0b3b46fcfe33   \n",
       "1   b6e6624f-5776-4d7e-a2c1-595b2b539f46   \n",
       "2   51343e16-05b6-4314-8169-59ff765bda0a   \n",
       "3   54ada9c7-e8f7-4ecc-b829-903be7e3ad82   \n",
       "4   be68cfb0-c47f-49a1-9e39-b0b3e57372fc   \n",
       "..                                   ...   \n",
       "95  5838a973-c162-47cd-a93d-e28fd92f1ca5   \n",
       "96  ba037761-7166-40e1-8f27-1c1561a18b9c   \n",
       "97  c5d83fdb-5a16-44c2-b392-8deb1a3c5b57   \n",
       "98  5718a50d-332c-4882-8736-2ca8989946dd   \n",
       "99  1451b12b-b8b7-47eb-a919-8043b8bafae9   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '9c32418f-ce23-425...   \n",
       "1   [{'system': 'GDC', 'value': 'b6e6624f-5776-4d7...   \n",
       "2   [{'system': 'GDC', 'value': '51343e16-05b6-431...   \n",
       "3   [{'system': 'GDC', 'value': '54ada9c7-e8f7-4ec...   \n",
       "4   [{'system': 'GDC', 'value': 'be68cfb0-c47f-49a...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '5838a973-c162-47c...   \n",
       "96  [{'system': 'GDC', 'value': 'ba037761-7166-40e...   \n",
       "97  [{'system': 'GDC', 'value': 'c5d83fdb-5a16-44c...   \n",
       "98  [{'system': 'GDC', 'value': '5718a50d-332c-488...   \n",
       "99  [{'system': 'GDC', 'value': '1451b12b-b8b7-47e...   \n",
       "\n",
       "   member_of_research_project          primary_diagnosis_condition  \\\n",
       "0                   GENIE-JHU                       Neoplasms, NOS   \n",
       "1                   GENIE-JHU  Malignant Lymphomas, NOS or Diffuse   \n",
       "2   EXCEPTIONAL_RESPONDERS-ER                       Neoplasms, NOS   \n",
       "3                   GENIE-NKI                       Neoplasms, NOS   \n",
       "4                   GENIE-JHU                       Neoplasms, NOS   \n",
       "..                        ...                                  ...   \n",
       "95                    CPTAC-3                              Gliomas   \n",
       "96                    CPTAC-3                              Gliomas   \n",
       "97                    CPTAC-3                              Gliomas   \n",
       "98                    CPTAC-3                              Gliomas   \n",
       "99                    CPTAC-3                       Not Applicable   \n",
       "\n",
       "   primary_diagnosis_site                    subject_id  \n",
       "0                   Brain               GENIE-JHU-03295  \n",
       "1                   Brain               GENIE-JHU-02253  \n",
       "2                   Brain                       ER-B0DO  \n",
       "3                   Brain                GENIE-NKI-8IN3  \n",
       "4                   Brain               GENIE-JHU-02615  \n",
       "..                    ...                           ...  \n",
       "95                  Brain                     C3L-01046  \n",
       "96                  Brain                     C3L-02970  \n",
       "97                  Brain                     C3L-01327  \n",
       "98                  Brain                     C3L-02984  \n",
       "99                  Brain  GTEX-NPJ7-0011-R10A-SM-HAKXW  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4564beee",
   "metadata": {},
   "source": [
    "Each row from the researchsubject endpoint results tells us about a subject in a given study. Using this endpoint we can find out information like what studies fit our search criteria, and also get data that we can filter to have only subjects from multiple studies, or only subjects from single studies.\n",
    "\n",
    "Any given subject will have one row per study they participated in. The subject_id in the last column of this view is the same as the `id` in the first column of the Subjects endpoint results. You can use this to combine information across endpoints, which is covered near the end of the <a href=\"../BuildingACohort/#merging-results-across-endpoints\">Cohort Building workflow</a> notebook.\n",
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47372646",
   "metadata": {},
   "source": [
    "### diagnosis\n",
    "\n",
    "The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "75029795",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4270 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: a0b30464-8dc1-4025-94d6-a57183878bc4\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 1756\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagnosisresults = myquery.diagnosis.run()\n",
    "diagnosisresults #1756"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2f80fa21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3845 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>morphology</th>\n",
       "      <th>stage</th>\n",
       "      <th>grade</th>\n",
       "      <th>method_of_diagnosis</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0607a306-8b24-5f44-bf0d-fbfcc9985b8a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0607a306-8b24-5f4...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>18383.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-08-0358</td>\n",
       "      <td>ec504558-e96a-44e9-835a-7b12680681ed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3c999848-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3c999848-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>19879.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3L-02704</td>\n",
       "      <td>104c2847-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>65dfa780-27c2-53e8-8112-df7ef3f6cf13</td>\n",
       "      <td>[{'system': 'GDC', 'value': '65dfa780-27c2-53e...</td>\n",
       "      <td>Oligodendroglioma, anaplastic</td>\n",
       "      <td>19149.0</td>\n",
       "      <td>9451/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-QH-A6CU</td>\n",
       "      <td>76cb16f9-37be-450e-b60a-5f0d0f1ec2bc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7336271c-f7db-5bd3-a590-7b7e3340bf9c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7336271c-f7db-5bd...</td>\n",
       "      <td>Oligodendroglioma, NOS</td>\n",
       "      <td>13968.0</td>\n",
       "      <td>9450/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7472</td>\n",
       "      <td>482d4078-7576-48cf-bf7c-29014935fd47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7eacc35e-fe59-58d6-936a-a92d72fb4f2b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7eacc35e-fe59-58d...</td>\n",
       "      <td>Astrocytoma, anaplastic</td>\n",
       "      <td>17692.0</td>\n",
       "      <td>9401/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-QH-A6XC</td>\n",
       "      <td>62088274-a45b-4616-b086-b0ddd64895f0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>d091e91b-ff5e-11e9-9a07-0a80fada099c</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'd091e91b-ff5e-11e...</td>\n",
       "      <td>Ependymoma, NOS</td>\n",
       "      <td>4188.0</td>\n",
       "      <td>9391/3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>G2</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>C34440</td>\n",
       "      <td>d08d8323-ff5e-11e9-9a07-0a80fada099c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>d09210a8-ff5e-11e9-9a07-0a80fada099c</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'd09210a8-ff5e-11e...</td>\n",
       "      <td>Glioma, NOS</td>\n",
       "      <td>3917.0</td>\n",
       "      <td>9400/3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>G1</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>C38991</td>\n",
       "      <td>d08d93b7-ff5e-11e9-9a07-0a80fada099c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>db947c5a-b0c8-5e53-93f1-d24953c8bd0a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'db947c5a-b0c8-5e5...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>15964.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-02-2483</td>\n",
       "      <td>a2ac9937-f351-4d78-9261-264bf6c21e0c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>e23338b6-3e43-577d-92a5-a4561610b305</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e23338b6-3e43-577...</td>\n",
       "      <td>Oligodendroglioma, anaplastic</td>\n",
       "      <td>8752.0</td>\n",
       "      <td>9451/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-6408</td>\n",
       "      <td>14375753-a69f-41e0-b910-2bb8e84cc43a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>f0bb2fb4-506b-567b-9359-7f429967cd9e</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f0bb2fb4-506b-567...</td>\n",
       "      <td>Oligodendroglioma, anaplastic</td>\n",
       "      <td>21341.0</td>\n",
       "      <td>9451/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-E1-A7YQ</td>\n",
       "      <td>75dbc8fb-4db8-4764-824c-eccf3a223884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0607a306-8b24-5f44-bf0d-fbfcc9985b8a   \n",
       "1   3c999848-2848-11ec-b712-0a4e2186f121   \n",
       "2   65dfa780-27c2-53e8-8112-df7ef3f6cf13   \n",
       "3   7336271c-f7db-5bd3-a590-7b7e3340bf9c   \n",
       "4   7eacc35e-fe59-58d6-936a-a92d72fb4f2b   \n",
       "..                                   ...   \n",
       "95  d091e91b-ff5e-11e9-9a07-0a80fada099c   \n",
       "96  d09210a8-ff5e-11e9-9a07-0a80fada099c   \n",
       "97  db947c5a-b0c8-5e53-93f1-d24953c8bd0a   \n",
       "98  e23338b6-3e43-577d-92a5-a4561610b305   \n",
       "99  f0bb2fb4-506b-567b-9359-7f429967cd9e   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '0607a306-8b24-5f4...   \n",
       "1   [{'system': 'PDC', 'value': '3c999848-2848-11e...   \n",
       "2   [{'system': 'GDC', 'value': '65dfa780-27c2-53e...   \n",
       "3   [{'system': 'GDC', 'value': '7336271c-f7db-5bd...   \n",
       "4   [{'system': 'GDC', 'value': '7eacc35e-fe59-58d...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'PDC', 'value': 'd091e91b-ff5e-11e...   \n",
       "96  [{'system': 'PDC', 'value': 'd09210a8-ff5e-11e...   \n",
       "97  [{'system': 'GDC', 'value': 'db947c5a-b0c8-5e5...   \n",
       "98  [{'system': 'GDC', 'value': 'e23338b6-3e43-577...   \n",
       "99  [{'system': 'GDC', 'value': 'f0bb2fb4-506b-567...   \n",
       "\n",
       "                primary_diagnosis  age_at_diagnosis morphology         stage  \\\n",
       "0                    Glioblastoma           18383.0     9440/3          None   \n",
       "1                    Glioblastoma           19879.0     9440/3  Not Reported   \n",
       "2   Oligodendroglioma, anaplastic           19149.0     9451/3          None   \n",
       "3          Oligodendroglioma, NOS           13968.0     9450/3          None   \n",
       "4         Astrocytoma, anaplastic           17692.0     9401/3          None   \n",
       "..                            ...               ...        ...           ...   \n",
       "95                Ependymoma, NOS            4188.0     9391/3       Unknown   \n",
       "96                    Glioma, NOS            3917.0     9400/3       Unknown   \n",
       "97                   Glioblastoma           15964.0     9440/3          None   \n",
       "98  Oligodendroglioma, anaplastic            8752.0     9451/3          None   \n",
       "99  Oligodendroglioma, anaplastic           21341.0     9451/3          None   \n",
       "\n",
       "           grade method_of_diagnosis    subject_id  \\\n",
       "0   not reported                None  TCGA-08-0358   \n",
       "1   Not Reported                         C3L-02704   \n",
       "2   not reported                None  TCGA-QH-A6CU   \n",
       "3   not reported                None  TCGA-HT-7472   \n",
       "4   not reported                None  TCGA-QH-A6XC   \n",
       "..           ...                 ...           ...   \n",
       "95            G2        Not Reported        C34440   \n",
       "96            G1        Not Reported        C38991   \n",
       "97  not reported                None  TCGA-02-2483   \n",
       "98  not reported                None  TCGA-DU-6408   \n",
       "99  not reported                None  TCGA-E1-A7YQ   \n",
       "\n",
       "                      researchsubject_id  \n",
       "0   ec504558-e96a-44e9-835a-7b12680681ed  \n",
       "1   104c2847-2139-11ea-aee1-0e1aae319e49  \n",
       "2   76cb16f9-37be-450e-b60a-5f0d0f1ec2bc  \n",
       "3   482d4078-7576-48cf-bf7c-29014935fd47  \n",
       "4   62088274-a45b-4616-b086-b0ddd64895f0  \n",
       "..                                   ...  \n",
       "95  d08d8323-ff5e-11e9-9a07-0a80fada099c  \n",
       "96  d08d93b7-ff5e-11e9-9a07-0a80fada099c  \n",
       "97  a2ac9937-f351-4d78-9261-264bf6c21e0c  \n",
       "98  14375753-a69f-41e0-b910-2bb8e84cc43a  \n",
       "99  75dbc8fb-4db8-4764-824c-eccf3a223884  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "diagnosisresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6008331",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Diagnosis Field Definitions</h3>\n",
    "\n",
    "<i>A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>primary_diagnosis:</b> The diagnosis instance that qualified a subject for inclusion on a ResearchProject.</li>\n",
    "  <li><b>age_at_diagnosis:</b> The age in days of the individual at the time of diagnosis.</li>\n",
    "  <li><b>morphology:</b> Code that represents the histology of the disease using the third edition of the <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a>, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms.</li>\n",
    "  <li><b>stage:</b> The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body. Different diseases may use different staging criteria, please refer to the originating data source to see what staging system is reported</li>\n",
    "  <li><b>grade:</b> The degree of abnormality of cancer cells, a measure of differentiation, the extent to which cancer cells are similar in appearance and function to healthy cells of the same tissue type. The degree of differentiation often relates to the clinical behavior of the particular tumor. Based on the microscopic findings, tumor grade is commonly described by one of four degrees of severity. Histopathologic grade of a tumor may be used to plan treatment and estimate the future course, outcome, and overall prognosis of disease. Certain types of cancers, such as soft tissue sarcoma, primary brain tumors, lymphomas, and breast have special grading systems.</li>\n",
    "  <li><b>method_of_diagnosis:</b> The method used to confirm the patients malignant diagnosis.</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5855d4dd",
   "metadata": {},
   "source": [
    "### treatment\n",
    "\n",
    "The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "964fb1ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3801 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 6956c77e-9853-4adc-afc7-9a8c848df311\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 2386\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatmentresults = myquery.treatment.run()\n",
    "treatmentresults #2386"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d81bdba0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3768 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>days_to_treatment_start</th>\n",
       "      <th>days_to_treatment_end</th>\n",
       "      <th>therapeutic_agent</th>\n",
       "      <th>treatment_anatomic_site</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_end_reason</th>\n",
       "      <th>number_of_cycles</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>researchsubject_diagnosis_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>01e3b1d2-99f0-5d32-9d1b-594429011f50</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01e3b1d2-99f0-5d3...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-FG-A70Z</td>\n",
       "      <td>98e82e98-341a-4bca-92c0-23c65c1bf94f</td>\n",
       "      <td>6db74e47-a996-5463-853a-6416b46a263a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>08b045d6-5d7c-5ae5-94ba-3e07a239f555</td>\n",
       "      <td>[{'system': 'GDC', 'value': '08b045d6-5d7c-5ae...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-A6S3</td>\n",
       "      <td>1e713e76-064d-46dc-a53c-20a17eb68341</td>\n",
       "      <td>a8065153-05ac-5138-91ff-cfd23a8a3459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>104f8ff0-48af-554a-9bfe-e5ed6b275bc9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '104f8ff0-48af-554...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-12-0821</td>\n",
       "      <td>eae44c1c-1628-4b58-8b90-d3372e3577d5</td>\n",
       "      <td>01af5bbe-5ce1-563c-934e-daa5124fb775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18c95fab-389e-5c64-949c-fbd2dd0bb2fe</td>\n",
       "      <td>[{'system': 'GDC', 'value': '18c95fab-389e-5c6...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-08-0350</td>\n",
       "      <td>dc624b20-2854-4235-ae7e-672156ee30e3</td>\n",
       "      <td>1b04b930-35a8-514d-8a99-8bac2010dc21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26a729a6-dcf9-526d-a977-8bd24f17f5bf</td>\n",
       "      <td>[{'system': 'GDC', 'value': '26a729a6-dcf9-526...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-A7TI</td>\n",
       "      <td>0efaf8f9-177c-4ffb-b03f-f0316adf5dc1</td>\n",
       "      <td>472558de-4cbd-52cc-b03e-3e80f8275eee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>d194224d-2bfe-5fcf-a053-4ebe00b0c65a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd194224d-2bfe-5fc...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-FG-5963</td>\n",
       "      <td>68f550ca-3a33-45e4-b1c9-14f99e543263</td>\n",
       "      <td>df157cdd-174a-539c-ae8f-698a3b3a46a2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>e6ffa328-c1be-5943-aae7-d9f0b741ac3e</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e6ffa328-c1be-594...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A6WI</td>\n",
       "      <td>0b283f95-dbd4-4c36-9201-2198a90ee0d6</td>\n",
       "      <td>63c1ba76-12ba-5d0c-86ed-dccdfec708e0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>f05078ba-26c4-5c33-830d-e44d95132310</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f05078ba-26c4-5c3...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A6WO</td>\n",
       "      <td>a2202233-4a83-4163-9532-b6edb9941453</td>\n",
       "      <td>6b6695a0-d3ef-56a1-b980-d98871774b00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>f85946e1-2a17-50d4-ad4d-d0b6f0b98b02</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f85946e1-2a17-50d...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-76-6661</td>\n",
       "      <td>19dd2c8f-05ca-44e8-8c19-bac0327f2ea9</td>\n",
       "      <td>eed788c8-1c8c-59e0-a038-c9cb74800396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>fbf4b995-2f75-5de8-a31e-085119d2de62</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'fbf4b995-2f75-5de...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7694</td>\n",
       "      <td>49a87025-d0a4-44b3-b6ad-563bf9651093</td>\n",
       "      <td>5d190050-b7b2-584c-92b8-19e2a7b6be92</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   01e3b1d2-99f0-5d32-9d1b-594429011f50   \n",
       "1   08b045d6-5d7c-5ae5-94ba-3e07a239f555   \n",
       "2   104f8ff0-48af-554a-9bfe-e5ed6b275bc9   \n",
       "3   18c95fab-389e-5c64-949c-fbd2dd0bb2fe   \n",
       "4   26a729a6-dcf9-526d-a977-8bd24f17f5bf   \n",
       "..                                   ...   \n",
       "95  d194224d-2bfe-5fcf-a053-4ebe00b0c65a   \n",
       "96  e6ffa328-c1be-5943-aae7-d9f0b741ac3e   \n",
       "97  f05078ba-26c4-5c33-830d-e44d95132310   \n",
       "98  f85946e1-2a17-50d4-ad4d-d0b6f0b98b02   \n",
       "99  fbf4b995-2f75-5de8-a31e-085119d2de62   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '01e3b1d2-99f0-5d3...   \n",
       "1   [{'system': 'GDC', 'value': '08b045d6-5d7c-5ae...   \n",
       "2   [{'system': 'GDC', 'value': '104f8ff0-48af-554...   \n",
       "3   [{'system': 'GDC', 'value': '18c95fab-389e-5c6...   \n",
       "4   [{'system': 'GDC', 'value': '26a729a6-dcf9-526...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': 'd194224d-2bfe-5fc...   \n",
       "96  [{'system': 'GDC', 'value': 'e6ffa328-c1be-594...   \n",
       "97  [{'system': 'GDC', 'value': 'f05078ba-26c4-5c3...   \n",
       "98  [{'system': 'GDC', 'value': 'f85946e1-2a17-50d...   \n",
       "99  [{'system': 'GDC', 'value': 'fbf4b995-2f75-5de...   \n",
       "\n",
       "                 treatment_type treatment_outcome days_to_treatment_start  \\\n",
       "0        Radiation Therapy, NOS              None                    None   \n",
       "1        Radiation Therapy, NOS              None                    None   \n",
       "2   Pharmaceutical Therapy, NOS              None                    None   \n",
       "3        Radiation Therapy, NOS              None                    None   \n",
       "4   Pharmaceutical Therapy, NOS              None                    None   \n",
       "..                          ...               ...                     ...   \n",
       "95       Radiation Therapy, NOS              None                    None   \n",
       "96  Pharmaceutical Therapy, NOS              None                    None   \n",
       "97  Pharmaceutical Therapy, NOS              None                    None   \n",
       "98  Pharmaceutical Therapy, NOS              None                    None   \n",
       "99  Pharmaceutical Therapy, NOS              None                    None   \n",
       "\n",
       "   days_to_treatment_end therapeutic_agent treatment_anatomic_site  \\\n",
       "0                   None              None                    None   \n",
       "1                   None              None                    None   \n",
       "2                   None              None                    None   \n",
       "3                   None              None                    None   \n",
       "4                   None              None                    None   \n",
       "..                   ...               ...                     ...   \n",
       "95                  None              None                    None   \n",
       "96                  None              None                    None   \n",
       "97                  None              None                    None   \n",
       "98                  None              None                    None   \n",
       "99                  None              None                    None   \n",
       "\n",
       "   treatment_effect treatment_end_reason number_of_cycles    subject_id  \\\n",
       "0              None                 None             None  TCGA-FG-A70Z   \n",
       "1              None                 None             None  TCGA-DU-A6S3   \n",
       "2              None                 None             None  TCGA-12-0821   \n",
       "3              None                 None             None  TCGA-08-0350   \n",
       "4              None                 None             None  TCGA-DU-A7TI   \n",
       "..              ...                  ...              ...           ...   \n",
       "95             None                 None             None  TCGA-FG-5963   \n",
       "96             None                 None             None  TCGA-S9-A6WI   \n",
       "97             None                 None             None  TCGA-S9-A6WO   \n",
       "98             None                 None             None  TCGA-76-6661   \n",
       "99             None                 None             None  TCGA-HT-7694   \n",
       "\n",
       "                      researchsubject_id          researchsubject_diagnosis_id  \n",
       "0   98e82e98-341a-4bca-92c0-23c65c1bf94f  6db74e47-a996-5463-853a-6416b46a263a  \n",
       "1   1e713e76-064d-46dc-a53c-20a17eb68341  a8065153-05ac-5138-91ff-cfd23a8a3459  \n",
       "2   eae44c1c-1628-4b58-8b90-d3372e3577d5  01af5bbe-5ce1-563c-934e-daa5124fb775  \n",
       "3   dc624b20-2854-4235-ae7e-672156ee30e3  1b04b930-35a8-514d-8a99-8bac2010dc21  \n",
       "4   0efaf8f9-177c-4ffb-b03f-f0316adf5dc1  472558de-4cbd-52cc-b03e-3e80f8275eee  \n",
       "..                                   ...                                   ...  \n",
       "95  68f550ca-3a33-45e4-b1c9-14f99e543263  df157cdd-174a-539c-ae8f-698a3b3a46a2  \n",
       "96  0b283f95-dbd4-4c36-9201-2198a90ee0d6  63c1ba76-12ba-5d0c-86ed-dccdfec708e0  \n",
       "97  a2202233-4a83-4163-9532-b6edb9941453  6b6695a0-d3ef-56a1-b980-d98871774b00  \n",
       "98  19dd2c8f-05ca-44e8-8c19-bac0327f2ea9  eed788c8-1c8c-59e0-a038-c9cb74800396  \n",
       "99  49a87025-d0a4-44b3-b6ad-563bf9651093  5d190050-b7b2-584c-92b8-19e2a7b6be92  \n",
       "\n",
       "[100 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "treatmentresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30804682",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Treatment Field Definitions</h3>\n",
    "\n",
    "<i><i> Medication administration or other treatment types. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>treatment_type:</b> The treatment type including medication/therapeutics or other procedures.</li>\n",
    "  <li><b>treatment_outcome:</b> The final outcome of the treatment.</li>\n",
    "  <li><b>days_to_treatment_start:</b> The timepoint at which the treatment started.</li>\n",
    "  <li><b>days_to_treatment_end:</b>The timepoint at which the treatment ended. </li>\n",
    "  <li><b>therapeutic_agent:</b> One or more therapeutic agents as part of this treatment.</li>\n",
    "  <li><b>treatment_anatomic_site:</b> The anatomical site that the treatment targets.</li>\n",
    "  <li><b>treatment_effect:</b>The effect of a treatment on the diagnosis or tumor. </li>\n",
    "  <li><b>treatment_end_reason:</b>The reason the treatment ended. </li>\n",
    "  <li><b>number_of_cycles:</b>The number of treatment cycles the subject received. </li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the researchsubject. Can be joined to the `id` field from researchsubject results</li>\n",
    "  <li><b>researchsubject_diagnosis_id:</b> An identifier for the diagnosis. Can be joined to the `id` field from diagnosis results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c00bec8",
   "metadata": {},
   "source": [
    "### specimen\n",
    "\n",
    "We can use this same query to see what specimens are available for brain tissue at the CDA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "51960eed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3966 ms\n",
      "\n",
      "            QueryID: 663f60e8-5922-4872-9adc-64dbc0a1cac8\n",
      "            \n",
      "            Offset: 0\n",
      "            Count: 100\n",
      "            Total Row Count: 39201\n",
      "            More pages: True\n",
      "            \n"
     ]
    }
   ],
   "source": [
    "specimenresults =  myquery.specimen.run()\n",
    "print(specimenresults) #39201"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f254547",
   "metadata": {},
   "source": [
    "Nearly 40,000 specimens meet our search criteria! We would typically expect this number to be much larger than our number of subjects or researchsubjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. Since we didn't specify any further filters, our results will return all of these as seperate speciments. Let's look at a few:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c55b0128",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>days_to_collection</th>\n",
       "      <th>primary_disease_type</th>\n",
       "      <th>anatomical_site</th>\n",
       "      <th>source_material_type</th>\n",
       "      <th>specimen_type</th>\n",
       "      <th>derived_from_specimen</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>008a69a1-9590-43a4-b5fd-14ab663721d5</td>\n",
       "      <td>[{'system': 'GDC', 'value': '008a69a1-9590-43a...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>4a7395a9-b4b3-438f-afc7-25c634ba5c75</td>\n",
       "      <td>TCGA-08-0355</td>\n",
       "      <td>bb8230b5-f984-4d28-9f10-b3c89bab1c28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00c7f311-41b3-474e-91d2-0cc234846336</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00c7f311-41b3-474...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>4cc6f946-09a0-403e-85c0-01f01c208113</td>\n",
       "      <td>TCGA-CS-6669</td>\n",
       "      <td>dc5bdd21-b23f-400e-bbe2-4c98a517fc9e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00ce2c37-cc9f-4868-8c9d-a76b0741651a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00ce2c37-cc9f-486...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>33d17296-27ea-4e0f-86e5-88c14d2775a3</td>\n",
       "      <td>TCGA-02-0086</td>\n",
       "      <td>8e3fbea7-d22e-40fc-8031-58fe6fe091bb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>01729f63-439c-43c1-8ee9-943f059c542c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01729f63-439c-43c...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>8d859a3c-c509-41df-a485-185746721d21</td>\n",
       "      <td>TCGA-12-1598</td>\n",
       "      <td>2b90648c-c99e-4d4d-8d9b-67ff2c4e6115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>01cee30c-fa9a-4ba8-b810-204bd1327004</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01cee30c-fa9a-4ba...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>63251572-4138-4664-85bd-2939293c1f5e</td>\n",
       "      <td>TCGA-06-0190</td>\n",
       "      <td>74139255-a635-4c87-814d-3dd04ed630a8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>38f4e4b7-a4cc-4d87-bfd0-c4356d92086e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '38f4e4b7-a4cc-4d8...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>261e8179-6213-45d4-b132-3e39ea8b86fb</td>\n",
       "      <td>TCGA-HT-7467</td>\n",
       "      <td>f4a1fc14-0d32-4e77-a8a3-fd8fea82188a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3e9f8f4b-bd8b-4b46-9ed6-b95518ab80ad</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3e9f8f4b-bd8b-4b4...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>cf7f3be6-3666-41a9-93a5-45c415087b98</td>\n",
       "      <td>TCGA-87-5896</td>\n",
       "      <td>ed879063-e603-4151-bf5e-22a4ee210281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3f1d6889-997a-4f7c-b74b-3def8db6db57</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f1d6889-997a-4f7...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>2239ec04-c589-448f-817b-7ef99aa65846</td>\n",
       "      <td>TCGA-02-0048</td>\n",
       "      <td>24eaab84-5d9c-4c12-af8b-32f4c6d8f81e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>3faaa60f-7fe5-4b3c-acf2-2915b6f177c4</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3faaa60f-7fe5-4b3...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>2fff9e3d-6b6c-566e-8a9d-f44647d76e4e</td>\n",
       "      <td>TCGA-P5-A731</td>\n",
       "      <td>335bc811-8bdc-4f66-bdd0-703b7cc7464b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>3fd3aafa-7af1-47a6-bd82-1e3598541a02</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3fd3aafa-7af1-47a...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>analyte</td>\n",
       "      <td>6f4a7910-4c64-477b-915c-c58f6b69fae1</td>\n",
       "      <td>TCGA-08-0390</td>\n",
       "      <td>108a8d8e-1ddb-4c0d-a05b-71573b33b9aa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   008a69a1-9590-43a4-b5fd-14ab663721d5   \n",
       "1   00c7f311-41b3-474e-91d2-0cc234846336   \n",
       "2   00ce2c37-cc9f-4868-8c9d-a76b0741651a   \n",
       "3   01729f63-439c-43c1-8ee9-943f059c542c   \n",
       "4   01cee30c-fa9a-4ba8-b810-204bd1327004   \n",
       "..                                   ...   \n",
       "95  38f4e4b7-a4cc-4d87-bfd0-c4356d92086e   \n",
       "96  3e9f8f4b-bd8b-4b46-9ed6-b95518ab80ad   \n",
       "97  3f1d6889-997a-4f7c-b74b-3def8db6db57   \n",
       "98  3faaa60f-7fe5-4b3c-acf2-2915b6f177c4   \n",
       "99  3fd3aafa-7af1-47a6-bd82-1e3598541a02   \n",
       "\n",
       "                                           identifier associated_project  \\\n",
       "0   [{'system': 'GDC', 'value': '008a69a1-9590-43a...           TCGA-GBM   \n",
       "1   [{'system': 'GDC', 'value': '00c7f311-41b3-474...           TCGA-LGG   \n",
       "2   [{'system': 'GDC', 'value': '00ce2c37-cc9f-486...           TCGA-GBM   \n",
       "3   [{'system': 'GDC', 'value': '01729f63-439c-43c...           TCGA-GBM   \n",
       "4   [{'system': 'GDC', 'value': '01cee30c-fa9a-4ba...           TCGA-GBM   \n",
       "..                                                ...                ...   \n",
       "95  [{'system': 'GDC', 'value': '38f4e4b7-a4cc-4d8...           TCGA-LGG   \n",
       "96  [{'system': 'GDC', 'value': '3e9f8f4b-bd8b-4b4...           TCGA-GBM   \n",
       "97  [{'system': 'GDC', 'value': '3f1d6889-997a-4f7...           TCGA-GBM   \n",
       "98  [{'system': 'GDC', 'value': '3faaa60f-7fe5-4b3...           TCGA-LGG   \n",
       "99  [{'system': 'GDC', 'value': '3fd3aafa-7af1-47a...           TCGA-GBM   \n",
       "\n",
       "    days_to_collection primary_disease_type anatomical_site  \\\n",
       "0                  NaN              Gliomas            None   \n",
       "1                  NaN              Gliomas            None   \n",
       "2                  NaN              Gliomas            None   \n",
       "3                  NaN              Gliomas            None   \n",
       "4                  NaN              Gliomas            None   \n",
       "..                 ...                  ...             ...   \n",
       "95                 NaN              Gliomas            None   \n",
       "96                 NaN              Gliomas            None   \n",
       "97                 NaN              Gliomas            None   \n",
       "98                 0.0              Gliomas            None   \n",
       "99                 NaN              Gliomas            None   \n",
       "\n",
       "    source_material_type specimen_type                 derived_from_specimen  \\\n",
       "0    Solid Tissue Normal       aliquot  4a7395a9-b4b3-438f-afc7-25c634ba5c75   \n",
       "1          Primary Tumor       aliquot  4cc6f946-09a0-403e-85c0-01f01c208113   \n",
       "2          Primary Tumor       aliquot  33d17296-27ea-4e0f-86e5-88c14d2775a3   \n",
       "3          Primary Tumor       aliquot  8d859a3c-c509-41df-a485-185746721d21   \n",
       "4   Blood Derived Normal       aliquot  63251572-4138-4664-85bd-2939293c1f5e   \n",
       "..                   ...           ...                                   ...   \n",
       "95         Primary Tumor       aliquot  261e8179-6213-45d4-b132-3e39ea8b86fb   \n",
       "96         Primary Tumor       aliquot  cf7f3be6-3666-41a9-93a5-45c415087b98   \n",
       "97         Primary Tumor       aliquot  2239ec04-c589-448f-817b-7ef99aa65846   \n",
       "98         Primary Tumor         slide  2fff9e3d-6b6c-566e-8a9d-f44647d76e4e   \n",
       "99         Primary Tumor       analyte  6f4a7910-4c64-477b-915c-c58f6b69fae1   \n",
       "\n",
       "      subject_id                    researchsubject_id  \n",
       "0   TCGA-08-0355  bb8230b5-f984-4d28-9f10-b3c89bab1c28  \n",
       "1   TCGA-CS-6669  dc5bdd21-b23f-400e-bbe2-4c98a517fc9e  \n",
       "2   TCGA-02-0086  8e3fbea7-d22e-40fc-8031-58fe6fe091bb  \n",
       "3   TCGA-12-1598  2b90648c-c99e-4d4d-8d9b-67ff2c4e6115  \n",
       "4   TCGA-06-0190  74139255-a635-4c87-814d-3dd04ed630a8  \n",
       "..           ...                                   ...  \n",
       "95  TCGA-HT-7467  f4a1fc14-0d32-4e77-a8a3-fd8fea82188a  \n",
       "96  TCGA-87-5896  ed879063-e603-4151-bf5e-22a4ee210281  \n",
       "97  TCGA-02-0048  24eaab84-5d9c-4c12-af8b-32f4c6d8f81e  \n",
       "98  TCGA-P5-A731  335bc811-8bdc-4f66-bdd0-703b7cc7464b  \n",
       "99  TCGA-08-0390  108a8d8e-1ddb-4c0d-a05b-71573b33b9aa  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specimenresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccb0f1c",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Specimen Field Definitions</h3>\n",
    "\n",
    "<i><i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>\n",
    " A given specimen will have only a single subject ID and a single research subject ID</i>\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>associated_project:</b> The Project associated with the specimen.</li>\n",
    "  <li><b>days_to_collection:</b> The number of days from the index date to either the date a sample was collected for a specific study or project, or the date a patient underwent a procedure (e.g. surgical resection) yielding a sample that was eventually used for research.</li>\n",
    "  <li><b>primary_disease_type:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "  <li><b>anatomical_site:</b> Per GDC Dictionary, the text term that represents the name of the primary disease site of the submitted tumor sample; recommend dropping tumor; biospecimen_anatomic_site.</li>\n",
    "  <li><b>source_material_type:</b> The general kind of material from which the specimen was derived, indicating the physical nature of the source material.</li>\n",
    "  <li><b>specimen_type:</b> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide.</li>\n",
    "  <li><b>derived_from_specimen:</b> A source/parent specimen from which this one was directly derived.</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n",
    "### file\n",
    "\n",
    "The file endpoint returns information about files that meet our search criteria, regardless of whether they are attached to subjects, research-subjects or specimens: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "94a0361f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4183 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: adbd32d8-8bfb-4934-ab74-582c91fa80bd\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 4549472\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.file.run() #4549472"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6006ec89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4187 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5e9a14f5-c5d8-4249-a7c6-5ab779544f83</td>\n",
       "      <td>[{'system': 'GDC', 'value': '5e9a14f5-c5d8-424...</td>\n",
       "      <td>50e0e4de-6ee2-4fa8-b7fe-5bb3947c563c.wxs.VarSc...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:5e9a14f5-c5d8-4249-a7c6-5ab77954...</td>\n",
       "      <td>162650</td>\n",
       "      <td>e0edbdec5cabeeabc08f38cbc99f32a2</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-DU-A7T6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>345b67b8-ec76-4cd3-9f26-25aa9e0c9c19</td>\n",
       "      <td>[{'system': 'GDC', 'value': '345b67b8-ec76-4cd...</td>\n",
       "      <td>TCGA-06-0881-01A-01-BS1.6f2cf2cc-dc8c-49ca-b36...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>drs://dg.4DFC:345b67b8-ec76-4cd3-9f26-25aa9e0c...</td>\n",
       "      <td>260020245</td>\n",
       "      <td>ce066d859c4c7d452e10ea048dc340af</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-06-0881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>027511f7-15e6-11ea-9bfa-0a42f3c845fe</td>\n",
       "      <td>[{'system': 'PDC', 'value': '027511f7-15e6-11e...</td>\n",
       "      <td>01CPTAC_GBM_W_PNNL_20190123_B1S1_f24.raw</td>\n",
       "      <td>Raw Mass Spectra</td>\n",
       "      <td>Proprietary</td>\n",
       "      <td>vendor-specific</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:027511f7-15e6-11ea-9bfa-0a42f3c8...</td>\n",
       "      <td>708638166</td>\n",
       "      <td>6e0470755a0ec6866519fd3a637f7cff</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3L-01834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3eb648c4-51d5-49e1-8c8e-596325c7deba</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3eb648c4-51d5-49e...</td>\n",
       "      <td>972ce08b-3d70-4e63-a572-b722905d6b4c.rna_seq.s...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Splice Junction Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:3eb648c4-51d5-49e1-8c8e-596325c7...</td>\n",
       "      <td>2229707</td>\n",
       "      <td>e2a4392ebdde3fbbfe79eff034de9031</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-DB-A64X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b82d3c39-39e6-4a1d-a5f5-1ffe4c3d452e</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'b82d3c39-39e6-4a1...</td>\n",
       "      <td>NCI-11Plex-19-F11-f10554.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:b82d3c39-39e6-4a1d-a5f5-1ffe4c3d...</td>\n",
       "      <td>169024251</td>\n",
       "      <td>9b2c7a26ea571db3f992fba668f2e306</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C907125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>a7590da5-eb4a-488a-bea0-6369a2167384</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a7590da5-eb4a-488...</td>\n",
       "      <td>ffec4a68-74c4-4d7e-bc42-c60aebba3881_noid_Red....</td>\n",
       "      <td>DNA Methylation</td>\n",
       "      <td>Masked Intensities</td>\n",
       "      <td>IDAT</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:a7590da5-eb4a-488a-bea0-6369a216...</td>\n",
       "      <td>8095293</td>\n",
       "      <td>e083ac98e5a6573458ec797ea0eab229</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-HT-8109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>0192d6c7-fb9b-4cc7-936a-4e7c9a766f7b</td>\n",
       "      <td>[{'system': 'PDC', 'value': '0192d6c7-fb9b-4cc...</td>\n",
       "      <td>NCI-11Plex-9-F3-z11239.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:0192d6c7-fb9b-4cc7-936a-4e7c9a76...</td>\n",
       "      <td>238729165</td>\n",
       "      <td>6a62ad00d9368abf036a22723a0581bd</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C25830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>447da7db-c642-455a-980c-2350d7360e42</td>\n",
       "      <td>[{'system': 'PDC', 'value': '447da7db-c642-455...</td>\n",
       "      <td>10CPTAC_GBM_W_PNNL_20190501_B3S2_f22.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:447da7db-c642-455a-980c-2350d736...</td>\n",
       "      <td>191567974</td>\n",
       "      <td>92f7b3817f3b58ec61041374abe35013</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3L-02900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>c1a2c48b-db96-4fb6-afeb-ee4846498cea</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c1a2c48b-db96-4fb...</td>\n",
       "      <td>57b638b7-09bf-4607-8d9b-9bd59694aa4e_noid_Grn....</td>\n",
       "      <td>DNA Methylation</td>\n",
       "      <td>Masked Intensities</td>\n",
       "      <td>IDAT</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>drs://dg.4DFC:c1a2c48b-db96-4fb6-afeb-ee484649...</td>\n",
       "      <td>13676217</td>\n",
       "      <td>6dadcd6c38dd0e1a2f8b762a2ff2f663</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs001287</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-02256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>2458c146-74ef-4c2a-b2b4-1bb6ab5292a9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2458c146-74ef-4c2...</td>\n",
       "      <td>fd0dc2d4-4a8e-4f37-8e1c-320193ace083.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>drs://dg.4DFC:2458c146-74ef-4c2a-b2b4-1bb6ab52...</td>\n",
       "      <td>4242748</td>\n",
       "      <td>47ad57ab57fb8688a3c5e625c26c1cdb</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-19-0957</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   5e9a14f5-c5d8-4249-a7c6-5ab779544f83   \n",
       "1   345b67b8-ec76-4cd3-9f26-25aa9e0c9c19   \n",
       "2   027511f7-15e6-11ea-9bfa-0a42f3c845fe   \n",
       "3   3eb648c4-51d5-49e1-8c8e-596325c7deba   \n",
       "4   b82d3c39-39e6-4a1d-a5f5-1ffe4c3d452e   \n",
       "..                                   ...   \n",
       "95  a7590da5-eb4a-488a-bea0-6369a2167384   \n",
       "96  0192d6c7-fb9b-4cc7-936a-4e7c9a766f7b   \n",
       "97  447da7db-c642-455a-980c-2350d7360e42   \n",
       "98  c1a2c48b-db96-4fb6-afeb-ee4846498cea   \n",
       "99  2458c146-74ef-4c2a-b2b4-1bb6ab5292a9   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '5e9a14f5-c5d8-424...   \n",
       "1   [{'system': 'GDC', 'value': '345b67b8-ec76-4cd...   \n",
       "2   [{'system': 'PDC', 'value': '027511f7-15e6-11e...   \n",
       "3   [{'system': 'GDC', 'value': '3eb648c4-51d5-49e...   \n",
       "4   [{'system': 'PDC', 'value': 'b82d3c39-39e6-4a1...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': 'a7590da5-eb4a-488...   \n",
       "96  [{'system': 'PDC', 'value': '0192d6c7-fb9b-4cc...   \n",
       "97  [{'system': 'PDC', 'value': '447da7db-c642-455...   \n",
       "98  [{'system': 'GDC', 'value': 'c1a2c48b-db96-4fb...   \n",
       "99  [{'system': 'GDC', 'value': '2458c146-74ef-4c2...   \n",
       "\n",
       "                                                label  \\\n",
       "0   50e0e4de-6ee2-4fa8-b7fe-5bb3947c563c.wxs.VarSc...   \n",
       "1   TCGA-06-0881-01A-01-BS1.6f2cf2cc-dc8c-49ca-b36...   \n",
       "2            01CPTAC_GBM_W_PNNL_20190123_B1S1_f24.raw   \n",
       "3   972ce08b-3d70-4e63-a572-b722905d6b4c.rna_seq.s...   \n",
       "4                    NCI-11Plex-19-F11-f10554.mzML.gz   \n",
       "..                                                ...   \n",
       "95  ffec4a68-74c4-4d7e-bc42-c60aebba3881_noid_Red....   \n",
       "96                     NCI-11Plex-9-F3-z11239.mzML.gz   \n",
       "97       10CPTAC_GBM_W_PNNL_20190501_B3S2_f22.mzML.gz   \n",
       "98  57b638b7-09bf-4607-8d9b-9bd59694aa4e_noid_Grn....   \n",
       "99  fd0dc2d4-4a8e-4f37-8e1c-320193ace083.rna_seq.a...   \n",
       "\n",
       "                  data_category                       data_type  \\\n",
       "0   Simple Nucleotide Variation      Annotated Somatic Mutation   \n",
       "1                   Biospecimen                     Slide Image   \n",
       "2              Raw Mass Spectra                     Proprietary   \n",
       "3       Transcriptome Profiling  Splice Junction Quantification   \n",
       "4        Processed Mass Spectra                   Open Standard   \n",
       "..                          ...                             ...   \n",
       "95              DNA Methylation              Masked Intensities   \n",
       "96       Processed Mass Spectra                   Open Standard   \n",
       "97       Processed Mass Spectra                   Open Standard   \n",
       "98              DNA Methylation              Masked Intensities   \n",
       "99      Transcriptome Profiling  Gene Expression Quantification   \n",
       "\n",
       "        file_format                                 associated_project  \\\n",
       "0               MAF                                           TCGA-LGG   \n",
       "1               SVS                                           TCGA-GBM   \n",
       "2   vendor-specific                                   CPTAC3-Discovery   \n",
       "3               TSV                                           TCGA-LGG   \n",
       "4              mzML  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "..              ...                                                ...   \n",
       "95             IDAT                                           TCGA-LGG   \n",
       "96             mzML  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "97             mzML                                   CPTAC3-Discovery   \n",
       "98             IDAT                                            CPTAC-3   \n",
       "99              TSV                                           TCGA-GBM   \n",
       "\n",
       "                                              drs_uri  byte_size  \\\n",
       "0   drs://dg.4DFC:5e9a14f5-c5d8-4249-a7c6-5ab77954...     162650   \n",
       "1   drs://dg.4DFC:345b67b8-ec76-4cd3-9f26-25aa9e0c...  260020245   \n",
       "2   drs://dg.4DFC:027511f7-15e6-11ea-9bfa-0a42f3c8...  708638166   \n",
       "3   drs://dg.4DFC:3eb648c4-51d5-49e1-8c8e-596325c7...    2229707   \n",
       "4   drs://dg.4DFC:b82d3c39-39e6-4a1d-a5f5-1ffe4c3d...  169024251   \n",
       "..                                                ...        ...   \n",
       "95  drs://dg.4DFC:a7590da5-eb4a-488a-bea0-6369a216...    8095293   \n",
       "96  drs://dg.4DFC:0192d6c7-fb9b-4cc7-936a-4e7c9a76...  238729165   \n",
       "97  drs://dg.4DFC:447da7db-c642-455a-980c-2350d736...  191567974   \n",
       "98  drs://dg.4DFC:c1a2c48b-db96-4fb6-afeb-ee484649...   13676217   \n",
       "99  drs://dg.4DFC:2458c146-74ef-4c2a-b2b4-1bb6ab52...    4242748   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   e0edbdec5cabeeabc08f38cbc99f32a2       Genomic             None   \n",
       "1   ce066d859c4c7d452e10ea048dc340af       Genomic             None   \n",
       "2   6e0470755a0ec6866519fd3a637f7cff     Proteomic             None   \n",
       "3   e2a4392ebdde3fbbfe79eff034de9031       Genomic             None   \n",
       "4   9b2c7a26ea571db3f992fba668f2e306     Proteomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  e083ac98e5a6573458ec797ea0eab229       Genomic             None   \n",
       "96  6a62ad00d9368abf036a22723a0581bd     Proteomic             None   \n",
       "97  92f7b3817f3b58ec61041374abe35013     Proteomic             None   \n",
       "98  6dadcd6c38dd0e1a2f8b762a2ff2f663       Genomic             None   \n",
       "99  47ad57ab57fb8688a3c5e625c26c1cdb       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number imaging_series researchsubject_specimen_id  \\\n",
       "0                    None           None                               \n",
       "1                    None           None                               \n",
       "2                    None           None                               \n",
       "3                    None           None                               \n",
       "4                    None           None                               \n",
       "..                    ...            ...                         ...   \n",
       "95                   None           None                               \n",
       "96                   None           None                               \n",
       "97                   None           None                               \n",
       "98              phs001287           None                               \n",
       "99                   None           None                               \n",
       "\n",
       "   researchsubject_id    subject_id  \n",
       "0                      TCGA-DU-A7T6  \n",
       "1                      TCGA-06-0881  \n",
       "2                         C3L-01834  \n",
       "3                      TCGA-DB-A64X  \n",
       "4                           C907125  \n",
       "..                ...           ...  \n",
       "95                     TCGA-HT-8109  \n",
       "96                           C25830  \n",
       "97                        C3L-02900  \n",
       "98                        C3N-02256  \n",
       "99                     TCGA-19-0957  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileresults = myquery.file.run()\n",
    "fileresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46fcb3e4",
   "metadata": {},
   "source": [
    "As you might expect, searching file gives us a huge number of results. This is great if you are surveying what kind of data is available, but is less useful for getting a coherent cohort. \n",
    "\n",
    "A better way to get files for a specific cohort is to chain your queries together, which we cover in the next tutorial <a href=\"../Chaining\">Chaining Endpoints</a>: Combine information from multiple endpoints, and build And/Or/Like and other advanced query strings.\n",
    "\n",
    "Another useful way to look at high level information is to use our counts feature which returns summary information rather than the full search results. Check out the <a href=\"../DataSummaries\">Data Summaries tutorial</a> to try it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "549257b1",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
