{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "614a478f",
   "metadata": {},
   "source": [
    "# Cohort Building"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8004e87",
   "metadata": {},
   "source": [
    "**Example use case:** \n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" alt=\"alt_text\" align=\"left\"\n",
    "\twidth=\"150\" height=\"150\" />\n",
    "Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e45a9b",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "\n",
    "Before Julia does any work, she needs to import these functions cdapython.\n",
    "She'll also need to import [pandas](https://pandas.pydata.org/) to get nice dataframes.\n",
    "Finally, she tells cdapython to report it's version so she can be sure she's using the one she means to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5265d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "import pandas as pd \n",
    "print(cdapython.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75eef23e",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "391bc9a7",
   "metadata": {},
   "source": [
    "## Finding Search Terms\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef0dd8e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['File.id',\n",
       " 'File.identifier.system',\n",
       " 'File.identifier.value',\n",
       " 'File.label',\n",
       " 'File.data_category',\n",
       " 'File.data_type',\n",
       " 'File.file_format',\n",
       " 'File.associated_project',\n",
       " 'File.drs_uri',\n",
       " 'File.byte_size',\n",
       " 'File.checksum',\n",
       " 'File.data_modality',\n",
       " 'File.imaging_modality',\n",
       " 'File.dbgap_accession_number',\n",
       " 'File.imaging_series',\n",
       " 'id',\n",
       " 'identifier.system',\n",
       " 'identifier.value',\n",
       " 'species',\n",
       " 'sex',\n",
       " 'race',\n",
       " 'ethnicity',\n",
       " 'days_to_birth',\n",
       " 'subject_associated_project',\n",
       " 'vital_status',\n",
       " 'days_to_death',\n",
       " 'cause_of_death',\n",
       " 'ResearchSubject.id',\n",
       " 'ResearchSubject.identifier.system',\n",
       " 'ResearchSubject.identifier.value',\n",
       " 'ResearchSubject.member_of_research_project',\n",
       " 'ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles',\n",
       " 'ResearchSubject.Specimen.id',\n",
       " 'ResearchSubject.Specimen.identifier.system',\n",
       " 'ResearchSubject.Specimen.identifier.value',\n",
       " 'ResearchSubject.Specimen.associated_project',\n",
       " 'ResearchSubject.Specimen.days_to_collection',\n",
       " 'ResearchSubject.Specimen.primary_disease_type',\n",
       " 'ResearchSubject.Specimen.anatomical_site',\n",
       " 'ResearchSubject.Specimen.source_material_type',\n",
       " 'ResearchSubject.Specimen.specimen_type',\n",
       " 'ResearchSubject.Specimen.derived_from_specimen',\n",
       " 'ResearchSubject.Specimen.derived_from_subject']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05eba2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "536970c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list(filters=\"diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63b4cf0",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "\n",
    "To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retrieving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.\n",
    "    \n",
    "</div>\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Since Julia is interested specifically in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4527dde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Brain',\n",
       " 'Cervix',\n",
       " 'Head - Face Or Neck, Nos',\n",
       " 'Lymph Node(s) Paraaortic',\n",
       " 'Other',\n",
       " 'Pelvis',\n",
       " 'Spine',\n",
       " 'Unknown']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\").to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "740e5955",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Abdomen',\n",
       " 'Abdomen, Mediastinum',\n",
       " 'Abdomen, Pelvis',\n",
       " 'Adrenal Glands',\n",
       " 'Adrenal gland',\n",
       " 'Anus and anal canal',\n",
       " 'Base of tongue',\n",
       " 'Bile Duct',\n",
       " 'Bladder',\n",
       " 'Bones, joints and articular cartilage of limbs',\n",
       " 'Bones, joints and articular cartilage of other and unspecified sites',\n",
       " 'Brain',\n",
       " 'Breast',\n",
       " 'Bronchus and lung',\n",
       " 'Cervix',\n",
       " 'Cervix uteri',\n",
       " 'Chest',\n",
       " 'Chest-Abdomen-Pelvis, Leg, TSpine',\n",
       " 'Colon',\n",
       " 'Connective, subcutaneous and other soft tissues',\n",
       " 'Corpus uteri',\n",
       " 'Ear',\n",
       " 'Esophagus',\n",
       " 'Extremities',\n",
       " 'Eye and adnexa',\n",
       " 'Floor of mouth',\n",
       " 'Gallbladder',\n",
       " 'Gum',\n",
       " 'Head',\n",
       " 'Head and Neck',\n",
       " 'Head-Neck',\n",
       " 'Heart, mediastinum, and pleura',\n",
       " 'Hematopoietic and reticuloendothelial systems',\n",
       " 'Hypopharynx',\n",
       " 'Intraocular',\n",
       " 'Kidney',\n",
       " 'Larynx',\n",
       " 'Lip',\n",
       " 'Liver',\n",
       " 'Liver and intrahepatic bile ducts',\n",
       " 'Lung',\n",
       " 'Lung Phantom',\n",
       " 'Lymph nodes',\n",
       " 'Marrow, Blood',\n",
       " 'Meninges',\n",
       " 'Mesothelium',\n",
       " 'Nasal cavity and middle ear',\n",
       " 'Nasopharynx',\n",
       " 'Not Reported',\n",
       " 'Oropharynx',\n",
       " 'Other and ill-defined digestive organs',\n",
       " 'Other and ill-defined sites',\n",
       " 'Other and ill-defined sites in lip, oral cavity and pharynx',\n",
       " 'Other and ill-defined sites within respiratory system and intrathoracic organs',\n",
       " 'Other and unspecified female genital organs',\n",
       " 'Other and unspecified major salivary glands',\n",
       " 'Other and unspecified male genital organs',\n",
       " 'Other and unspecified parts of biliary tract',\n",
       " 'Other and unspecified parts of mouth',\n",
       " 'Other and unspecified parts of tongue',\n",
       " 'Other and unspecified urinary organs',\n",
       " 'Other endocrine glands and related structures',\n",
       " 'Ovary',\n",
       " 'Palate',\n",
       " 'Pancreas',\n",
       " 'Pancreas ',\n",
       " 'Pelvis, Prostate, Anus',\n",
       " 'Penis',\n",
       " 'Peripheral nerves and autonomic nervous system',\n",
       " 'Phantom',\n",
       " 'Prostate',\n",
       " 'Prostate gland',\n",
       " 'Rectosigmoid junction',\n",
       " 'Rectum',\n",
       " 'Renal pelvis',\n",
       " 'Retroperitoneum and peritoneum',\n",
       " 'Skin',\n",
       " 'Small intestine',\n",
       " 'Spinal cord, cranial nerves, and other parts of central nervous system',\n",
       " 'Stomach',\n",
       " 'Testicles',\n",
       " 'Testis',\n",
       " 'Thymus',\n",
       " 'Thyroid',\n",
       " 'Thyroid gland',\n",
       " 'Tonsil',\n",
       " 'Trachea',\n",
       " 'Unknown',\n",
       " 'Ureter',\n",
       " 'Uterus',\n",
       " 'Uterus, NOS',\n",
       " 'Vagina',\n",
       " 'Various',\n",
       " 'Various (11 locations)',\n",
       " 'Vulva']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b005036b",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e6b8dc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia sees that \"treatment_anatomic_site\" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the \"primary_diagnosis_site\" results. As she was initially looking for \"uterine\", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the \"ResearchSubject.primary_diagnosis_site\" for 'uter' as that should cover all variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31064125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix uteri', 'Corpus uteri', 'Uterus', 'Uterus, NOS']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"uter\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9311a49e",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Just to be sure, Julia also searches for any other instances of \"cervix\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2038a8cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix', 'Cervix uteri']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"cerv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c4de58",
   "metadata": {},
   "source": [
    "## Building a Query\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the \"treatment_anatomic_site\", only one term is of interest, so she uses the `=` operator to get only exact matches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "951fcc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = \"Cervix\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cb5f72",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "However, for \"primary_diagnosis_site\", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36cfd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Dsite = Q('ResearchSubject.primary_diagnosis_site = \"%uter%\" OR ResearchSubject.primary_diagnosis_site = \"%cerv%\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349af6f2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia adds her two queries together into one large one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9f5f9e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA = Tsite.OR(Dsite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f5cb55",
   "metadata": {},
   "source": [
    "## Looking at Summary Data\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "355b1706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3489 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">specimen_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">40793</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "specimen_count : \u001b[1;36m40793\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">treatment_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3049</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "treatment_count : \u001b[1;36m3049\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">diagnosis_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3685</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "diagnosis_count : \u001b[1;36m3685\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">researchsubject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "researchsubject_count : \u001b[1;36m4869\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">subject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3742</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "subject_count : \u001b[1;36m3742\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ce25fc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "55b0cdeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3633 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m4869\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">324646</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m324646\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a6254_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a6254_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a6254_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a6254_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_a6254_row0_col1\" class=\"data row0 col1\" >3591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a6254_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_a6254_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a6254_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_a6254_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_9e9a4_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_9e9a4_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_9e9a4_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row0_col0\" class=\"data row0 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_9e9a4_row0_col1\" class=\"data row0 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row1_col0\" class=\"data row1 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row1_col1\" class=\"data row1 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row2_col0\" class=\"data row2 col0\" >Adenomas and Adenocarcinomas</td>\n",
       "      <td id=\"T_9e9a4_row2_col1\" class=\"data row2 col1\" >1672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row3_col0\" class=\"data row3 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row3_col1\" class=\"data row3 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row4_col0\" class=\"data row4 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row4_col1\" class=\"data row4 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row5_col0\" class=\"data row5 col0\" >None</td>\n",
       "      <td id=\"T_9e9a4_row5_col1\" class=\"data row5 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row6_col0\" class=\"data row6 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_9e9a4_row6_col1\" class=\"data row6 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row7_col0\" class=\"data row7 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row7_col1\" class=\"data row7 col1\" >188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row8_col0\" class=\"data row8 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row8_col1\" class=\"data row8 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row9_col0\" class=\"data row9 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_9e9a4_row9_col1\" class=\"data row9 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row10_col0\" class=\"data row10 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_9e9a4_row10_col1\" class=\"data row10 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row11_col0\" class=\"data row11 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row11_col1\" class=\"data row11 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row12_col0\" class=\"data row12 col0\" >Not Reported</td>\n",
       "      <td id=\"T_9e9a4_row12_col1\" class=\"data row12 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row13_col0\" class=\"data row13 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_9e9a4_row13_col1\" class=\"data row13 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e9a4_row14_col0\" class=\"data row14 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_9e9a4_row14_col1\" class=\"data row14 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_1ac69_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_1ac69_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_1ac69_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_1ac69_row0_col0\" class=\"data row0 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_1ac69_row0_col1\" class=\"data row0 col1\" >780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_1ac69_row1_col0\" class=\"data row1 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_1ac69_row1_col1\" class=\"data row1 col1\" >2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_1ac69_row2_col0\" class=\"data row2 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_1ac69_row2_col1\" class=\"data row2 col1\" >915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_1ac69_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_1ac69_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_1ac69_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_1ac69_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a323e2",
   "metadata": {},
   "source": [
    "## Refining Queries\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0d526198",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3364 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3197</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m3197\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">298263</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m298263\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_592a5_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_592a5_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_592a5_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_592a5_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_592a5_row0_col1\" class=\"data row0 col1\" >1919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_592a5_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_592a5_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_592a5_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_592a5_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_e4abf_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_e4abf_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_e4abf_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row0_col0\" class=\"data row0 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row0_col1\" class=\"data row0 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row1_col0\" class=\"data row1 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row1_col1\" class=\"data row1 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row2_col0\" class=\"data row2 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row2_col1\" class=\"data row2 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row3_col0\" class=\"data row3 col0\" >Not Reported</td>\n",
       "      <td id=\"T_e4abf_row3_col1\" class=\"data row3 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row4_col0\" class=\"data row4 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_e4abf_row4_col1\" class=\"data row4 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row5_col0\" class=\"data row5 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_e4abf_row5_col1\" class=\"data row5 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row6_col0\" class=\"data row6 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row6_col1\" class=\"data row6 col1\" >188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row7_col0\" class=\"data row7 col0\" >None</td>\n",
       "      <td id=\"T_e4abf_row7_col1\" class=\"data row7 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row8_col0\" class=\"data row8 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_e4abf_row8_col1\" class=\"data row8 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row9_col0\" class=\"data row9 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_e4abf_row9_col1\" class=\"data row9 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row10_col0\" class=\"data row10 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row10_col1\" class=\"data row10 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row11_col0\" class=\"data row11 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_e4abf_row11_col1\" class=\"data row11 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row12_col0\" class=\"data row12 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_e4abf_row12_col1\" class=\"data row12 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_e4abf_row13_col0\" class=\"data row13 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_e4abf_row13_col1\" class=\"data row13 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_dbbbb_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_dbbbb_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_dbbbb_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_dbbbb_row0_col0\" class=\"data row0 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_dbbbb_row0_col1\" class=\"data row0 col1\" >962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dbbbb_row1_col0\" class=\"data row1 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_dbbbb_row1_col1\" class=\"data row1 col1\" >688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dbbbb_row2_col0\" class=\"data row2 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_dbbbb_row2_col1\" class=\"data row2 col1\" >373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dbbbb_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_dbbbb_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dbbbb_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_dbbbb_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Noadeno = Q('ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"')\n",
    "\n",
    "NoAdenoData = ALLDATA.AND(Noadeno)\n",
    "\n",
    "NoAdenoData.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a0191d",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work. Since she's mostly interested in looking at the kinds of data available from each endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d186b837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3486 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1cb371d5-20ca-41c...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0021035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26d54a63-24da-444f-bbb7-1f40d6335c78</td>\n",
       "      <td>[{'system': 'GDC', 'value': '26d54a63-24da-444...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0010283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3f24b940-868d-40b3-916f-268bd66b7a5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f24b940-868d-40b...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-000076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>474438a9-a134-43fd-8434-8a657b63b3db</td>\n",
       "      <td>[{'system': 'GDC', 'value': '474438a9-a134-43f...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0019659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-010402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>8bb89c0f-0f04-485e-8e08-801042dfcba9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8bb89c0f-0f04-485...</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>TCGA-EY-A3QX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>8bf18925-f53b-48b5-8245-e5b6c6a1cfd5</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8bf18925-f53b-48b...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>GENIE-UHN-993502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>9ff09e88-9897-4dcf-a99e-9f1bfbcecbf1</td>\n",
       "      <td>[{'system': 'GDC', 'value': '9ff09e88-9897-4dc...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-UHN-776060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>C3L-01604__cptac_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-01604__cptac_...</td>\n",
       "      <td>cptac_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>C3L-01604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>C3L-02353__cptac_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-02353__cptac_...</td>\n",
       "      <td>cptac_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>C3L-02353</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0   \n",
       "1   26d54a63-24da-444f-bbb7-1f40d6335c78   \n",
       "2   3f24b940-868d-40b3-916f-268bd66b7a5c   \n",
       "3   474438a9-a134-43fd-8434-8a657b63b3db   \n",
       "4   4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d   \n",
       "..                                   ...   \n",
       "95  8bb89c0f-0f04-485e-8e08-801042dfcba9   \n",
       "96  8bf18925-f53b-48b5-8245-e5b6c6a1cfd5   \n",
       "97  9ff09e88-9897-4dcf-a99e-9f1bfbcecbf1   \n",
       "98                 C3L-01604__cptac_ucec   \n",
       "99                 C3L-02353__cptac_ucec   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '1cb371d5-20ca-41c...   \n",
       "1   [{'system': 'GDC', 'value': '26d54a63-24da-444...   \n",
       "2   [{'system': 'GDC', 'value': '3f24b940-868d-40b...   \n",
       "3   [{'system': 'GDC', 'value': '474438a9-a134-43f...   \n",
       "4   [{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '8bb89c0f-0f04-485...   \n",
       "96  [{'system': 'GDC', 'value': '8bf18925-f53b-48b...   \n",
       "97  [{'system': 'GDC', 'value': '9ff09e88-9897-4dc...   \n",
       "98  [{'system': 'IDC', 'value': 'C3L-01604__cptac_...   \n",
       "99  [{'system': 'IDC', 'value': 'C3L-02353__cptac_...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                   GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "1                   GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "2                  GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "3                   GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "4                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "95                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   \n",
       "96                  GENIE-UHN                Squamous Cell Neoplasms   \n",
       "97                  GENIE-UHN              Epithelial Neoplasms, NOS   \n",
       "98                 cptac_ucec                                   None   \n",
       "99                 cptac_ucec                                   None   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0            Corpus uteri  GENIE-MSK-P-0021035  \n",
       "1            Corpus uteri  GENIE-MSK-P-0010283  \n",
       "2             Uterus, NOS    GENIE-DFCI-000076  \n",
       "3             Uterus, NOS  GENIE-MSK-P-0019659  \n",
       "4             Uterus, NOS    GENIE-DFCI-010402  \n",
       "..                    ...                  ...  \n",
       "95           Corpus uteri         TCGA-EY-A3QX  \n",
       "96           Cervix uteri     GENIE-UHN-993502  \n",
       "97           Corpus uteri     GENIE-UHN-776060  \n",
       "98                 Uterus            C3L-01604  \n",
       "99                 Uterus            C3L-02353  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086697b3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8d0f5e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3862 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD10521</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD10521'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD15235</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD15235'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD16470</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD16470'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3N-00860</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-00860'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3N-00872</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-00872'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GENIE-MSK-P-0000382</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00003...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23010.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GENIE-MSK-P-0001564</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00015...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-18993.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GENIE-MSK-P-0004976</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00049...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20454.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GENIE-MSK-P-0011101</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00111...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24837.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GENIE-MSK-P-0011145</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00111...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23741.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id                                         identifier  \\\n",
       "0               AD10521            [{'system': 'GDC', 'value': 'AD10521'}]   \n",
       "1               AD15235            [{'system': 'GDC', 'value': 'AD15235'}]   \n",
       "2               AD16470            [{'system': 'GDC', 'value': 'AD16470'}]   \n",
       "3             C3N-00860          [{'system': 'IDC', 'value': 'C3N-00860'}]   \n",
       "4             C3N-00872          [{'system': 'IDC', 'value': 'C3N-00872'}]   \n",
       "..                  ...                                                ...   \n",
       "95  GENIE-MSK-P-0000382  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00003...   \n",
       "96  GENIE-MSK-P-0001564  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00015...   \n",
       "97  GENIE-MSK-P-0004976  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00049...   \n",
       "98  GENIE-MSK-P-0011101  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00111...   \n",
       "99  GENIE-MSK-P-0011145  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00111...   \n",
       "\n",
       "         species     sex          race               ethnicity  days_to_birth  \\\n",
       "0   homo sapiens  female  not reported            not reported            NaN   \n",
       "1   homo sapiens  female  not reported            not reported            NaN   \n",
       "2   homo sapiens  female  not reported            not reported            NaN   \n",
       "3   homo sapiens    None          None                    None            NaN   \n",
       "4   homo sapiens    None          None                    None            NaN   \n",
       "..           ...     ...           ...                     ...            ...   \n",
       "95  homo sapiens  female         white  not hispanic or latino       -23010.0   \n",
       "96  homo sapiens  female         white  not hispanic or latino       -18993.0   \n",
       "97  homo sapiens  female       Unknown  not hispanic or latino       -20454.0   \n",
       "98  homo sapiens  female         white  not hispanic or latino       -24837.0   \n",
       "99  homo sapiens  female         white  not hispanic or latino       -23741.0   \n",
       "\n",
       "   subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                     [FM-AD]  Not Reported            NaN           None  \n",
       "1                     [FM-AD]  Not Reported            NaN           None  \n",
       "2                     [FM-AD]  Not Reported            NaN           None  \n",
       "3                [cptac_ucec]          None            NaN           None  \n",
       "4                [cptac_ucec]          None            NaN           None  \n",
       "..                        ...           ...            ...            ...  \n",
       "95                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "96                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "97                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "98                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "99                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.subject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec76132",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "04e04136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4030 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>031aea12-2e34-42d9-8aeb-2f279db4bc96</td>\n",
       "      <td>[{'system': 'GDC', 'value': '031aea12-2e34-42d...</td>\n",
       "      <td>7357a5af-22c2-47db-9e8f-cf9573253e58.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:031aea12-2e34-42d9-8aeb-2f279db4...</td>\n",
       "      <td>104470</td>\n",
       "      <td>29ea3df3c376bd588ad81b25da485f72</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>c9444e17-2283-4cc2-81be-a247434cc9f2</td>\n",
       "      <td>f1b9e9d1-e08f-47be-9584-f968ca0e507c</td>\n",
       "      <td>HTMCP-03-06-02242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>04780388-c545-4c11-a495-215e16a19420</td>\n",
       "      <td>[{'system': 'GDC', 'value': '04780388-c545-4c1...</td>\n",
       "      <td>a9873ccc-fb22-4ca1-a4d7-c4905a4ab339.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:04780388-c545-4c11-a495-215e16a1...</td>\n",
       "      <td>119886</td>\n",
       "      <td>9ae5a4d8e330882aae98e2f5b9e1aa43</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>de873ade-747a-4c4c-92a3-444d4b3f1447</td>\n",
       "      <td>f33eb114-1258-4fde-be0c-4a245f58e1ce</td>\n",
       "      <td>HTMCP-03-06-02128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0b5cd4dc-5081-466a-9bcc-bc717b3254af</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b5cd4dc-5081-466...</td>\n",
       "      <td>2db011f6-0005-4fab-ad90-a0e02afc1a0b.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:0b5cd4dc-5081-466a-9bcc-bc717b32...</td>\n",
       "      <td>67148</td>\n",
       "      <td>724e8649e0213a3ff98e4f47cffe324a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>a6dca3aa-fb41-49b4-a3c2-68c55c1a1e50</td>\n",
       "      <td>226914c5-e486-4041-9b0a-83ba8baae0e7</td>\n",
       "      <td>HTMCP-03-06-02003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0f6bff1c-a0e0-47f6-b4a9-6eff171b7e82</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0f6bff1c-a0e0-47f...</td>\n",
       "      <td>e05cb0b1-3d3b-47d0-b513-aeeba47b313d.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:0f6bff1c-a0e0-47f6-b4a9-6eff171b...</td>\n",
       "      <td>106577</td>\n",
       "      <td>d7208a402ce015d4c5f8cc7ee5aaf290</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>dc61fea6-cee3-48dc-9ff1-27a4e635d665</td>\n",
       "      <td>29754644-2d3d-4e04-830c-4b1c845df651</td>\n",
       "      <td>HTMCP-03-06-02180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>213b10f5-f321-4783-becb-d7732e356d47</td>\n",
       "      <td>[{'system': 'GDC', 'value': '213b10f5-f321-478...</td>\n",
       "      <td>ba783552-1267-45d8-8bb9-3c1ef687a6ef.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:213b10f5-f321-4783-becb-d7732e35...</td>\n",
       "      <td>60762</td>\n",
       "      <td>bfd55fb0a72927a01eda936547528967</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>75a6157f-7644-495b-b834-9e9c5adbcc89</td>\n",
       "      <td>74b5fdf0-acb1-446e-a604-df12010a7384</td>\n",
       "      <td>HTMCP-03-06-02097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>613e2cfc-83b8-40b1-b22c-41fa6ea92364</td>\n",
       "      <td>[{'system': 'GDC', 'value': '613e2cfc-83b8-40b...</td>\n",
       "      <td>74dcdb96-608d-4768-94e7-10d3203d0721.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:613e2cfc-83b8-40b1-b22c-41fa6ea9...</td>\n",
       "      <td>186690</td>\n",
       "      <td>a07299bd5a8e61454058eae5613c521b</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>2a0ae2ba-f41b-4365-a56e-5f9ca43ab956</td>\n",
       "      <td>29d2a6dd-fb70-46ca-bfa3-0a00025970b7</td>\n",
       "      <td>HTMCP-03-06-02428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>6c97257c-13cf-48b8-a2a1-01aceab7bb5b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6c97257c-13cf-48b...</td>\n",
       "      <td>dc304249-7be1-4035-800a-ade9b2b5fc67.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:6c97257c-13cf-48b8-a2a1-01aceab7...</td>\n",
       "      <td>102572</td>\n",
       "      <td>dbceb045a7533e9ed0d0338c3efbd103</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>216980f2-62d2-464e-9fec-7dc1986debec</td>\n",
       "      <td>226914c5-e486-4041-9b0a-83ba8baae0e7</td>\n",
       "      <td>HTMCP-03-06-02003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>7aaae0a8-ee16-4817-8e02-40dc032f20db</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7aaae0a8-ee16-481...</td>\n",
       "      <td>852776ab-970b-4f18-8f68-4efac4044955.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:7aaae0a8-ee16-4817-8e02-40dc032f...</td>\n",
       "      <td>71685</td>\n",
       "      <td>8a7c00e35763dd3b8335a39a64951316</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>631f7f49-3aed-45cc-82d0-75b9dfb015ef</td>\n",
       "      <td>59673b19-2e14-441f-9624-626ed0a3530b</td>\n",
       "      <td>HTMCP-03-06-02125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>9d3ec32e-f978-46b6-86a7-0f0f222e7b66</td>\n",
       "      <td>[{'system': 'GDC', 'value': '9d3ec32e-f978-46b...</td>\n",
       "      <td>786bfbd3-b161-402e-bfaa-af3646d7cf8e.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:9d3ec32e-f978-46b6-86a7-0f0f222e...</td>\n",
       "      <td>600398</td>\n",
       "      <td>7f430f95a5b784e75086768a14e69779</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>c7b71dc2-019f-4739-b03d-9cb2a4c984cc</td>\n",
       "      <td>ec6c18ae-1e62-4c7b-84b9-8b795b9da775</td>\n",
       "      <td>HTMCP-03-06-02109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>a634ee3d-2b72-4169-9af8-0111220abb53</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a634ee3d-2b72-416...</td>\n",
       "      <td>d1ae08d8-1f85-4202-adaf-a968015eeefe.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:a634ee3d-2b72-4169-9af8-0111220a...</td>\n",
       "      <td>56356</td>\n",
       "      <td>b7fb62f07f388b5fd80903444ab5a6f7</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>c1d4b32e-5e07-48f6-81dc-df28944db94c</td>\n",
       "      <td>458cbd14-a7ce-4d65-9caf-680c4c66e024</td>\n",
       "      <td>HTMCP-03-06-02214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   031aea12-2e34-42d9-8aeb-2f279db4bc96   \n",
       "1   04780388-c545-4c11-a495-215e16a19420   \n",
       "2   0b5cd4dc-5081-466a-9bcc-bc717b3254af   \n",
       "3   0f6bff1c-a0e0-47f6-b4a9-6eff171b7e82   \n",
       "4   213b10f5-f321-4783-becb-d7732e356d47   \n",
       "..                                   ...   \n",
       "95  613e2cfc-83b8-40b1-b22c-41fa6ea92364   \n",
       "96  6c97257c-13cf-48b8-a2a1-01aceab7bb5b   \n",
       "97  7aaae0a8-ee16-4817-8e02-40dc032f20db   \n",
       "98  9d3ec32e-f978-46b6-86a7-0f0f222e7b66   \n",
       "99  a634ee3d-2b72-4169-9af8-0111220abb53   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '031aea12-2e34-42d...   \n",
       "1   [{'system': 'GDC', 'value': '04780388-c545-4c1...   \n",
       "2   [{'system': 'GDC', 'value': '0b5cd4dc-5081-466...   \n",
       "3   [{'system': 'GDC', 'value': '0f6bff1c-a0e0-47f...   \n",
       "4   [{'system': 'GDC', 'value': '213b10f5-f321-478...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '613e2cfc-83b8-40b...   \n",
       "96  [{'system': 'GDC', 'value': '6c97257c-13cf-48b...   \n",
       "97  [{'system': 'GDC', 'value': '7aaae0a8-ee16-481...   \n",
       "98  [{'system': 'GDC', 'value': '9d3ec32e-f978-46b...   \n",
       "99  [{'system': 'GDC', 'value': 'a634ee3d-2b72-416...   \n",
       "\n",
       "                                                label  \\\n",
       "0   7357a5af-22c2-47db-9e8f-cf9573253e58.wgs.BRASS...   \n",
       "1   a9873ccc-fb22-4ca1-a4d7-c4905a4ab339.wgs.BRASS...   \n",
       "2   2db011f6-0005-4fab-ad90-a0e02afc1a0b.wgs.BRASS...   \n",
       "3   e05cb0b1-3d3b-47d0-b513-aeeba47b313d.wgs.BRASS...   \n",
       "4   ba783552-1267-45d8-8bb9-3c1ef687a6ef.wgs.BRASS...   \n",
       "..                                                ...   \n",
       "95  74dcdb96-608d-4768-94e7-10d3203d0721.wgs.BRASS...   \n",
       "96  dc304249-7be1-4035-800a-ade9b2b5fc67.wgs.BRASS...   \n",
       "97  852776ab-970b-4f18-8f68-4efac4044955.wgs.BRASS...   \n",
       "98  786bfbd3-b161-402e-bfaa-af3646d7cf8e.wgs.BRASS...   \n",
       "99  d1ae08d8-1f85-4202-adaf-a968015eeefe.wgs.BRASS...   \n",
       "\n",
       "                   data_category                 data_type file_format  \\\n",
       "0   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "1   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "2   Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "3   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "4   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "..                           ...                       ...         ...   \n",
       "95  Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "96  Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "97  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "98  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "99  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "\n",
       "   associated_project                                            drs_uri  \\\n",
       "0       CGCI-HTMCP-CC  drs://dg.4DFC:031aea12-2e34-42d9-8aeb-2f279db4...   \n",
       "1       CGCI-HTMCP-CC  drs://dg.4DFC:04780388-c545-4c11-a495-215e16a1...   \n",
       "2       CGCI-HTMCP-CC  drs://dg.4DFC:0b5cd4dc-5081-466a-9bcc-bc717b32...   \n",
       "3       CGCI-HTMCP-CC  drs://dg.4DFC:0f6bff1c-a0e0-47f6-b4a9-6eff171b...   \n",
       "4       CGCI-HTMCP-CC  drs://dg.4DFC:213b10f5-f321-4783-becb-d7732e35...   \n",
       "..                ...                                                ...   \n",
       "95      CGCI-HTMCP-CC  drs://dg.4DFC:613e2cfc-83b8-40b1-b22c-41fa6ea9...   \n",
       "96      CGCI-HTMCP-CC  drs://dg.4DFC:6c97257c-13cf-48b8-a2a1-01aceab7...   \n",
       "97      CGCI-HTMCP-CC  drs://dg.4DFC:7aaae0a8-ee16-4817-8e02-40dc032f...   \n",
       "98      CGCI-HTMCP-CC  drs://dg.4DFC:9d3ec32e-f978-46b6-86a7-0f0f222e...   \n",
       "99      CGCI-HTMCP-CC  drs://dg.4DFC:a634ee3d-2b72-4169-9af8-0111220a...   \n",
       "\n",
       "    byte_size                          checksum data_modality  \\\n",
       "0      104470  29ea3df3c376bd588ad81b25da485f72       Genomic   \n",
       "1      119886  9ae5a4d8e330882aae98e2f5b9e1aa43       Genomic   \n",
       "2       67148  724e8649e0213a3ff98e4f47cffe324a       Genomic   \n",
       "3      106577  d7208a402ce015d4c5f8cc7ee5aaf290       Genomic   \n",
       "4       60762  bfd55fb0a72927a01eda936547528967       Genomic   \n",
       "..        ...                               ...           ...   \n",
       "95     186690  a07299bd5a8e61454058eae5613c521b       Genomic   \n",
       "96     102572  dbceb045a7533e9ed0d0338c3efbd103       Genomic   \n",
       "97      71685  8a7c00e35763dd3b8335a39a64951316       Genomic   \n",
       "98     600398  7f430f95a5b784e75086768a14e69779       Genomic   \n",
       "99      56356  b7fb62f07f388b5fd80903444ab5a6f7       Genomic   \n",
       "\n",
       "   imaging_modality dbgap_accession_number imaging_series  \\\n",
       "0              None              phs000528           None   \n",
       "1              None              phs000528           None   \n",
       "2              None              phs000528           None   \n",
       "3              None              phs000528           None   \n",
       "4              None              phs000528           None   \n",
       "..              ...                    ...            ...   \n",
       "95             None              phs000528           None   \n",
       "96             None              phs000528           None   \n",
       "97             None              phs000528           None   \n",
       "98             None              phs000528           None   \n",
       "99             None              phs000528           None   \n",
       "\n",
       "             researchsubject_specimen_id  \\\n",
       "0   c9444e17-2283-4cc2-81be-a247434cc9f2   \n",
       "1   de873ade-747a-4c4c-92a3-444d4b3f1447   \n",
       "2   a6dca3aa-fb41-49b4-a3c2-68c55c1a1e50   \n",
       "3   dc61fea6-cee3-48dc-9ff1-27a4e635d665   \n",
       "4   75a6157f-7644-495b-b834-9e9c5adbcc89   \n",
       "..                                   ...   \n",
       "95  2a0ae2ba-f41b-4365-a56e-5f9ca43ab956   \n",
       "96  216980f2-62d2-464e-9fec-7dc1986debec   \n",
       "97  631f7f49-3aed-45cc-82d0-75b9dfb015ef   \n",
       "98  c7b71dc2-019f-4739-b03d-9cb2a4c984cc   \n",
       "99  c1d4b32e-5e07-48f6-81dc-df28944db94c   \n",
       "\n",
       "                      researchsubject_id         subject_id  \n",
       "0   f1b9e9d1-e08f-47be-9584-f968ca0e507c  HTMCP-03-06-02242  \n",
       "1   f33eb114-1258-4fde-be0c-4a245f58e1ce  HTMCP-03-06-02128  \n",
       "2   226914c5-e486-4041-9b0a-83ba8baae0e7  HTMCP-03-06-02003  \n",
       "3   29754644-2d3d-4e04-830c-4b1c845df651  HTMCP-03-06-02180  \n",
       "4   74b5fdf0-acb1-446e-a604-df12010a7384  HTMCP-03-06-02097  \n",
       "..                                   ...                ...  \n",
       "95  29d2a6dd-fb70-46ca-bfa3-0a00025970b7  HTMCP-03-06-02428  \n",
       "96  226914c5-e486-4041-9b0a-83ba8baae0e7  HTMCP-03-06-02003  \n",
       "97  59673b19-2e14-441f-9624-626ed0a3530b  HTMCP-03-06-02125  \n",
       "98  ec6c18ae-1e62-4c7b-84b9-8b795b9da775  HTMCP-03-06-02109  \n",
       "99  458cbd14-a7ce-4d65-9caf-680c4c66e024  HTMCP-03-06-02214  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.file.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf9f2d3",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retrieve this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6aadbe",
   "metadata": {},
   "source": [
    "## Working with Results (pagination)\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c2cec2bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3579 ms\n"
     ]
    }
   ],
   "source": [
    "researchsubs = NoAdenoData.researchsubject.run()\n",
    "rsdf = pd.DataFrame()\n",
    "for i in researchsubs.paginator(to_df=True):\n",
    "    rsdf = pd.concat([rsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a1258057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3520 ms\n"
     ]
    }
   ],
   "source": [
    "subs = NoAdenoData.subject.run()\n",
    "subsdf = pd.DataFrame()\n",
    "for i in subs.paginator(to_df=True):\n",
    "    subsdf = pd.concat([subsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "04cd73df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1cb371d5-20ca-41c...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0021035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26d54a63-24da-444f-bbb7-1f40d6335c78</td>\n",
       "      <td>[{'system': 'GDC', 'value': '26d54a63-24da-444...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0010283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3f24b940-868d-40b3-916f-268bd66b7a5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f24b940-868d-40b...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-000076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>474438a9-a134-43fd-8434-8a657b63b3db</td>\n",
       "      <td>[{'system': 'GDC', 'value': '474438a9-a134-43f...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0019659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-010402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>bf1f593d-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'bf1f593d-118a-11e...</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>C3N-01217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>de7dabc7-ba48-4ac9-828c-02ec47351d6a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'de7dabc7-ba48-4ac...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-004071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>e1cb5a54-594b-4b2a-a9da-e206e2103c5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e1cb5a54-594b-4b2...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD14736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>e40bbfbd-b273-4b17-ac5b-d57125bc9ba8</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e40bbfbd-b273-4b1...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-006773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>e9230570-37a0-4bb6-bf21-170e830fe9dc</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e9230570-37a0-4bb...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02447</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0   \n",
       "1   26d54a63-24da-444f-bbb7-1f40d6335c78   \n",
       "2   3f24b940-868d-40b3-916f-268bd66b7a5c   \n",
       "3   474438a9-a134-43fd-8434-8a657b63b3db   \n",
       "4   4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d   \n",
       "..                                   ...   \n",
       "92  bf1f593d-118a-11e9-afb9-0a9c39d33490   \n",
       "93  de7dabc7-ba48-4ac9-828c-02ec47351d6a   \n",
       "94  e1cb5a54-594b-4b2a-a9da-e206e2103c5c   \n",
       "95  e40bbfbd-b273-4b17-ac5b-d57125bc9ba8   \n",
       "96  e9230570-37a0-4bb6-bf21-170e830fe9dc   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '1cb371d5-20ca-41c...   \n",
       "1   [{'system': 'GDC', 'value': '26d54a63-24da-444...   \n",
       "2   [{'system': 'GDC', 'value': '3f24b940-868d-40b...   \n",
       "3   [{'system': 'GDC', 'value': '474438a9-a134-43f...   \n",
       "4   [{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...   \n",
       "..                                                ...   \n",
       "92  [{'system': 'PDC', 'value': 'bf1f593d-118a-11e...   \n",
       "93  [{'system': 'GDC', 'value': 'de7dabc7-ba48-4ac...   \n",
       "94  [{'system': 'GDC', 'value': 'e1cb5a54-594b-4b2...   \n",
       "95  [{'system': 'GDC', 'value': 'e40bbfbd-b273-4b1...   \n",
       "96  [{'system': 'GDC', 'value': 'e9230570-37a0-4bb...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                   GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "1                   GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "2                  GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "3                   GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "4                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "92           CPTAC3-Discovery   Uterine Corpus Endometrial Carcinoma   \n",
       "93                 GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "94                      FM-AD  Cystic, Mucinous and Serous Neoplasms   \n",
       "95                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "96              CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0            Corpus uteri  GENIE-MSK-P-0021035  \n",
       "1            Corpus uteri  GENIE-MSK-P-0010283  \n",
       "2             Uterus, NOS    GENIE-DFCI-000076  \n",
       "3             Uterus, NOS  GENIE-MSK-P-0019659  \n",
       "4             Uterus, NOS    GENIE-DFCI-010402  \n",
       "..                    ...                  ...  \n",
       "92            Uterus, NOS            C3N-01217  \n",
       "93            Uterus, NOS    GENIE-DFCI-004071  \n",
       "94            Uterus, NOS              AD14736  \n",
       "95            Uterus, NOS    GENIE-DFCI-006773  \n",
       "96           Cervix uteri    HTMCP-03-06-02447  \n",
       "\n",
       "[3197 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rsdf # view the researchsubject dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "92a6f811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD10521</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD10521'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD15235</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD15235'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD16470</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD16470'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3N-00860</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-00860'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3N-00872</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-00872'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-EK-A3GM</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A3GM'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>-23879.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-EY-A1GH</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EY-A1GH'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-25802.0</td>\n",
       "      <td>[tcga_ucec, TCGA-UCEC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-HB-A43Z</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-HB-A43Z'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21233.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-KJ-A3U4</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-KJ-A3U4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>-20392.0</td>\n",
       "      <td>[tcga_ucec, TCGA-UCEC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-N1-A6IA</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N1-A6IA'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-15112.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>224.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2609 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                         identifier  \\\n",
       "0        AD10521            [{'system': 'GDC', 'value': 'AD10521'}]   \n",
       "1        AD15235            [{'system': 'GDC', 'value': 'AD15235'}]   \n",
       "2        AD16470            [{'system': 'GDC', 'value': 'AD16470'}]   \n",
       "3      C3N-00860          [{'system': 'IDC', 'value': 'C3N-00860'}]   \n",
       "4      C3N-00872          [{'system': 'IDC', 'value': 'C3N-00872'}]   \n",
       "..           ...                                                ...   \n",
       "4   TCGA-EK-A3GM  [{'system': 'GDC', 'value': 'TCGA-EK-A3GM'}, {...   \n",
       "5   TCGA-EY-A1GH  [{'system': 'GDC', 'value': 'TCGA-EY-A1GH'}, {...   \n",
       "6   TCGA-HB-A43Z  [{'system': 'GDC', 'value': 'TCGA-HB-A43Z'}, {...   \n",
       "7   TCGA-KJ-A3U4  [{'system': 'GDC', 'value': 'TCGA-KJ-A3U4'}, {...   \n",
       "8   TCGA-N1-A6IA  [{'system': 'GDC', 'value': 'TCGA-N1-A6IA'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "0   homo sapiens  female               not reported            not reported   \n",
       "1   homo sapiens  female               not reported            not reported   \n",
       "2   homo sapiens  female               not reported            not reported   \n",
       "3   homo sapiens    None                       None                    None   \n",
       "4   homo sapiens    None                       None                    None   \n",
       "..           ...     ...                        ...                     ...   \n",
       "4   homo sapiens  female                      white      hispanic or latino   \n",
       "5   homo sapiens  female                      white  not hispanic or latino   \n",
       "6   homo sapiens  female                      white  not hispanic or latino   \n",
       "7   homo sapiens  female                      white      hispanic or latino   \n",
       "8   homo sapiens  female  black or african american  not hispanic or latino   \n",
       "\n",
       "    days_to_birth subject_associated_project  vital_status  days_to_death  \\\n",
       "0             NaN                    [FM-AD]  Not Reported            NaN   \n",
       "1             NaN                    [FM-AD]  Not Reported            NaN   \n",
       "2             NaN                    [FM-AD]  Not Reported            NaN   \n",
       "3             NaN               [cptac_ucec]          None            NaN   \n",
       "4             NaN               [cptac_ucec]          None            NaN   \n",
       "..            ...                        ...           ...            ...   \n",
       "4        -23879.0     [tcga_cesc, TCGA-CESC]         Alive            NaN   \n",
       "5        -25802.0     [tcga_ucec, TCGA-UCEC]         Alive            NaN   \n",
       "6        -21233.0     [tcga_sarc, TCGA-SARC]         Alive            NaN   \n",
       "7        -20392.0     [tcga_ucec, TCGA-UCEC]         Alive            NaN   \n",
       "8        -15112.0     [tcga_sarc, TCGA-SARC]          Dead          224.0   \n",
       "\n",
       "   cause_of_death  \n",
       "0            None  \n",
       "1            None  \n",
       "2            None  \n",
       "3            None  \n",
       "4            None  \n",
       "..            ...  \n",
       "4            None  \n",
       "5            None  \n",
       "6            None  \n",
       "7            None  \n",
       "8            None  \n",
       "\n",
       "[2609 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsdf # view the subject dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bcbe86",
   "metadata": {},
   "source": [
    "## Merging Results across Endpoints\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Then Julia uses the `subject_id` and `id` fields in each result to merge them together into one big dataset. She also specifies that any columns that are in both tables should be kept and have a suffix added to their name. This will help her to check that her merge worked correctly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9b7a3383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_rs</th>\n",
       "      <th>identifier_rs</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>id_sub</th>\n",
       "      <th>identifier_sub</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1cb371d5-20ca-41c...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0021035</td>\n",
       "      <td>GENIE-MSK-P-0021035</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00210...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21915.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26d54a63-24da-444f-bbb7-1f40d6335c78</td>\n",
       "      <td>[{'system': 'GDC', 'value': '26d54a63-24da-444...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0010283</td>\n",
       "      <td>GENIE-MSK-P-0010283</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00102...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-28854.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3f24b940-868d-40b3-916f-268bd66b7a5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f24b940-868d-40b...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-000076</td>\n",
       "      <td>GENIE-DFCI-000076</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-000076'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17897.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>474438a9-a134-43fd-8434-8a657b63b3db</td>\n",
       "      <td>[{'system': 'GDC', 'value': '474438a9-a134-43f...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0019659</td>\n",
       "      <td>GENIE-MSK-P-0019659</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00196...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21184.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-010402</td>\n",
       "      <td>GENIE-DFCI-010402</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-010402'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23010.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>a79ece94-3030-4e33-a9aa-623f430cd379</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a79ece94-3030-4e3...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02103</td>\n",
       "      <td>HTMCP-03-06-02103</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'HTMCP-03-06-02103'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CGCI-HTMCP-CC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>415.0</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193</th>\n",
       "      <td>de7dabc7-ba48-4ac9-828c-02ec47351d6a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'de7dabc7-ba48-4ac...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-004071</td>\n",
       "      <td>GENIE-DFCI-004071</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-004071'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24471.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3194</th>\n",
       "      <td>e1cb5a54-594b-4b2a-a9da-e206e2103c5c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e1cb5a54-594b-4b2...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD14736</td>\n",
       "      <td>AD14736</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD14736'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3195</th>\n",
       "      <td>e40bbfbd-b273-4b17-ac5b-d57125bc9ba8</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e40bbfbd-b273-4b1...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-006773</td>\n",
       "      <td>GENIE-DFCI-006773</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-006773'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>hispanic or latino</td>\n",
       "      <td>-10227.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3196</th>\n",
       "      <td>e9230570-37a0-4bb6-bf21-170e830fe9dc</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e9230570-37a0-4bb...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02447</td>\n",
       "      <td>HTMCP-03-06-02447</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'HTMCP-03-06-02447'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21829.0</td>\n",
       "      <td>[CGCI-HTMCP-CC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     id_rs  \\\n",
       "0     1cb371d5-20ca-41c3-8cf8-0979fa7bd9e0   \n",
       "1     26d54a63-24da-444f-bbb7-1f40d6335c78   \n",
       "2     3f24b940-868d-40b3-916f-268bd66b7a5c   \n",
       "3     474438a9-a134-43fd-8434-8a657b63b3db   \n",
       "4     4a02ccbd-ed6b-497c-9e36-b7575ee2ee5d   \n",
       "...                                    ...   \n",
       "3192  a79ece94-3030-4e33-a9aa-623f430cd379   \n",
       "3193  de7dabc7-ba48-4ac9-828c-02ec47351d6a   \n",
       "3194  e1cb5a54-594b-4b2a-a9da-e206e2103c5c   \n",
       "3195  e40bbfbd-b273-4b17-ac5b-d57125bc9ba8   \n",
       "3196  e9230570-37a0-4bb6-bf21-170e830fe9dc   \n",
       "\n",
       "                                          identifier_rs  \\\n",
       "0     [{'system': 'GDC', 'value': '1cb371d5-20ca-41c...   \n",
       "1     [{'system': 'GDC', 'value': '26d54a63-24da-444...   \n",
       "2     [{'system': 'GDC', 'value': '3f24b940-868d-40b...   \n",
       "3     [{'system': 'GDC', 'value': '474438a9-a134-43f...   \n",
       "4     [{'system': 'GDC', 'value': '4a02ccbd-ed6b-497...   \n",
       "...                                                 ...   \n",
       "3192  [{'system': 'GDC', 'value': 'a79ece94-3030-4e3...   \n",
       "3193  [{'system': 'GDC', 'value': 'de7dabc7-ba48-4ac...   \n",
       "3194  [{'system': 'GDC', 'value': 'e1cb5a54-594b-4b2...   \n",
       "3195  [{'system': 'GDC', 'value': 'e40bbfbd-b273-4b1...   \n",
       "3196  [{'system': 'GDC', 'value': 'e9230570-37a0-4bb...   \n",
       "\n",
       "     member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                     GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "1                     GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "2                    GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "3                     GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "4                    GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "...                         ...                                    ...   \n",
       "3192              CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "3193                 GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "3194                      FM-AD  Cystic, Mucinous and Serous Neoplasms   \n",
       "3195                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "3196              CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "\n",
       "     primary_diagnosis_site           subject_id               id_sub  \\\n",
       "0              Corpus uteri  GENIE-MSK-P-0021035  GENIE-MSK-P-0021035   \n",
       "1              Corpus uteri  GENIE-MSK-P-0010283  GENIE-MSK-P-0010283   \n",
       "2               Uterus, NOS    GENIE-DFCI-000076    GENIE-DFCI-000076   \n",
       "3               Uterus, NOS  GENIE-MSK-P-0019659  GENIE-MSK-P-0019659   \n",
       "4               Uterus, NOS    GENIE-DFCI-010402    GENIE-DFCI-010402   \n",
       "...                     ...                  ...                  ...   \n",
       "3192           Cervix uteri    HTMCP-03-06-02103    HTMCP-03-06-02103   \n",
       "3193            Uterus, NOS    GENIE-DFCI-004071    GENIE-DFCI-004071   \n",
       "3194            Uterus, NOS              AD14736              AD14736   \n",
       "3195            Uterus, NOS    GENIE-DFCI-006773    GENIE-DFCI-006773   \n",
       "3196           Cervix uteri    HTMCP-03-06-02447    HTMCP-03-06-02447   \n",
       "\n",
       "                                         identifier_sub       species     sex  \\\n",
       "0     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00210...  homo sapiens  female   \n",
       "1     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00102...  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'GENIE-DFCI-000076'}]  homo sapiens  female   \n",
       "3     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00196...  homo sapiens  female   \n",
       "4     [{'system': 'GDC', 'value': 'GENIE-DFCI-010402'}]  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "3192  [{'system': 'GDC', 'value': 'HTMCP-03-06-02103'}]  homo sapiens  female   \n",
       "3193  [{'system': 'GDC', 'value': 'GENIE-DFCI-004071'}]  homo sapiens  female   \n",
       "3194            [{'system': 'GDC', 'value': 'AD14736'}]  homo sapiens  female   \n",
       "3195  [{'system': 'GDC', 'value': 'GENIE-DFCI-006773'}]  homo sapiens  female   \n",
       "3196  [{'system': 'GDC', 'value': 'HTMCP-03-06-02447'}]  homo sapiens  female   \n",
       "\n",
       "                           race               ethnicity  days_to_birth  \\\n",
       "0                         white  not hispanic or latino       -21915.0   \n",
       "1                         white  not hispanic or latino       -28854.0   \n",
       "2                         white  not hispanic or latino       -17897.0   \n",
       "3                         white  not hispanic or latino       -21184.0   \n",
       "4                         white  not hispanic or latino       -23010.0   \n",
       "...                         ...                     ...            ...   \n",
       "3192                    Unknown                 Unknown            NaN   \n",
       "3193                      white  not hispanic or latino       -24471.0   \n",
       "3194               not reported            not reported            NaN   \n",
       "3195                      white      hispanic or latino       -10227.0   \n",
       "3196  black or african american  not hispanic or latino       -21829.0   \n",
       "\n",
       "     subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                   [GENIE-MSK]  Not Reported            NaN           None  \n",
       "1                   [GENIE-MSK]  Not Reported            NaN           None  \n",
       "2                  [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3                   [GENIE-MSK]  Not Reported            NaN           None  \n",
       "4                  [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "...                         ...           ...            ...            ...  \n",
       "3192            [CGCI-HTMCP-CC]          Dead          415.0        Unknown  \n",
       "3193               [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3194                    [FM-AD]  Not Reported            NaN           None  \n",
       "3195               [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3196            [CGCI-HTMCP-CC]         Alive            NaN           None  \n",
       "\n",
       "[3197 rows x 17 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmetadata = pd.merge(rsdf,\n",
    "                subsdf,\n",
    "                left_on=\"subject_id\",\n",
    "                right_on='id',\n",
    "                suffixes=(\"_rs\", \"_sub\"))\n",
    "\n",
    "allmetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2e053e",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "`subject_id` from the research subject results seems to perfectly match the id data from the subject table, `id_sub`. Julia then checks to see that her dataframe is the right size. She had 3197 researchsubject rows, so she expects 3197 rows here as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e5874a36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id_rs                          3197\n",
       "identifier_rs                  3197\n",
       "member_of_research_project     3197\n",
       "primary_diagnosis_condition    2022\n",
       "primary_diagnosis_site         3197\n",
       "subject_id                     3197\n",
       "id_sub                         3197\n",
       "identifier_sub                 3197\n",
       "species                        3197\n",
       "sex                            3025\n",
       "race                           3025\n",
       "ethnicity                      3025\n",
       "days_to_birth                  2740\n",
       "subject_associated_project     3197\n",
       "vital_status                   3025\n",
       "days_to_death                   482\n",
       "cause_of_death                  309\n",
       "dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmetadata.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024da831",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Satisfied with her results, Julia saves the data out to a csv so she can browse it with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b6628de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata.to_csv(\"allmetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246644d3",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia knows from her subject count summary that there are more than 200,000 files associated with her subjects, which is likely far more than she needs. To help her decide what files she wants, Julia uses endpoint chaining to get summary information about the files that are assigned to researchsubjects for her search criteria:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ae1ae079",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3409 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">298263</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   total : \u001b[1;36m298263\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_97401_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_97401_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_97401_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_97401_row0_col0\" class=\"data row0 col0\" >IDC</td>\n",
       "      <td id=\"T_97401_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_97401_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_97401_row1_col1\" class=\"data row1 col1\" >2560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_97401_row2_col0\" class=\"data row2 col0\" >GDC</td>\n",
       "      <td id=\"T_97401_row2_col1\" class=\"data row2 col1\" >31274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_98838_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_98838_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_98838_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row0_col0\" class=\"data row0 col0\" >Imaging</td>\n",
       "      <td id=\"T_98838_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row1_col0\" class=\"data row1 col0\" >Simple Nucleotide Variation</td>\n",
       "      <td id=\"T_98838_row1_col1\" class=\"data row1 col1\" >11745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row2_col0\" class=\"data row2 col0\" >Structural Variation</td>\n",
       "      <td id=\"T_98838_row2_col1\" class=\"data row2 col1\" >2192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row3_col0\" class=\"data row3 col0\" >Sequencing Reads</td>\n",
       "      <td id=\"T_98838_row3_col1\" class=\"data row3 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row4_col0\" class=\"data row4 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_98838_row4_col1\" class=\"data row4 col1\" >2866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row5_col0\" class=\"data row5 col0\" >Transcriptome Profiling</td>\n",
       "      <td id=\"T_98838_row5_col1\" class=\"data row5 col1\" >2820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row6_col0\" class=\"data row6 col0\" >DNA Methylation</td>\n",
       "      <td id=\"T_98838_row6_col1\" class=\"data row6 col1\" >1947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row7_col0\" class=\"data row7 col0\" >Peptide Spectral Matches</td>\n",
       "      <td id=\"T_98838_row7_col1\" class=\"data row7 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row8_col0\" class=\"data row8 col0\" >Copy Number Variation</td>\n",
       "      <td id=\"T_98838_row8_col1\" class=\"data row8 col1\" >4079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row9_col0\" class=\"data row9 col0\" >Processed Mass Spectra</td>\n",
       "      <td id=\"T_98838_row9_col1\" class=\"data row9 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row10_col0\" class=\"data row10 col0\" >Clinical</td>\n",
       "      <td id=\"T_98838_row10_col1\" class=\"data row10 col1\" >774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row11_col0\" class=\"data row11 col0\" >Raw Mass Spectra</td>\n",
       "      <td id=\"T_98838_row11_col1\" class=\"data row11 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row12_col0\" class=\"data row12 col0\" >Proteome Profiling</td>\n",
       "      <td id=\"T_98838_row12_col1\" class=\"data row12 col1\" >339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_98838_row13_col0\" class=\"data row13 col0\" >Somatic Structural Variation</td>\n",
       "      <td id=\"T_98838_row13_col1\" class=\"data row13 col1\" >370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_ff236_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_ff236_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_ff236_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row0_col0\" class=\"data row0 col0\" >DICOM</td>\n",
       "      <td id=\"T_ff236_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row1_col0\" class=\"data row1 col0\" >MAF</td>\n",
       "      <td id=\"T_ff236_row1_col1\" class=\"data row1 col1\" >5235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row2_col0\" class=\"data row2 col0\" >TSV</td>\n",
       "      <td id=\"T_ff236_row2_col1\" class=\"data row2 col1\" >3980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row3_col0\" class=\"data row3 col0\" >SVS</td>\n",
       "      <td id=\"T_ff236_row3_col1\" class=\"data row3 col1\" >1111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row4_col0\" class=\"data row4 col0\" >TXT</td>\n",
       "      <td id=\"T_ff236_row4_col1\" class=\"data row4 col1\" >4827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row5_col0\" class=\"data row5 col0\" >IDAT</td>\n",
       "      <td id=\"T_ff236_row5_col1\" class=\"data row5 col1\" >1298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row6_col0\" class=\"data row6 col0\" >BAM</td>\n",
       "      <td id=\"T_ff236_row6_col1\" class=\"data row6 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row7_col0\" class=\"data row7 col0\" >BCR SSF XML</td>\n",
       "      <td id=\"T_ff236_row7_col1\" class=\"data row7 col1\" >517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row8_col0\" class=\"data row8 col0\" >BEDPE</td>\n",
       "      <td id=\"T_ff236_row8_col1\" class=\"data row8 col1\" >1504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row9_col0\" class=\"data row9 col0\" >BCR XML</td>\n",
       "      <td id=\"T_ff236_row9_col1\" class=\"data row9 col1\" >1217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row10_col0\" class=\"data row10 col0\" >mzML</td>\n",
       "      <td id=\"T_ff236_row10_col1\" class=\"data row10 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row11_col0\" class=\"data row11 col0\" >VCF</td>\n",
       "      <td id=\"T_ff236_row11_col1\" class=\"data row11 col1\" >6652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row12_col0\" class=\"data row12 col0\" >BCR Auxiliary XML</td>\n",
       "      <td id=\"T_ff236_row12_col1\" class=\"data row12 col1\" >474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row13_col0\" class=\"data row13 col0\" >vendor-specific</td>\n",
       "      <td id=\"T_ff236_row13_col1\" class=\"data row13 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row14_col0\" class=\"data row14 col0\" >tsv</td>\n",
       "      <td id=\"T_ff236_row14_col1\" class=\"data row14 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row15_col0\" class=\"data row15 col0\" >BCR PPS XML</td>\n",
       "      <td id=\"T_ff236_row15_col1\" class=\"data row15 col1\" >193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row16_col0\" class=\"data row16 col0\" >mzIdentML</td>\n",
       "      <td id=\"T_ff236_row16_col1\" class=\"data row16 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row17_col0\" class=\"data row17 col0\" >BCR OMF XML</td>\n",
       "      <td id=\"T_ff236_row17_col1\" class=\"data row17 col1\" >39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row18_col0\" class=\"data row18 col0\" >BCR Biotab</td>\n",
       "      <td id=\"T_ff236_row18_col1\" class=\"data row18 col1\" >76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row19_col0\" class=\"data row19 col0\" >XLSX</td>\n",
       "      <td id=\"T_ff236_row19_col1\" class=\"data row19 col1\" >7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row20_col0\" class=\"data row20 col0\" >PDF</td>\n",
       "      <td id=\"T_ff236_row20_col1\" class=\"data row20 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ff236_row21_col0\" class=\"data row21 col0\" >CDC JSON</td>\n",
       "      <td id=\"T_ff236_row21_col1\" class=\"data row21 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_c2df3_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_c2df3_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_c2df3_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row0_col0\" class=\"data row0 col0\" >Gene Level Copy Number Scores</td>\n",
       "      <td id=\"T_c2df3_row0_col1\" class=\"data row0 col1\" >809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row1_col0\" class=\"data row1 col0\" >Raw Simple Somatic Mutation</td>\n",
       "      <td id=\"T_c2df3_row1_col1\" class=\"data row1 col1\" >3397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row2_col0\" class=\"data row2 col0\" >None</td>\n",
       "      <td id=\"T_c2df3_row2_col1\" class=\"data row2 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row3_col0\" class=\"data row3 col0\" >Gene Expression Quantification</td>\n",
       "      <td id=\"T_c2df3_row3_col1\" class=\"data row3 col1\" >710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row4_col0\" class=\"data row4 col0\" >Text</td>\n",
       "      <td id=\"T_c2df3_row4_col1\" class=\"data row4 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row5_col0\" class=\"data row5 col0\" >Gene Level Copy Number</td>\n",
       "      <td id=\"T_c2df3_row5_col1\" class=\"data row5 col1\" >637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row6_col0\" class=\"data row6 col0\" >Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_c2df3_row6_col1\" class=\"data row6 col1\" >6043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row7_col0\" class=\"data row7 col0\" >Biospecimen Supplement</td>\n",
       "      <td id=\"T_c2df3_row7_col1\" class=\"data row7 col1\" >1755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row8_col0\" class=\"data row8 col0\" >Aggregated Somatic Mutation</td>\n",
       "      <td id=\"T_c2df3_row8_col1\" class=\"data row8 col1\" >732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row9_col0\" class=\"data row9 col0\" >Aligned Reads</td>\n",
       "      <td id=\"T_c2df3_row9_col1\" class=\"data row9 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row10_col0\" class=\"data row10 col0\" >Masked Intensities</td>\n",
       "      <td id=\"T_c2df3_row10_col1\" class=\"data row10 col1\" >1298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row11_col0\" class=\"data row11 col0\" >Splice Junction Quantification</td>\n",
       "      <td id=\"T_c2df3_row11_col1\" class=\"data row11 col1\" >710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row12_col0\" class=\"data row12 col0\" >Slide Image</td>\n",
       "      <td id=\"T_c2df3_row12_col1\" class=\"data row12 col1\" >1111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row13_col0\" class=\"data row13 col0\" >Isoform Expression Quantification</td>\n",
       "      <td id=\"T_c2df3_row13_col1\" class=\"data row13 col1\" >700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row14_col0\" class=\"data row14 col0\" >Masked Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_c2df3_row14_col1\" class=\"data row14 col1\" >1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row15_col0\" class=\"data row15 col0\" >Proprietary</td>\n",
       "      <td id=\"T_c2df3_row15_col1\" class=\"data row15 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row16_col0\" class=\"data row16 col0\" >Copy Number Segment</td>\n",
       "      <td id=\"T_c2df3_row16_col1\" class=\"data row16 col1\" >1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row17_col0\" class=\"data row17 col0\" >Open Standard</td>\n",
       "      <td id=\"T_c2df3_row17_col1\" class=\"data row17 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row18_col0\" class=\"data row18 col0\" >Clinical Supplement</td>\n",
       "      <td id=\"T_c2df3_row18_col1\" class=\"data row18 col1\" >773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row19_col0\" class=\"data row19 col0\" >Transcript Fusion</td>\n",
       "      <td id=\"T_c2df3_row19_col1\" class=\"data row19 col1\" >2278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row20_col0\" class=\"data row20 col0\" >Protein Expression Quantification</td>\n",
       "      <td id=\"T_c2df3_row20_col1\" class=\"data row20 col1\" >339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row21_col0\" class=\"data row21 col0\" >Masked Copy Number Segment</td>\n",
       "      <td id=\"T_c2df3_row21_col1\" class=\"data row21 col1\" >998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row22_col0\" class=\"data row22 col0\" >Allele-specific Copy Number Segment</td>\n",
       "      <td id=\"T_c2df3_row22_col1\" class=\"data row22 col1\" >495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row23_col0\" class=\"data row23 col0\" >Structural Rearrangement</td>\n",
       "      <td id=\"T_c2df3_row23_col1\" class=\"data row23 col1\" >284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row24_col0\" class=\"data row24 col0\" >Masked Somatic Mutation</td>\n",
       "      <td id=\"T_c2df3_row24_col1\" class=\"data row24 col1\" >549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row25_col0\" class=\"data row25 col0\" >miRNA Expression Quantification</td>\n",
       "      <td id=\"T_c2df3_row25_col1\" class=\"data row25 col1\" >700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row26_col0\" class=\"data row26 col0\" >Methylation Beta Value</td>\n",
       "      <td id=\"T_c2df3_row26_col1\" class=\"data row26 col1\" >649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c2df3_row27_col0\" class=\"data row27 col0\" >Pathology Report</td>\n",
       "      <td id=\"T_c2df3_row27_col1\" class=\"data row27 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111f47c0",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Julia decides that a good place to start would be with Slide Images. There's only 1111, so she should be able to quickly scan through them over the next few days and see if they will be useful. So she adds one more filter on her search:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d4a170b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3569 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1111</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m1111\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_02a1e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_02a1e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_02a1e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_02a1e_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_02a1e_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_f5ae7_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_f5ae7_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f5ae7_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_f5ae7_row0_col0\" class=\"data row0 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_f5ae7_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_436ee_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_436ee_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_436ee_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_436ee_row0_col0\" class=\"data row0 col0\" >SVS</td>\n",
       "      <td id=\"T_436ee_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_cf4e8_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_cf4e8_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_cf4e8_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_cf4e8_row0_col0\" class=\"data row0 col0\" >Slide Image</td>\n",
       "      <td id=\"T_cf4e8_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JustSlides = Q('file.data_type = \"Slide Image\"')\n",
    "NoadenoJustSlides = NoAdenoData.AND(JustSlides)\n",
    "NoadenoJustSlides.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e385faf",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Finally, Julia uses the pagenation function again to get all the slide files, and merges her metadata file with this file information. This way she will be able to review what phenotypes each slide is associated with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f45f6503",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3734 ms\n"
     ]
    }
   ],
   "source": [
    "slides = NoadenoJustSlides.researchsubject.file.run()\n",
    "slidesdf = pd.DataFrame()\n",
    "for i in slides.paginator(to_df=True):\n",
    "    slidesdf = pd.concat([slidesdf, i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0d1f7a21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>...</th>\n",
       "      <th>identifier_sub</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31dbb27e-40d9-421d-9dc1-e46014798478</td>\n",
       "      <td>[{'system': 'GDC', 'value': '31dbb27e-40d9-421...</td>\n",
       "      <td>TCGA-EA-A3HS-01A-01-TSA.70C289D1-DA4E-4351-AF6...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:31dbb27e-40d9-421d-9dc1-e4601479...</td>\n",
       "      <td>397383025</td>\n",
       "      <td>1bb62627b5439493aa58da2163adfcee</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EA-A3HS'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-13117.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c89ec421-a3f9-4f88-85a1-eaf0eb8337e5</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c89ec421-a3f9-4f8...</td>\n",
       "      <td>TCGA-EA-A3HS-01Z-00-DX1.F74FF288-29EB-4AEF-BC6...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:c89ec421-a3f9-4f88-85a1-eaf0eb83...</td>\n",
       "      <td>1899807371</td>\n",
       "      <td>12bedf131ac5414599272920366ec9cf</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EA-A3HS'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-13117.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>56af89c9-5059-425c-bf6a-49f50b61e128</td>\n",
       "      <td>[{'system': 'GDC', 'value': '56af89c9-5059-425...</td>\n",
       "      <td>TCGA-C5-A1BQ-01C-01-TS1.98EDB638-ED5E-47FB-833...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:56af89c9-5059-425c-bf6a-49f50b61...</td>\n",
       "      <td>157482701</td>\n",
       "      <td>00dd7016b7f336c716b7945e511b316a</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-C5-A1BQ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24018.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>604.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b6c8e60d-65a5-40af-83ca-1d9b5c4889ea</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b6c8e60d-65a5-40a...</td>\n",
       "      <td>TCGA-C5-A1BQ-01Z-00-DX1.72164F1F-956F-4C46-A7B...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:b6c8e60d-65a5-40af-83ca-1d9b5c48...</td>\n",
       "      <td>457023672</td>\n",
       "      <td>6c8c02b7002ec849793a1ab2051d1b10</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-C5-A1BQ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24018.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>604.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6779a257-513c-4ff8-ae34-28db3e832265</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6779a257-513c-4ff...</td>\n",
       "      <td>TCGA-3B-A9I1-01A-01-TS1.D5A4C7EE-A821-4792-A41...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-SARC</td>\n",
       "      <td>drs://dg.4DFC:6779a257-513c-4ff8-ae34-28db3e83...</td>\n",
       "      <td>93966751</td>\n",
       "      <td>c961d45bd551695d8327ff0bbe2f7017</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-3B-A9I1'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20629.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>567.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1106</th>\n",
       "      <td>1da21c74-a418-4379-ba1a-cdebfcc21c7e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1da21c74-a418-437...</td>\n",
       "      <td>TCGA-AJ-A2QM-01A-01-TSA.A7357359-DE36-4AB7-991...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:1da21c74-a418-4379-ba1a-cdebfcc2...</td>\n",
       "      <td>104552321</td>\n",
       "      <td>0604561d915143572662c3b58a5ff198</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AJ-A2QM'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24776.0</td>\n",
       "      <td>[tcga_ucec, TCGA-UCEC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1107</th>\n",
       "      <td>2adfdc32-8b41-418e-beb7-b0ab8544ba6b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2adfdc32-8b41-418...</td>\n",
       "      <td>TCGA-EK-A2PL-01A-01-TS1.2A3E42CE-BFFE-4761-90F...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:2adfdc32-8b41-418e-beb7-b0ab8544...</td>\n",
       "      <td>640115960</td>\n",
       "      <td>0043c9d66b32e2780c91530d56c8606a</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2PL'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-13384.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1108</th>\n",
       "      <td>977bd855-9430-4f41-97ca-6b5010b67076</td>\n",
       "      <td>[{'system': 'GDC', 'value': '977bd855-9430-4f4...</td>\n",
       "      <td>TCGA-KD-A5QU-01A-01-TS1.46D8773F-5E6F-49B1-963...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-SARC</td>\n",
       "      <td>drs://dg.4DFC:977bd855-9430-4f41-97ca-6b5010b6...</td>\n",
       "      <td>87301173</td>\n",
       "      <td>c24e969802885328178a764134045706</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-KD-A5QU'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-15018.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109</th>\n",
       "      <td>358728ce-157b-4002-8377-5391781a3d57</td>\n",
       "      <td>[{'system': 'GDC', 'value': '358728ce-157b-400...</td>\n",
       "      <td>TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...</td>\n",
       "      <td>116519647</td>\n",
       "      <td>592cb544b9ee7331d52d0ab68505d956</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17752.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110</th>\n",
       "      <td>bc892c68-06dc-4d3b-85b1-e185b2aa1427</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'bc892c68-06dc-4d3...</td>\n",
       "      <td>TCGA-EK-A3GN-01A-01-TS1.6F36A1A7-A8D7-434D-AC1...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:bc892c68-06dc-4d3b-85b1-e185b2aa...</td>\n",
       "      <td>159587192</td>\n",
       "      <td>94cdffcf5ca6ae6d759df7d64509d826</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A3GN'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17425.0</td>\n",
       "      <td>[tcga_cesc, TCGA-CESC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1111 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        id  \\\n",
       "0     31dbb27e-40d9-421d-9dc1-e46014798478   \n",
       "1     c89ec421-a3f9-4f88-85a1-eaf0eb8337e5   \n",
       "2     56af89c9-5059-425c-bf6a-49f50b61e128   \n",
       "3     b6c8e60d-65a5-40af-83ca-1d9b5c4889ea   \n",
       "4     6779a257-513c-4ff8-ae34-28db3e832265   \n",
       "...                                    ...   \n",
       "1106  1da21c74-a418-4379-ba1a-cdebfcc21c7e   \n",
       "1107  2adfdc32-8b41-418e-beb7-b0ab8544ba6b   \n",
       "1108  977bd855-9430-4f41-97ca-6b5010b67076   \n",
       "1109  358728ce-157b-4002-8377-5391781a3d57   \n",
       "1110  bc892c68-06dc-4d3b-85b1-e185b2aa1427   \n",
       "\n",
       "                                             identifier  \\\n",
       "0     [{'system': 'GDC', 'value': '31dbb27e-40d9-421...   \n",
       "1     [{'system': 'GDC', 'value': 'c89ec421-a3f9-4f8...   \n",
       "2     [{'system': 'GDC', 'value': '56af89c9-5059-425...   \n",
       "3     [{'system': 'GDC', 'value': 'b6c8e60d-65a5-40a...   \n",
       "4     [{'system': 'GDC', 'value': '6779a257-513c-4ff...   \n",
       "...                                                 ...   \n",
       "1106  [{'system': 'GDC', 'value': '1da21c74-a418-437...   \n",
       "1107  [{'system': 'GDC', 'value': '2adfdc32-8b41-418...   \n",
       "1108  [{'system': 'GDC', 'value': '977bd855-9430-4f4...   \n",
       "1109  [{'system': 'GDC', 'value': '358728ce-157b-400...   \n",
       "1110  [{'system': 'GDC', 'value': 'bc892c68-06dc-4d3...   \n",
       "\n",
       "                                                  label data_category  \\\n",
       "0     TCGA-EA-A3HS-01A-01-TSA.70C289D1-DA4E-4351-AF6...   Biospecimen   \n",
       "1     TCGA-EA-A3HS-01Z-00-DX1.F74FF288-29EB-4AEF-BC6...   Biospecimen   \n",
       "2     TCGA-C5-A1BQ-01C-01-TS1.98EDB638-ED5E-47FB-833...   Biospecimen   \n",
       "3     TCGA-C5-A1BQ-01Z-00-DX1.72164F1F-956F-4C46-A7B...   Biospecimen   \n",
       "4     TCGA-3B-A9I1-01A-01-TS1.D5A4C7EE-A821-4792-A41...   Biospecimen   \n",
       "...                                                 ...           ...   \n",
       "1106  TCGA-AJ-A2QM-01A-01-TSA.A7357359-DE36-4AB7-991...   Biospecimen   \n",
       "1107  TCGA-EK-A2PL-01A-01-TS1.2A3E42CE-BFFE-4761-90F...   Biospecimen   \n",
       "1108  TCGA-KD-A5QU-01A-01-TS1.46D8773F-5E6F-49B1-963...   Biospecimen   \n",
       "1109  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   \n",
       "1110  TCGA-EK-A3GN-01A-01-TS1.6F36A1A7-A8D7-434D-AC1...   Biospecimen   \n",
       "\n",
       "        data_type file_format associated_project  \\\n",
       "0     Slide Image         SVS          TCGA-CESC   \n",
       "1     Slide Image         SVS          TCGA-CESC   \n",
       "2     Slide Image         SVS          TCGA-CESC   \n",
       "3     Slide Image         SVS          TCGA-CESC   \n",
       "4     Slide Image         SVS          TCGA-SARC   \n",
       "...           ...         ...                ...   \n",
       "1106  Slide Image         SVS          TCGA-UCEC   \n",
       "1107  Slide Image         SVS          TCGA-CESC   \n",
       "1108  Slide Image         SVS          TCGA-SARC   \n",
       "1109  Slide Image         SVS          TCGA-CESC   \n",
       "1110  Slide Image         SVS          TCGA-CESC   \n",
       "\n",
       "                                                drs_uri   byte_size  \\\n",
       "0     drs://dg.4DFC:31dbb27e-40d9-421d-9dc1-e4601479...   397383025   \n",
       "1     drs://dg.4DFC:c89ec421-a3f9-4f88-85a1-eaf0eb83...  1899807371   \n",
       "2     drs://dg.4DFC:56af89c9-5059-425c-bf6a-49f50b61...   157482701   \n",
       "3     drs://dg.4DFC:b6c8e60d-65a5-40af-83ca-1d9b5c48...   457023672   \n",
       "4     drs://dg.4DFC:6779a257-513c-4ff8-ae34-28db3e83...    93966751   \n",
       "...                                                 ...         ...   \n",
       "1106  drs://dg.4DFC:1da21c74-a418-4379-ba1a-cdebfcc2...   104552321   \n",
       "1107  drs://dg.4DFC:2adfdc32-8b41-418e-beb7-b0ab8544...   640115960   \n",
       "1108  drs://dg.4DFC:977bd855-9430-4f41-97ca-6b5010b6...    87301173   \n",
       "1109  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...   116519647   \n",
       "1110  drs://dg.4DFC:bc892c68-06dc-4d3b-85b1-e185b2aa...   159587192   \n",
       "\n",
       "                              checksum  ...  \\\n",
       "0     1bb62627b5439493aa58da2163adfcee  ...   \n",
       "1     12bedf131ac5414599272920366ec9cf  ...   \n",
       "2     00dd7016b7f336c716b7945e511b316a  ...   \n",
       "3     6c8c02b7002ec849793a1ab2051d1b10  ...   \n",
       "4     c961d45bd551695d8327ff0bbe2f7017  ...   \n",
       "...                                ...  ...   \n",
       "1106  0604561d915143572662c3b58a5ff198  ...   \n",
       "1107  0043c9d66b32e2780c91530d56c8606a  ...   \n",
       "1108  c24e969802885328178a764134045706  ...   \n",
       "1109  592cb544b9ee7331d52d0ab68505d956  ...   \n",
       "1110  94cdffcf5ca6ae6d759df7d64509d826  ...   \n",
       "\n",
       "                                         identifier_sub       species     sex  \\\n",
       "0     [{'system': 'GDC', 'value': 'TCGA-EA-A3HS'}, {...  homo sapiens  female   \n",
       "1     [{'system': 'GDC', 'value': 'TCGA-EA-A3HS'}, {...  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'TCGA-C5-A1BQ'}, {...  homo sapiens  female   \n",
       "3     [{'system': 'GDC', 'value': 'TCGA-C5-A1BQ'}, {...  homo sapiens  female   \n",
       "4     [{'system': 'GDC', 'value': 'TCGA-3B-A9I1'}, {...  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "1106  [{'system': 'GDC', 'value': 'TCGA-AJ-A2QM'}, {...  homo sapiens  female   \n",
       "1107  [{'system': 'GDC', 'value': 'TCGA-EK-A2PL'}, {...  homo sapiens  female   \n",
       "1108  [{'system': 'GDC', 'value': 'TCGA-KD-A5QU'}, {...  homo sapiens  female   \n",
       "1109  [{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...  homo sapiens  female   \n",
       "1110  [{'system': 'GDC', 'value': 'TCGA-EK-A3GN'}, {...  homo sapiens  female   \n",
       "\n",
       "                           race               ethnicity days_to_birth  \\\n",
       "0                         white  not hispanic or latino      -13117.0   \n",
       "1                         white  not hispanic or latino      -13117.0   \n",
       "2                         white  not hispanic or latino      -24018.0   \n",
       "3                         white  not hispanic or latino      -24018.0   \n",
       "4     black or african american  not hispanic or latino      -20629.0   \n",
       "...                         ...                     ...           ...   \n",
       "1106                      white  not hispanic or latino      -24776.0   \n",
       "1107                      white            not reported      -13384.0   \n",
       "1108                      white  not hispanic or latino      -15018.0   \n",
       "1109                      white  not hispanic or latino      -17752.0   \n",
       "1110                      white  not hispanic or latino      -17425.0   \n",
       "\n",
       "     subject_associated_project vital_status days_to_death cause_of_death  \n",
       "0        [tcga_cesc, TCGA-CESC]        Alive           NaN           None  \n",
       "1        [tcga_cesc, TCGA-CESC]        Alive           NaN           None  \n",
       "2        [tcga_cesc, TCGA-CESC]         Dead         604.0           None  \n",
       "3        [tcga_cesc, TCGA-CESC]         Dead         604.0           None  \n",
       "4        [tcga_sarc, TCGA-SARC]         Dead         567.0           None  \n",
       "...                         ...          ...           ...            ...  \n",
       "1106     [tcga_ucec, TCGA-UCEC]        Alive           NaN           None  \n",
       "1107     [tcga_cesc, TCGA-CESC]        Alive           NaN           None  \n",
       "1108     [tcga_sarc, TCGA-SARC]         Dead        1073.0           None  \n",
       "1109     [tcga_cesc, TCGA-CESC]        Alive           NaN           None  \n",
       "1110     [tcga_cesc, TCGA-CESC]        Alive           NaN           None  \n",
       "\n",
       "[1111 rows x 32 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata = pd.merge(slidesdf, \n",
    "                         allmetadata, \n",
    "                         left_on=(\"subject_id\",\"researchsubject_id\"),\n",
    "                         right_on=(\"subject_id\", \"id_rs\"),\n",
    "                         suffixes=(\"_slide\", \"_all\"))\n",
    "slidemetadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "306e479c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                             1111\n",
       "identifier                     1111\n",
       "label                          1111\n",
       "data_category                  1111\n",
       "data_type                      1111\n",
       "file_format                    1111\n",
       "associated_project             1111\n",
       "drs_uri                        1111\n",
       "byte_size                      1111\n",
       "checksum                       1111\n",
       "data_modality                  1111\n",
       "imaging_modality                  0\n",
       "dbgap_accession_number            0\n",
       "imaging_series                    0\n",
       "subject_id                     1111\n",
       "researchsubject_id             1111\n",
       "id_rs                          1111\n",
       "identifier_rs                  1111\n",
       "member_of_research_project     1111\n",
       "primary_diagnosis_condition    1111\n",
       "primary_diagnosis_site         1111\n",
       "id_sub                         1111\n",
       "identifier_sub                 1111\n",
       "species                        1111\n",
       "sex                            1099\n",
       "race                           1099\n",
       "ethnicity                      1099\n",
       "days_to_birth                  1093\n",
       "subject_associated_project     1111\n",
       "vital_status                   1099\n",
       "days_to_death                   366\n",
       "cause_of_death                    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "369a3068",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Julia saves this dataframe to a csv as well, and now she has all the information she needs to begin work on her project. She can use the `drs_id` column information to directly download the images she is interested in using a DRS resolver, or she can input the DRS IDs at a cloud workspace such as [Terra](https://terra.bio/) or the [Cancer Genomics Cloud](https://www.cancergenomicscloud.org/) to view the images online. In either case, she has all the metadata she needs to get started, and can save this notebook of her work in case she'd like to come back and modify her search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8ef6e0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "slidemetadata.to_csv(\"slidemetadata.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
