{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "614a478f",
   "metadata": {},
   "source": [
    "# Cohort Building"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8004e87",
   "metadata": {},
   "source": [
    "**Example use case:** \n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" alt=\"alt_text\" align=\"left\"\n",
    "\twidth=\"150\" height=\"150\" />\n",
    "Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e45a9b",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "\n",
    "Before Julia does any work, she needs to import these functions cdapython.\n",
    "She'll also need to import [pandas](https://pandas.pydata.org/) to get nice dataframes.\n",
    "Finally, she tells cdapython to report it's version so she can be sure she's using the one she means to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5265d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "import pandas as pd \n",
    "print(Q.get_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22f48d3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()\n",
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75eef23e",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "391bc9a7",
   "metadata": {},
   "source": [
    "## Finding Search Terms\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef0dd8e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['id',\n",
       " 'identifier.system',\n",
       " 'identifier.value',\n",
       " 'species',\n",
       " 'sex',\n",
       " 'race',\n",
       " 'ethnicity',\n",
       " 'days_to_birth',\n",
       " 'subject_associated_project',\n",
       " 'vital_status',\n",
       " 'days_to_death',\n",
       " 'cause_of_death',\n",
       " 'ResearchSubject.id',\n",
       " 'ResearchSubject.identifier.system',\n",
       " 'ResearchSubject.identifier.value',\n",
       " 'ResearchSubject.member_of_research_project',\n",
       " 'ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles',\n",
       " 'ResearchSubject.Specimen.id',\n",
       " 'ResearchSubject.Specimen.identifier.system',\n",
       " 'ResearchSubject.Specimen.identifier.value',\n",
       " 'ResearchSubject.Specimen.associated_project',\n",
       " 'ResearchSubject.Specimen.days_to_collection',\n",
       " 'ResearchSubject.Specimen.primary_disease_type',\n",
       " 'ResearchSubject.Specimen.anatomical_site',\n",
       " 'ResearchSubject.Specimen.source_material_type',\n",
       " 'ResearchSubject.Specimen.specimen_type',\n",
       " 'ResearchSubject.Specimen.derived_from_specimen',\n",
       " 'ResearchSubject.Specimen.derived_from_subject',\n",
       " 'File.id',\n",
       " 'File.identifier.system',\n",
       " 'File.identifier.value',\n",
       " 'File.label',\n",
       " 'File.data_category',\n",
       " 'File.data_type',\n",
       " 'File.file_format',\n",
       " 'File.associated_project',\n",
       " 'File.drs_uri',\n",
       " 'File.byte_size',\n",
       " 'File.checksum',\n",
       " 'File.data_modality',\n",
       " 'File.imaging_modality',\n",
       " 'File.dbgap_accession_number',\n",
       " 'File.imaging_series']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05eba2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "536970c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list(filters=\"diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63b4cf0",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "\n",
    "To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retreiving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.\n",
    "    \n",
    "</div>\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Since Julia is interested specificially in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4527dde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Brain',\n",
       " 'Cervix',\n",
       " 'Head - Face Or Neck, Nos',\n",
       " 'Lymph Node(s) Paraaortic',\n",
       " 'Other',\n",
       " 'Pelvis',\n",
       " 'Spine',\n",
       " 'Unknown']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\").to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "740e5955",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Abdomen',\n",
       " 'Abdomen, Mediastinum',\n",
       " 'Abdomen, Pelvis',\n",
       " 'Adrenal Glands',\n",
       " 'Adrenal gland',\n",
       " 'Anus and anal canal',\n",
       " 'Base of tongue',\n",
       " 'Bile Duct',\n",
       " 'Bladder',\n",
       " 'Bones, joints and articular cartilage of limbs',\n",
       " 'Bones, joints and articular cartilage of other and unspecified sites',\n",
       " 'Brain',\n",
       " 'Breast',\n",
       " 'Bronchus and lung',\n",
       " 'Cervix',\n",
       " 'Cervix uteri',\n",
       " 'Chest',\n",
       " 'Chest-Abdomen-Pelvis, Leg, TSpine',\n",
       " 'Colon',\n",
       " 'Connective, subcutaneous and other soft tissues',\n",
       " 'Corpus uteri',\n",
       " 'Ear',\n",
       " 'Esophagus',\n",
       " 'Extremities',\n",
       " 'Eye and adnexa',\n",
       " 'Floor of mouth',\n",
       " 'Gallbladder',\n",
       " 'Gum',\n",
       " 'Head',\n",
       " 'Head and Neck',\n",
       " 'Head-Neck',\n",
       " 'Heart, mediastinum, and pleura',\n",
       " 'Hematopoietic and reticuloendothelial systems',\n",
       " 'Hypopharynx',\n",
       " 'Intraocular',\n",
       " 'Kidney',\n",
       " 'Larynx',\n",
       " 'Lip',\n",
       " 'Liver',\n",
       " 'Liver and intrahepatic bile ducts',\n",
       " 'Lung',\n",
       " 'Lung Phantom',\n",
       " 'Lymph nodes',\n",
       " 'Marrow, Blood',\n",
       " 'Meninges',\n",
       " 'Mesothelium',\n",
       " 'Nasal cavity and middle ear',\n",
       " 'Nasopharynx',\n",
       " 'Not Reported',\n",
       " 'Oropharynx',\n",
       " 'Other and ill-defined digestive organs',\n",
       " 'Other and ill-defined sites',\n",
       " 'Other and ill-defined sites in lip, oral cavity and pharynx',\n",
       " 'Other and ill-defined sites within respiratory system and intrathoracic organs',\n",
       " 'Other and unspecified female genital organs',\n",
       " 'Other and unspecified major salivary glands',\n",
       " 'Other and unspecified male genital organs',\n",
       " 'Other and unspecified parts of biliary tract',\n",
       " 'Other and unspecified parts of mouth',\n",
       " 'Other and unspecified parts of tongue',\n",
       " 'Other and unspecified urinary organs',\n",
       " 'Other endocrine glands and related structures',\n",
       " 'Ovary',\n",
       " 'Palate',\n",
       " 'Pancreas',\n",
       " 'Pancreas ',\n",
       " 'Pelvis, Prostate, Anus',\n",
       " 'Penis',\n",
       " 'Peripheral nerves and autonomic nervous system',\n",
       " 'Phantom',\n",
       " 'Prostate',\n",
       " 'Prostate gland',\n",
       " 'Rectosigmoid junction',\n",
       " 'Rectum',\n",
       " 'Renal pelvis',\n",
       " 'Retroperitoneum and peritoneum',\n",
       " 'Skin',\n",
       " 'Small intestine',\n",
       " 'Spinal cord, cranial nerves, and other parts of central nervous system',\n",
       " 'Stomach',\n",
       " 'Testicles',\n",
       " 'Testis',\n",
       " 'Thymus',\n",
       " 'Thyroid',\n",
       " 'Thyroid gland',\n",
       " 'Tonsil',\n",
       " 'Trachea',\n",
       " 'Unknown',\n",
       " 'Ureter',\n",
       " 'Uterus',\n",
       " 'Uterus, NOS',\n",
       " 'Vagina',\n",
       " 'Various',\n",
       " 'Various (11 locations)',\n",
       " 'Vulva']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b005036b",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e6b8dc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia sees that \"treatment_anatomic_site\" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the \"primary_diagnosis_site\" results. As she was initially looking for \"uterine\", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the \"ResearchSubject.primary_diagnosis_site\" for 'uter' as that should cover all variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "31064125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix uteri', 'Corpus uteri', 'Uterus', 'Uterus, NOS']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"uter\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9311a49e",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Just to be sure, Julia also searches for any other instances of \"cervix\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2038a8cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix', 'Cervix uteri']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"cerv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c4de58",
   "metadata": {},
   "source": [
    "## Building a Query\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the \"treatment_anatomic_site\", only one term is of interest, so she uses the `=` operator to get only exact matches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "951fcc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = \"Cervix\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cb5f72",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "However, for \"primary_diagnosis_site\", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "36cfd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Dsite = Q('ResearchSubject.primary_diagnosis_site = \"%uter%\" OR ResearchSubject.primary_diagnosis_site = \"%cerv%\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349af6f2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia adds her two queries together into one large one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9f5f9e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA = Tsite.OR(Dsite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f5cb55",
   "metadata": {},
   "source": [
    "## Looking at Summary Data\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "355b1706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3741 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">specimen_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">40793</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "specimen_count : \u001b[1;36m40793\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">treatment_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3049</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "treatment_count : \u001b[1;36m3049\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">diagnosis_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3685</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "diagnosis_count : \u001b[1;36m3685\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">researchsubject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "researchsubject_count : \u001b[1;36m4869\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">subject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3742</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "subject_count : \u001b[1;36m3742\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.count.run()\n",
    "#specimen_count : 40793\n",
    "#treatment_count : 3049\n",
    "#diagnosis_count : 3685\n",
    "#researchsubject_count : 4869\n",
    "#subject_count : 3742"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ce25fc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "55b0cdeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3562 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m4869\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">324646</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m324646\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_50fde_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_50fde_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_50fde_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_50fde_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_50fde_row0_col1\" class=\"data row0 col1\" >3591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_50fde_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_50fde_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_50fde_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_50fde_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_a7892_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a7892_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a7892_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row0_col0\" class=\"data row0 col0\" >Adenomas and Adenocarcinomas</td>\n",
       "      <td id=\"T_a7892_row0_col1\" class=\"data row0 col1\" >1672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row1_col0\" class=\"data row1 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_a7892_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row2_col0\" class=\"data row2 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_a7892_row2_col1\" class=\"data row2 col1\" >188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row3_col0\" class=\"data row3 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_a7892_row3_col1\" class=\"data row3 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row4_col0\" class=\"data row4 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_a7892_row4_col1\" class=\"data row4 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row5_col0\" class=\"data row5 col0\" >Not Reported</td>\n",
       "      <td id=\"T_a7892_row5_col1\" class=\"data row5 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row6_col0\" class=\"data row6 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_a7892_row6_col1\" class=\"data row6 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row7_col0\" class=\"data row7 col0\" >None</td>\n",
       "      <td id=\"T_a7892_row7_col1\" class=\"data row7 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row8_col0\" class=\"data row8 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_a7892_row8_col1\" class=\"data row8 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row9_col0\" class=\"data row9 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_a7892_row9_col1\" class=\"data row9 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row10_col0\" class=\"data row10 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_a7892_row10_col1\" class=\"data row10 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row11_col0\" class=\"data row11 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_a7892_row11_col1\" class=\"data row11 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row12_col0\" class=\"data row12 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_a7892_row12_col1\" class=\"data row12 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row13_col0\" class=\"data row13 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_a7892_row13_col1\" class=\"data row13 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a7892_row14_col0\" class=\"data row14 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_a7892_row14_col1\" class=\"data row14 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_875eb_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_875eb_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_875eb_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_875eb_row0_col0\" class=\"data row0 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_875eb_row0_col1\" class=\"data row0 col1\" >915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_875eb_row1_col0\" class=\"data row1 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_875eb_row1_col1\" class=\"data row1 col1\" >780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_875eb_row2_col0\" class=\"data row2 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_875eb_row2_col1\" class=\"data row2 col1\" >2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_875eb_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_875eb_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_875eb_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_875eb_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.researchsubject.count.run()\n",
    "# system\tcount\n",
    "# GDC\t3591\n",
    "# PDC\t104\n",
    "# IDC\t1174 \n",
    "# primary_diagnosis_condition\tcount\n",
    "# Adenomas and Adenocarcinomas\t1672\n",
    "# Uterine Corpus Endometrial Carcinoma\t104\n",
    "# Myomatous Neoplasms\t188\n",
    "# Cystic, Mucinous and Serous Neoplasms\t487\n",
    "# Squamous Cell Neoplasms\t609\n",
    "# Not Reported\t12\n",
    "# Complex Mixed and Stromal Neoplasms\t320\n",
    "# None\t1175\n",
    "# Epithelial Neoplasms, NOS\t230\n",
    "# Soft Tissue Tumors and Sarcomas, NOS\t14\n",
    "# Complex Epithelial Neoplasms\t27\n",
    "# Trophoblastic neoplasms\t13\n",
    "# Neoplasms, NOS\t12\n",
    "# Mesonephromas\t5\n",
    "# Neuroepitheliomatous Neoplasms\t1 \n",
    "# primary_diagnosis_site\tcount\n",
    "# Cervix uteri\t915\n",
    "# Corpus uteri\t780\n",
    "# Uterus, NOS\t2000\n",
    "# Uterus\t867\n",
    "# Cervix\t307"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a323e2",
   "metadata": {},
   "source": [
    "## Refining Queries\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0d526198",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3560 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3197</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m3197\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">298263</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m298263\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_82b73_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_82b73_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_82b73_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_82b73_row0_col0\" class=\"data row0 col0\" >PDC</td>\n",
       "      <td id=\"T_82b73_row0_col1\" class=\"data row0 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_82b73_row1_col0\" class=\"data row1 col0\" >GDC</td>\n",
       "      <td id=\"T_82b73_row1_col1\" class=\"data row1 col1\" >1919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_82b73_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_82b73_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_eb83e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_eb83e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_eb83e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row0_col0\" class=\"data row0 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row0_col1\" class=\"data row0 col1\" >188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row1_col0\" class=\"data row1 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row1_col1\" class=\"data row1 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row2_col0\" class=\"data row2 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_eb83e_row2_col1\" class=\"data row2 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row3_col0\" class=\"data row3 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row3_col1\" class=\"data row3 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row4_col0\" class=\"data row4 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row4_col1\" class=\"data row4 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row5_col0\" class=\"data row5 col0\" >None</td>\n",
       "      <td id=\"T_eb83e_row5_col1\" class=\"data row5 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row6_col0\" class=\"data row6 col0\" >Not Reported</td>\n",
       "      <td id=\"T_eb83e_row6_col1\" class=\"data row6 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row7_col0\" class=\"data row7 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_eb83e_row7_col1\" class=\"data row7 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row8_col0\" class=\"data row8 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row8_col1\" class=\"data row8 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row9_col0\" class=\"data row9 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_eb83e_row9_col1\" class=\"data row9 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row10_col0\" class=\"data row10 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_eb83e_row10_col1\" class=\"data row10 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row11_col0\" class=\"data row11 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_eb83e_row11_col1\" class=\"data row11 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row12_col0\" class=\"data row12 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_eb83e_row12_col1\" class=\"data row12 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_eb83e_row13_col0\" class=\"data row13 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_eb83e_row13_col1\" class=\"data row13 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_95bdd_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_95bdd_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_95bdd_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_95bdd_row0_col0\" class=\"data row0 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_95bdd_row0_col1\" class=\"data row0 col1\" >688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_95bdd_row1_col0\" class=\"data row1 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_95bdd_row1_col1\" class=\"data row1 col1\" >373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_95bdd_row2_col0\" class=\"data row2 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_95bdd_row2_col1\" class=\"data row2 col1\" >962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_95bdd_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_95bdd_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_95bdd_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_95bdd_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Noadeno = Q('ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"')\n",
    "\n",
    "NoAdenoData = ALLDATA.AND(Noadeno)\n",
    "\n",
    "NoAdenoData.researchsubject.count.run()\n",
    "#   total : 3197    \n",
    "#    files : 298263   \n",
    "# system\tcount\n",
    "# PDC\t104\n",
    "# GDC\t1919\n",
    "# IDC\t1174 \n",
    "# primary_diagnosis_condition\tcount\n",
    "# Myomatous Neoplasms\t188\n",
    "# Squamous Cell Neoplasms\t609\n",
    "# Uterine Corpus Endometrial Carcinoma\t104\n",
    "# Cystic, Mucinous and Serous Neoplasms\t487\n",
    "# Complex Mixed and Stromal Neoplasms\t320\n",
    "# None\t1175\n",
    "# Not Reported\t12\n",
    "# Epithelial Neoplasms, NOS\t230\n",
    "# Complex Epithelial Neoplasms\t27\n",
    "# Soft Tissue Tumors and Sarcomas, NOS\t14\n",
    "# Neuroepitheliomatous Neoplasms\t1\n",
    "# Trophoblastic neoplasms\t13\n",
    "# Neoplasms, NOS\t12\n",
    "# Mesonephromas\t5 \n",
    "# primary_diagnosis_site\tcount\n",
    "# Cervix uteri\t688\n",
    "# Corpus uteri\t373\n",
    "# Uterus, NOS\t962\n",
    "# Uterus\t867\n",
    "# Cervix\t307"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a0191d",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work. Since she's mostly interested in looking at the kinds of data available from each endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d186b837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3622 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1f20b0c8-11c1-4a3c-84fc-d485aac49bc8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD16769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29d93fb1-0b3d-4d13-8799-2dcf3e14be04</td>\n",
       "      <td>[{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-006032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3bccf0e3-d467-4477-adfd-b04d71f3eb86</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3bccf0e3-d467-447...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-009614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3c25bf98-1a11-41dc-8d65-182b0e8c5978</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3c25bf98-1a11-41d...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0014679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4ed1616e-8c62-4104-bcbe-26d7059f04d0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4ed1616e-8c62-410...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-C5-A7X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>3789ae3f-286b-42f0-a805-320ad69ea1df</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3789ae3f-286b-42f...</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>TCGA-B5-A3S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3ac62513-b5f4-4063-aabd-dad08a1f56fb</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3ac62513-b5f4-406...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-EA-A3Y4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3ba89750-e310-4771-88e7-b5bf289354e0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3ba89750-e310-477...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>AD16507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>5040ec9a-5050-4295-b792-38863eded12b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '5040ec9a-5050-429...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>AD3542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>5be58e2e-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>[{'system': 'PDC', 'value': '5be58e2e-118a-11e...</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>C3L-00963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   \n",
       "1   29d93fb1-0b3d-4d13-8799-2dcf3e14be04   \n",
       "2   3bccf0e3-d467-4477-adfd-b04d71f3eb86   \n",
       "3   3c25bf98-1a11-41dc-8d65-182b0e8c5978   \n",
       "4   4ed1616e-8c62-4104-bcbe-26d7059f04d0   \n",
       "..                                   ...   \n",
       "95  3789ae3f-286b-42f0-a805-320ad69ea1df   \n",
       "96  3ac62513-b5f4-4063-aabd-dad08a1f56fb   \n",
       "97  3ba89750-e310-4771-88e7-b5bf289354e0   \n",
       "98  5040ec9a-5050-4295-b792-38863eded12b   \n",
       "99  5be58e2e-118a-11e9-afb9-0a9c39d33490   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   \n",
       "1   [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   \n",
       "2   [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   \n",
       "3   [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   \n",
       "4   [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '3789ae3f-286b-42f...   \n",
       "96  [{'system': 'GDC', 'value': '3ac62513-b5f4-406...   \n",
       "97  [{'system': 'GDC', 'value': '3ba89750-e310-477...   \n",
       "98  [{'system': 'GDC', 'value': '5040ec9a-5050-429...   \n",
       "99  [{'system': 'PDC', 'value': '5be58e2e-118a-11e...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                       FM-AD    Complex Mixed and Stromal Neoplasms   \n",
       "1                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "2                  GENIE-DFCI                    Myomatous Neoplasms   \n",
       "3                   GENIE-MSK                    Myomatous Neoplasms   \n",
       "4                   TCGA-CESC                Squamous Cell Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "95                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   \n",
       "96                  TCGA-CESC                Squamous Cell Neoplasms   \n",
       "97                      FM-AD                Squamous Cell Neoplasms   \n",
       "98                      FM-AD                Squamous Cell Neoplasms   \n",
       "99           CPTAC3-Discovery   Uterine Corpus Endometrial Carcinoma   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0             Uterus, NOS              AD16769  \n",
       "1             Uterus, NOS    GENIE-DFCI-006032  \n",
       "2             Uterus, NOS    GENIE-DFCI-009614  \n",
       "3             Uterus, NOS  GENIE-MSK-P-0014679  \n",
       "4            Cervix uteri         TCGA-C5-A7X3  \n",
       "..                    ...                  ...  \n",
       "95           Corpus uteri         TCGA-B5-A3S1  \n",
       "96           Cervix uteri         TCGA-EA-A3Y4  \n",
       "97           Cervix uteri              AD16507  \n",
       "98           Cervix uteri               AD3542  \n",
       "99            Uterus, NOS            C3L-00963  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086697b3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8d0f5e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3596 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD95</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD95'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-02894</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-02894'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3N-03005</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-03005'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-DFCI-002257</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22280.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-DFCI-003511</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24837.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TCGA-EY-A3QX</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EY-A3QX'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23663.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>989.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-N5-A4RU</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N5-A4RU'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20025.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1526.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-NG-A4VU</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-NG-A4VU'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23323.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>442.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-PG-A7D5</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-PG-A7D5'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23006.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-VS-A959</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A959'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-27897.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id                                         identifier  \\\n",
       "0                AD95               [{'system': 'GDC', 'value': 'AD95'}]   \n",
       "1           C3L-02894          [{'system': 'IDC', 'value': 'C3L-02894'}]   \n",
       "2           C3N-03005          [{'system': 'IDC', 'value': 'C3N-03005'}]   \n",
       "3   GENIE-DFCI-002257  [{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]   \n",
       "4   GENIE-DFCI-003511  [{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]   \n",
       "..                ...                                                ...   \n",
       "95       TCGA-EY-A3QX  [{'system': 'GDC', 'value': 'TCGA-EY-A3QX'}, {...   \n",
       "96       TCGA-N5-A4RU  [{'system': 'GDC', 'value': 'TCGA-N5-A4RU'}, {...   \n",
       "97       TCGA-NG-A4VU  [{'system': 'GDC', 'value': 'TCGA-NG-A4VU'}, {...   \n",
       "98       TCGA-PG-A7D5  [{'system': 'GDC', 'value': 'TCGA-PG-A7D5'}, {...   \n",
       "99       TCGA-VS-A959  [{'system': 'GDC', 'value': 'TCGA-VS-A959'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "0   homo sapiens  female               not reported            not reported   \n",
       "1   Homo sapiens    None                       None                    None   \n",
       "2   Homo sapiens    None                       None                    None   \n",
       "3   homo sapiens  female                      white  not hispanic or latino   \n",
       "4   homo sapiens  female                      white  not hispanic or latino   \n",
       "..           ...     ...                        ...                     ...   \n",
       "95  homo sapiens  female                      white  not hispanic or latino   \n",
       "96  homo sapiens  female                      white  not hispanic or latino   \n",
       "97  homo sapiens  female                      white  not hispanic or latino   \n",
       "98  homo sapiens  female  black or african american  not hispanic or latino   \n",
       "99  homo sapiens  female               not reported            not reported   \n",
       "\n",
       "    days_to_birth subject_associated_project  vital_status  days_to_death  \\\n",
       "0             NaN                    [FM-AD]  Not Reported            NaN   \n",
       "1             NaN               [cptac_ucec]          None            NaN   \n",
       "2             NaN               [cptac_ucec]          None            NaN   \n",
       "3        -22280.0               [GENIE-DFCI]  Not Reported            NaN   \n",
       "4        -24837.0               [GENIE-DFCI]  Not Reported            NaN   \n",
       "..            ...                        ...           ...            ...   \n",
       "95       -23663.0     [TCGA-UCEC, tcga_ucec]          Dead          989.0   \n",
       "96       -20025.0       [TCGA-UCS, tcga_ucs]          Dead         1526.0   \n",
       "97       -23323.0       [TCGA-UCS, tcga_ucs]          Dead          442.0   \n",
       "98       -23006.0     [TCGA-UCEC, tcga_ucec]         Alive            NaN   \n",
       "99       -27897.0     [TCGA-CESC, tcga_cesc]         Alive            NaN   \n",
       "\n",
       "   cause_of_death  \n",
       "0            None  \n",
       "1            None  \n",
       "2            None  \n",
       "3            None  \n",
       "4            None  \n",
       "..            ...  \n",
       "95           None  \n",
       "96           None  \n",
       "97           None  \n",
       "98           None  \n",
       "99           None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.subject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec76132",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "04e04136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4225 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0b3ca53a-102f-4e81-845f-cad1afeabcb9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b3ca53a-102f-4e8...</td>\n",
       "      <td>baff6124-e0ed-4ae8-abb2-614fc68e6728.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:0b3ca53a-102f-4e81-845f-cad1afea...</td>\n",
       "      <td>186898</td>\n",
       "      <td>c2d7d5de9497268525e9a25753e108e8</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>1f1933a7-64e4-4bb5-934d-e61ab9c7eb10</td>\n",
       "      <td>055ef10b-309e-4105-8379-ef6282d30c3a</td>\n",
       "      <td>HTMCP-03-06-02036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>308d0d13-ba05-46b4-a46d-ec02e9768f63</td>\n",
       "      <td>[{'system': 'GDC', 'value': '308d0d13-ba05-46b...</td>\n",
       "      <td>32ac189b-859b-4f44-8040-c68b5bfc499d.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:308d0d13-ba05-46b4-a46d-ec02e976...</td>\n",
       "      <td>64370</td>\n",
       "      <td>f7eb03972ab6cca2e63ee209598d46e7</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>fdb8c4e2-fc29-444e-a441-dded10f769fd</td>\n",
       "      <td>54d01763-63d0-43a7-85f2-d447c6d9950e</td>\n",
       "      <td>HTMCP-03-06-02076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a1e661f5-3158-4748-8e4c-c262f6981e05</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a1e661f5-3158-474...</td>\n",
       "      <td>TCGA-2W-A8YY-01A-21-A40H-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:a1e661f5-3158-4748-8e4c-c262f698...</td>\n",
       "      <td>23833</td>\n",
       "      <td>a2c830ca73e7b2dba189b70c7ae96c05</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>c9afd150-ea3a-4219-b634-6e097c3b2a4f</td>\n",
       "      <td>5aeac31a-176a-4f93-a376-a93a670821bb</td>\n",
       "      <td>TCGA-2W-A8YY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>a6de4f29-79fb-4139-b320-b664d2d5387c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a6de4f29-79fb-413...</td>\n",
       "      <td>16afeee8-917a-4682-abf6-3d599345fc28.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:a6de4f29-79fb-4139-b320-b664d2d5...</td>\n",
       "      <td>38046</td>\n",
       "      <td>714ef4e3d21edd894a5e4e972e983170</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>900e0e61-e387-4be4-ba67-99423a0a23aa</td>\n",
       "      <td>1ffd0e35-68fd-4eaa-8b2b-a337fbdff5b9</td>\n",
       "      <td>HTMCP-03-06-02026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b30cd89d-895b-4d7c-b2dc-9cb469c9b39f</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b30cd89d-895b-4d7...</td>\n",
       "      <td>c2ad8dfa-b36b-4bf8-8db2-1a3beb9ec181.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:b30cd89d-895b-4d7c-b2dc-9cb469c9...</td>\n",
       "      <td>48137</td>\n",
       "      <td>01d4836e1af4a4e60e36a8a40178b6ba</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>6d2955d3-764c-4457-9bd2-b4d94b26f80c</td>\n",
       "      <td>8859b1ba-2d81-43e5-a845-4966ca6866da</td>\n",
       "      <td>HTMCP-03-06-02320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>fed0013e-e5e2-4fce-b1f1-e7bc942b4a4a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'fed0013e-e5e2-4fc...</td>\n",
       "      <td>a415048b-118c-4e8e-a44d-ca9ea524eedd.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:fed0013e-e5e2-4fce-b1f1-e7bc942b...</td>\n",
       "      <td>109420</td>\n",
       "      <td>f63e7a05c74d325a1247d4e351ab477e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>ff4bae0a-b24e-461d-a802-28b27ad870bf</td>\n",
       "      <td>a284a345-42f1-4f3c-9833-13fc835e131a</td>\n",
       "      <td>HTMCP-03-06-02174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>0b5cd4dc-5081-466a-9bcc-bc717b3254af</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b5cd4dc-5081-466...</td>\n",
       "      <td>2db011f6-0005-4fab-ad90-a0e02afc1a0b.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:0b5cd4dc-5081-466a-9bcc-bc717b32...</td>\n",
       "      <td>67148</td>\n",
       "      <td>724e8649e0213a3ff98e4f47cffe324a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>b669fa70-1392-4a9c-9d12-a56eb50b6e8d</td>\n",
       "      <td>226914c5-e486-4041-9b0a-83ba8baae0e7</td>\n",
       "      <td>HTMCP-03-06-02003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>0f6bff1c-a0e0-47f6-b4a9-6eff171b7e82</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0f6bff1c-a0e0-47f...</td>\n",
       "      <td>e05cb0b1-3d3b-47d0-b513-aeeba47b313d.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:0f6bff1c-a0e0-47f6-b4a9-6eff171b...</td>\n",
       "      <td>106577</td>\n",
       "      <td>d7208a402ce015d4c5f8cc7ee5aaf290</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>4a72437d-3711-4e49-9751-c327ca836f18</td>\n",
       "      <td>29754644-2d3d-4e04-830c-4b1c845df651</td>\n",
       "      <td>HTMCP-03-06-02180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1844c0ab-f430-45bc-9865-4078cf57c5c4</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1844c0ab-f430-45b...</td>\n",
       "      <td>bcdf7a5f-2726-4b83-9a56-0ddf6e5d48c4.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:1844c0ab-f430-45bc-9865-4078cf57...</td>\n",
       "      <td>87697</td>\n",
       "      <td>0e5e07ff37a2fdccba492db6a9de1c7a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>9456911a-d588-436f-a77c-cc7754e218c7</td>\n",
       "      <td>57d7c3a9-0f1d-4ec1-8815-fea5d4565d3b</td>\n",
       "      <td>HTMCP-03-06-02435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1981e3bd-c534-4259-8124-cd37a51fb725</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1981e3bd-c534-425...</td>\n",
       "      <td>d6381f8b-f7e0-49c6-9b1b-481079e84410.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:1981e3bd-c534-4259-8124-cd37a51f...</td>\n",
       "      <td>60710</td>\n",
       "      <td>844eab2061cf2becd8860a63118b69d9</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>None</td>\n",
       "      <td>b714eb14-8230-42c3-9846-3942d1d27172</td>\n",
       "      <td>2c776c4f-fade-4035-8de6-e91c45c8347a</td>\n",
       "      <td>HTMCP-03-06-02089</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0b3ca53a-102f-4e81-845f-cad1afeabcb9   \n",
       "1   308d0d13-ba05-46b4-a46d-ec02e9768f63   \n",
       "2   a1e661f5-3158-4748-8e4c-c262f6981e05   \n",
       "3   a6de4f29-79fb-4139-b320-b664d2d5387c   \n",
       "4   b30cd89d-895b-4d7c-b2dc-9cb469c9b39f   \n",
       "..                                   ...   \n",
       "95  fed0013e-e5e2-4fce-b1f1-e7bc942b4a4a   \n",
       "96  0b5cd4dc-5081-466a-9bcc-bc717b3254af   \n",
       "97  0f6bff1c-a0e0-47f6-b4a9-6eff171b7e82   \n",
       "98  1844c0ab-f430-45bc-9865-4078cf57c5c4   \n",
       "99  1981e3bd-c534-4259-8124-cd37a51fb725   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '0b3ca53a-102f-4e8...   \n",
       "1   [{'system': 'GDC', 'value': '308d0d13-ba05-46b...   \n",
       "2   [{'system': 'GDC', 'value': 'a1e661f5-3158-474...   \n",
       "3   [{'system': 'GDC', 'value': 'a6de4f29-79fb-413...   \n",
       "4   [{'system': 'GDC', 'value': 'b30cd89d-895b-4d7...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': 'fed0013e-e5e2-4fc...   \n",
       "96  [{'system': 'GDC', 'value': '0b5cd4dc-5081-466...   \n",
       "97  [{'system': 'GDC', 'value': '0f6bff1c-a0e0-47f...   \n",
       "98  [{'system': 'GDC', 'value': '1844c0ab-f430-45b...   \n",
       "99  [{'system': 'GDC', 'value': '1981e3bd-c534-425...   \n",
       "\n",
       "                                                label  \\\n",
       "0   baff6124-e0ed-4ae8-abb2-614fc68e6728.wgs.BRASS...   \n",
       "1   32ac189b-859b-4f44-8040-c68b5bfc499d.wgs.BRASS...   \n",
       "2           TCGA-2W-A8YY-01A-21-A40H-20_RPPA_data.tsv   \n",
       "3   16afeee8-917a-4682-abf6-3d599345fc28.wgs.BRASS...   \n",
       "4   c2ad8dfa-b36b-4bf8-8db2-1a3beb9ec181.wgs.BRASS...   \n",
       "..                                                ...   \n",
       "95  a415048b-118c-4e8e-a44d-ca9ea524eedd.wgs.BRASS...   \n",
       "96  2db011f6-0005-4fab-ad90-a0e02afc1a0b.wgs.BRASS...   \n",
       "97  e05cb0b1-3d3b-47d0-b513-aeeba47b313d.wgs.BRASS...   \n",
       "98  bcdf7a5f-2726-4b83-9a56-0ddf6e5d48c4.wgs.BRASS...   \n",
       "99  d6381f8b-f7e0-49c6-9b1b-481079e84410.wgs.BRASS...   \n",
       "\n",
       "                   data_category                          data_type  \\\n",
       "0   Somatic Structural Variation           Structural Rearrangement   \n",
       "1   Somatic Structural Variation           Structural Rearrangement   \n",
       "2             Proteome Profiling  Protein Expression Quantification   \n",
       "3   Somatic Structural Variation           Structural Rearrangement   \n",
       "4   Somatic Structural Variation           Structural Rearrangement   \n",
       "..                           ...                                ...   \n",
       "95  Somatic Structural Variation           Structural Rearrangement   \n",
       "96  Somatic Structural Variation           Structural Rearrangement   \n",
       "97  Somatic Structural Variation           Structural Rearrangement   \n",
       "98  Somatic Structural Variation           Structural Rearrangement   \n",
       "99  Somatic Structural Variation           Structural Rearrangement   \n",
       "\n",
       "   file_format associated_project  \\\n",
       "0        BEDPE      CGCI-HTMCP-CC   \n",
       "1          VCF      CGCI-HTMCP-CC   \n",
       "2          TSV          TCGA-CESC   \n",
       "3          VCF      CGCI-HTMCP-CC   \n",
       "4          VCF      CGCI-HTMCP-CC   \n",
       "..         ...                ...   \n",
       "95         VCF      CGCI-HTMCP-CC   \n",
       "96         VCF      CGCI-HTMCP-CC   \n",
       "97       BEDPE      CGCI-HTMCP-CC   \n",
       "98       BEDPE      CGCI-HTMCP-CC   \n",
       "99         VCF      CGCI-HTMCP-CC   \n",
       "\n",
       "                                              drs_uri  byte_size  \\\n",
       "0   drs://dg.4DFC:0b3ca53a-102f-4e81-845f-cad1afea...     186898   \n",
       "1   drs://dg.4DFC:308d0d13-ba05-46b4-a46d-ec02e976...      64370   \n",
       "2   drs://dg.4DFC:a1e661f5-3158-4748-8e4c-c262f698...      23833   \n",
       "3   drs://dg.4DFC:a6de4f29-79fb-4139-b320-b664d2d5...      38046   \n",
       "4   drs://dg.4DFC:b30cd89d-895b-4d7c-b2dc-9cb469c9...      48137   \n",
       "..                                                ...        ...   \n",
       "95  drs://dg.4DFC:fed0013e-e5e2-4fce-b1f1-e7bc942b...     109420   \n",
       "96  drs://dg.4DFC:0b5cd4dc-5081-466a-9bcc-bc717b32...      67148   \n",
       "97  drs://dg.4DFC:0f6bff1c-a0e0-47f6-b4a9-6eff171b...     106577   \n",
       "98  drs://dg.4DFC:1844c0ab-f430-45bc-9865-4078cf57...      87697   \n",
       "99  drs://dg.4DFC:1981e3bd-c534-4259-8124-cd37a51f...      60710   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   c2d7d5de9497268525e9a25753e108e8       Genomic             None   \n",
       "1   f7eb03972ab6cca2e63ee209598d46e7       Genomic             None   \n",
       "2   a2c830ca73e7b2dba189b70c7ae96c05       Genomic             None   \n",
       "3   714ef4e3d21edd894a5e4e972e983170       Genomic             None   \n",
       "4   01d4836e1af4a4e60e36a8a40178b6ba       Genomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  f63e7a05c74d325a1247d4e351ab477e       Genomic             None   \n",
       "96  724e8649e0213a3ff98e4f47cffe324a       Genomic             None   \n",
       "97  d7208a402ce015d4c5f8cc7ee5aaf290       Genomic             None   \n",
       "98  0e5e07ff37a2fdccba492db6a9de1c7a       Genomic             None   \n",
       "99  844eab2061cf2becd8860a63118b69d9       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number imaging_series  \\\n",
       "0               phs000528           None   \n",
       "1               phs000528           None   \n",
       "2                    None           None   \n",
       "3               phs000528           None   \n",
       "4               phs000528           None   \n",
       "..                    ...            ...   \n",
       "95              phs000528           None   \n",
       "96              phs000528           None   \n",
       "97              phs000528           None   \n",
       "98              phs000528           None   \n",
       "99              phs000528           None   \n",
       "\n",
       "             researchsubject_specimen_id  \\\n",
       "0   1f1933a7-64e4-4bb5-934d-e61ab9c7eb10   \n",
       "1   fdb8c4e2-fc29-444e-a441-dded10f769fd   \n",
       "2   c9afd150-ea3a-4219-b634-6e097c3b2a4f   \n",
       "3   900e0e61-e387-4be4-ba67-99423a0a23aa   \n",
       "4   6d2955d3-764c-4457-9bd2-b4d94b26f80c   \n",
       "..                                   ...   \n",
       "95  ff4bae0a-b24e-461d-a802-28b27ad870bf   \n",
       "96  b669fa70-1392-4a9c-9d12-a56eb50b6e8d   \n",
       "97  4a72437d-3711-4e49-9751-c327ca836f18   \n",
       "98  9456911a-d588-436f-a77c-cc7754e218c7   \n",
       "99  b714eb14-8230-42c3-9846-3942d1d27172   \n",
       "\n",
       "                      researchsubject_id         subject_id  \n",
       "0   055ef10b-309e-4105-8379-ef6282d30c3a  HTMCP-03-06-02036  \n",
       "1   54d01763-63d0-43a7-85f2-d447c6d9950e  HTMCP-03-06-02076  \n",
       "2   5aeac31a-176a-4f93-a376-a93a670821bb       TCGA-2W-A8YY  \n",
       "3   1ffd0e35-68fd-4eaa-8b2b-a337fbdff5b9  HTMCP-03-06-02026  \n",
       "4   8859b1ba-2d81-43e5-a845-4966ca6866da  HTMCP-03-06-02320  \n",
       "..                                   ...                ...  \n",
       "95  a284a345-42f1-4f3c-9833-13fc835e131a  HTMCP-03-06-02174  \n",
       "96  226914c5-e486-4041-9b0a-83ba8baae0e7  HTMCP-03-06-02003  \n",
       "97  29754644-2d3d-4e04-830c-4b1c845df651  HTMCP-03-06-02180  \n",
       "98  57d7c3a9-0f1d-4ec1-8815-fea5d4565d3b  HTMCP-03-06-02435  \n",
       "99  2c776c4f-fade-4035-8de6-e91c45c8347a  HTMCP-03-06-02089  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.file.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf9f2d3",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6aadbe",
   "metadata": {},
   "source": [
    "## Working with Results (pagination)\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c2cec2bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3592 ms\n"
     ]
    }
   ],
   "source": [
    "researchsubs = NoAdenoData.researchsubject.run()\n",
    "rsdf = pd.DataFrame()\n",
    "for i in researchsubs.paginator(to_df=True):\n",
    "    rsdf = pd.concat([rsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a1258057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3596 ms\n"
     ]
    }
   ],
   "source": [
    "subs = NoAdenoData.subject.run()\n",
    "subsdf = pd.DataFrame()\n",
    "for i in subs.paginator(to_df=True):\n",
    "    subsdf = pd.concat([subsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "04cd73df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1f20b0c8-11c1-4a3c-84fc-d485aac49bc8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD16769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29d93fb1-0b3d-4d13-8799-2dcf3e14be04</td>\n",
       "      <td>[{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-006032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3bccf0e3-d467-4477-adfd-b04d71f3eb86</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3bccf0e3-d467-447...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-009614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3c25bf98-1a11-41dc-8d65-182b0e8c5978</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3c25bf98-1a11-41d...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0014679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4ed1616e-8c62-4104-bcbe-26d7059f04d0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4ed1616e-8c62-410...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-C5-A7X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0000080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>d5ee39e8-a1d2-4244-9c38-07d7c5c1baba</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd5ee39e8-a1d2-424...</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>TCGA-B5-A5OE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>e5e66e4f-46af-4d88-a238-a8eb8e9dca09</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-DFCI-037165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>f8c840a8-2891-4c72-b9d8-9eb9c4351cf7</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-UHN-242802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>fbd1ffac-d944-4356-b224-663696112a0b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'fbd1ffac-d944-435...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0012305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   \n",
       "1   29d93fb1-0b3d-4d13-8799-2dcf3e14be04   \n",
       "2   3bccf0e3-d467-4477-adfd-b04d71f3eb86   \n",
       "3   3c25bf98-1a11-41dc-8d65-182b0e8c5978   \n",
       "4   4ed1616e-8c62-4104-bcbe-26d7059f04d0   \n",
       "..                                   ...   \n",
       "92  c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c   \n",
       "93  d5ee39e8-a1d2-4244-9c38-07d7c5c1baba   \n",
       "94  e5e66e4f-46af-4d88-a238-a8eb8e9dca09   \n",
       "95  f8c840a8-2891-4c72-b9d8-9eb9c4351cf7   \n",
       "96  fbd1ffac-d944-4356-b224-663696112a0b   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   \n",
       "1   [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   \n",
       "2   [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   \n",
       "3   [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   \n",
       "4   [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   \n",
       "..                                                ...   \n",
       "92  [{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...   \n",
       "93  [{'system': 'GDC', 'value': 'd5ee39e8-a1d2-424...   \n",
       "94  [{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...   \n",
       "95  [{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...   \n",
       "96  [{'system': 'GDC', 'value': 'fbd1ffac-d944-435...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                       FM-AD    Complex Mixed and Stromal Neoplasms   \n",
       "1                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "2                  GENIE-DFCI                    Myomatous Neoplasms   \n",
       "3                   GENIE-MSK                    Myomatous Neoplasms   \n",
       "4                   TCGA-CESC                Squamous Cell Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "92                  GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "93                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   \n",
       "94                 GENIE-DFCI              Epithelial Neoplasms, NOS   \n",
       "95                  GENIE-UHN  Cystic, Mucinous and Serous Neoplasms   \n",
       "96                  GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0             Uterus, NOS              AD16769  \n",
       "1             Uterus, NOS    GENIE-DFCI-006032  \n",
       "2             Uterus, NOS    GENIE-DFCI-009614  \n",
       "3             Uterus, NOS  GENIE-MSK-P-0014679  \n",
       "4            Cervix uteri         TCGA-C5-A7X3  \n",
       "..                    ...                  ...  \n",
       "92            Uterus, NOS  GENIE-MSK-P-0000080  \n",
       "93           Corpus uteri         TCGA-B5-A5OE  \n",
       "94           Corpus uteri    GENIE-DFCI-037165  \n",
       "95            Uterus, NOS     GENIE-UHN-242802  \n",
       "96           Corpus uteri  GENIE-MSK-P-0012305  \n",
       "\n",
       "[3197 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rsdf # view the researchsubject dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "92a6f811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD95</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD95'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-02894</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-02894'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3N-03005</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-03005'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-DFCI-002257</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22280.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-DFCI-003511</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24837.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-D1-A1NZ</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-D1-A1NZ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22027.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-D1-A2G5</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-D1-A2G5'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-18322.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-IS-A3K7</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-IS-A3K7'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-23022.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-NF-A4WU</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-NF-A4WU'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21901.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-X2-A95T</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-X2-A95T'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-18079.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2609 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id                                         identifier  \\\n",
       "0                AD95               [{'system': 'GDC', 'value': 'AD95'}]   \n",
       "1           C3L-02894          [{'system': 'IDC', 'value': 'C3L-02894'}]   \n",
       "2           C3N-03005          [{'system': 'IDC', 'value': 'C3N-03005'}]   \n",
       "3   GENIE-DFCI-002257  [{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]   \n",
       "4   GENIE-DFCI-003511  [{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]   \n",
       "..                ...                                                ...   \n",
       "4        TCGA-D1-A1NZ  [{'system': 'GDC', 'value': 'TCGA-D1-A1NZ'}, {...   \n",
       "5        TCGA-D1-A2G5  [{'system': 'GDC', 'value': 'TCGA-D1-A2G5'}, {...   \n",
       "6        TCGA-IS-A3K7  [{'system': 'GDC', 'value': 'TCGA-IS-A3K7'}, {...   \n",
       "7        TCGA-NF-A4WU  [{'system': 'GDC', 'value': 'TCGA-NF-A4WU'}, {...   \n",
       "8        TCGA-X2-A95T  [{'system': 'GDC', 'value': 'TCGA-X2-A95T'}, {...   \n",
       "\n",
       "         species     sex          race               ethnicity  days_to_birth  \\\n",
       "0   homo sapiens  female  not reported            not reported            NaN   \n",
       "1   Homo sapiens    None          None                    None            NaN   \n",
       "2   Homo sapiens    None          None                    None            NaN   \n",
       "3   homo sapiens  female         white  not hispanic or latino       -22280.0   \n",
       "4   homo sapiens  female         white  not hispanic or latino       -24837.0   \n",
       "..           ...     ...           ...                     ...            ...   \n",
       "4   homo sapiens  female         white  not hispanic or latino       -22027.0   \n",
       "5   homo sapiens  female         white  not hispanic or latino       -18322.0   \n",
       "6   homo sapiens  female         white            not reported       -23022.0   \n",
       "7   homo sapiens  female         white  not hispanic or latino       -21901.0   \n",
       "8   homo sapiens  female         white            not reported       -18079.0   \n",
       "\n",
       "   subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                     [FM-AD]  Not Reported            NaN           None  \n",
       "1                [cptac_ucec]          None            NaN           None  \n",
       "2                [cptac_ucec]          None            NaN           None  \n",
       "3                [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "4                [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "..                        ...           ...            ...            ...  \n",
       "4      [TCGA-UCEC, tcga_ucec]         Alive            NaN           None  \n",
       "5      [TCGA-UCEC, tcga_ucec]         Alive            NaN           None  \n",
       "6      [tcga_sarc, TCGA-SARC]         Alive            NaN           None  \n",
       "7        [TCGA-UCS, tcga_ucs]         Alive            NaN           None  \n",
       "8      [tcga_sarc, TCGA-SARC]         Alive            NaN           None  \n",
       "\n",
       "[2609 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsdf # view the subject dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bcbe86",
   "metadata": {},
   "source": [
    "## Merging Results across Endpoints\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Then Julia uses the `id` fields in each result to merge them together into one big dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9b7a3383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_x</th>\n",
       "      <th>identifier_x</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>id_y</th>\n",
       "      <th>identifier_y</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1f20b0c8-11c1-4a3c-84fc-d485aac49bc8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD16769</td>\n",
       "      <td>AD16769</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD16769'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29d93fb1-0b3d-4d13-8799-2dcf3e14be04</td>\n",
       "      <td>[{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-006032</td>\n",
       "      <td>GENIE-DFCI-006032</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-006032'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-31046.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3bccf0e3-d467-4477-adfd-b04d71f3eb86</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3bccf0e3-d467-447...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-009614</td>\n",
       "      <td>GENIE-DFCI-009614</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-009614'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17897.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3c25bf98-1a11-41dc-8d65-182b0e8c5978</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3c25bf98-1a11-41d...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0014679</td>\n",
       "      <td>GENIE-MSK-P-0014679</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00146...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-16071.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4ed1616e-8c62-4104-bcbe-26d7059f04d0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4ed1616e-8c62-410...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-C5-A7X3</td>\n",
       "      <td>TCGA-C5-A7X3</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-C5-A7X3'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-25665.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>284.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>baf1b1f8-36ae-4370-98dd-23bda7ed147e</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'baf1b1f8-36ae-437...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0005695</td>\n",
       "      <td>GENIE-MSK-P-0005695</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00056...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24471.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193</th>\n",
       "      <td>c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0000080</td>\n",
       "      <td>GENIE-MSK-P-0000080</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00000...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24837.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3194</th>\n",
       "      <td>e5e66e4f-46af-4d88-a238-a8eb8e9dca09</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-DFCI-037165</td>\n",
       "      <td>GENIE-DFCI-037165</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-037165'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24471.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3195</th>\n",
       "      <td>f8c840a8-2891-4c72-b9d8-9eb9c4351cf7</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-UHN-242802</td>\n",
       "      <td>GENIE-UHN-242802</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-UHN-242802'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not allowed to collect</td>\n",
       "      <td>-21915.0</td>\n",
       "      <td>[GENIE-UHN]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3196</th>\n",
       "      <td>fbd1ffac-d944-4356-b224-663696112a0b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'fbd1ffac-d944-435...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-MSK-P-0012305</td>\n",
       "      <td>GENIE-MSK-P-0012305</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00123...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-26298.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id_x  \\\n",
       "0     1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   \n",
       "1     29d93fb1-0b3d-4d13-8799-2dcf3e14be04   \n",
       "2     3bccf0e3-d467-4477-adfd-b04d71f3eb86   \n",
       "3     3c25bf98-1a11-41dc-8d65-182b0e8c5978   \n",
       "4     4ed1616e-8c62-4104-bcbe-26d7059f04d0   \n",
       "...                                    ...   \n",
       "3192  baf1b1f8-36ae-4370-98dd-23bda7ed147e   \n",
       "3193  c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c   \n",
       "3194  e5e66e4f-46af-4d88-a238-a8eb8e9dca09   \n",
       "3195  f8c840a8-2891-4c72-b9d8-9eb9c4351cf7   \n",
       "3196  fbd1ffac-d944-4356-b224-663696112a0b   \n",
       "\n",
       "                                           identifier_x  \\\n",
       "0     [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   \n",
       "1     [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   \n",
       "2     [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   \n",
       "3     [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   \n",
       "4     [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   \n",
       "...                                                 ...   \n",
       "3192  [{'system': 'GDC', 'value': 'baf1b1f8-36ae-437...   \n",
       "3193  [{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...   \n",
       "3194  [{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...   \n",
       "3195  [{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...   \n",
       "3196  [{'system': 'GDC', 'value': 'fbd1ffac-d944-435...   \n",
       "\n",
       "     member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                         FM-AD    Complex Mixed and Stromal Neoplasms   \n",
       "1                    GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   \n",
       "2                    GENIE-DFCI                    Myomatous Neoplasms   \n",
       "3                     GENIE-MSK                    Myomatous Neoplasms   \n",
       "4                     TCGA-CESC                Squamous Cell Neoplasms   \n",
       "...                         ...                                    ...   \n",
       "3192                  GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "3193                  GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "3194                 GENIE-DFCI              Epithelial Neoplasms, NOS   \n",
       "3195                  GENIE-UHN  Cystic, Mucinous and Serous Neoplasms   \n",
       "3196                  GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "\n",
       "     primary_diagnosis_site           subject_id                 id_y  \\\n",
       "0               Uterus, NOS              AD16769              AD16769   \n",
       "1               Uterus, NOS    GENIE-DFCI-006032    GENIE-DFCI-006032   \n",
       "2               Uterus, NOS    GENIE-DFCI-009614    GENIE-DFCI-009614   \n",
       "3               Uterus, NOS  GENIE-MSK-P-0014679  GENIE-MSK-P-0014679   \n",
       "4              Cervix uteri         TCGA-C5-A7X3         TCGA-C5-A7X3   \n",
       "...                     ...                  ...                  ...   \n",
       "3192            Uterus, NOS  GENIE-MSK-P-0005695  GENIE-MSK-P-0005695   \n",
       "3193            Uterus, NOS  GENIE-MSK-P-0000080  GENIE-MSK-P-0000080   \n",
       "3194           Corpus uteri    GENIE-DFCI-037165    GENIE-DFCI-037165   \n",
       "3195            Uterus, NOS     GENIE-UHN-242802     GENIE-UHN-242802   \n",
       "3196           Corpus uteri  GENIE-MSK-P-0012305  GENIE-MSK-P-0012305   \n",
       "\n",
       "                                           identifier_y       species     sex  \\\n",
       "0               [{'system': 'GDC', 'value': 'AD16769'}]  homo sapiens  female   \n",
       "1     [{'system': 'GDC', 'value': 'GENIE-DFCI-006032'}]  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'GENIE-DFCI-009614'}]  homo sapiens  female   \n",
       "3     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00146...  homo sapiens  female   \n",
       "4     [{'system': 'GDC', 'value': 'TCGA-C5-A7X3'}, {...  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "3192  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00056...  homo sapiens  female   \n",
       "3193  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00000...  homo sapiens  female   \n",
       "3194  [{'system': 'GDC', 'value': 'GENIE-DFCI-037165'}]  homo sapiens  female   \n",
       "3195   [{'system': 'GDC', 'value': 'GENIE-UHN-242802'}]  homo sapiens  female   \n",
       "3196  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00123...  homo sapiens  female   \n",
       "\n",
       "                           race               ethnicity  days_to_birth  \\\n",
       "0                  not reported            not reported            NaN   \n",
       "1                         white  not hispanic or latino       -31046.0   \n",
       "2                         white  not hispanic or latino       -17897.0   \n",
       "3                         white  not hispanic or latino       -16071.0   \n",
       "4                         white  not hispanic or latino       -25665.0   \n",
       "...                         ...                     ...            ...   \n",
       "3192                      asian  not hispanic or latino       -24471.0   \n",
       "3193                      white  not hispanic or latino       -24837.0   \n",
       "3194                      white  not hispanic or latino       -24471.0   \n",
       "3195  black or african american  not allowed to collect       -21915.0   \n",
       "3196                      white  not hispanic or latino       -26298.0   \n",
       "\n",
       "     subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                       [FM-AD]  Not Reported            NaN           None  \n",
       "1                  [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "2                  [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3                   [GENIE-MSK]  Not Reported            NaN           None  \n",
       "4        [TCGA-CESC, tcga_cesc]          Dead          284.0           None  \n",
       "...                         ...           ...            ...            ...  \n",
       "3192                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "3193                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "3194               [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3195                [GENIE-UHN]  Not Reported            NaN           None  \n",
       "3196                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "\n",
       "[3197 rows x 17 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmetadata = pd.merge(rsdf,\n",
    "                subsdf,\n",
    "                left_on=\"subject_id\",\n",
    "                right_on='id')\n",
    "\n",
    "allmetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024da831",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "And saves it out to a csv so she can browse it with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6628de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata.to_csv(\"allmetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246644d3",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia knows from her subject count summary that there are more than 200,000 files associated with her subjects, which is likely far more than she needs. To help her decide what files she wants, Julia uses endpoint chaining to get summary information about the files that are assigned to researchsubjects for her search criteria:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ae1ae079",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3505 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">298263</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   total : \u001b[1;36m298263\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_54982_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_54982_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_54982_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_54982_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_54982_row0_col1\" class=\"data row0 col1\" >31274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54982_row1_col0\" class=\"data row1 col0\" >IDC</td>\n",
       "      <td id=\"T_54982_row1_col1\" class=\"data row1 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54982_row2_col0\" class=\"data row2 col0\" >PDC</td>\n",
       "      <td id=\"T_54982_row2_col1\" class=\"data row2 col1\" >2560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_d261b_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_d261b_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d261b_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row0_col0\" class=\"data row0 col0\" >Imaging</td>\n",
       "      <td id=\"T_d261b_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row1_col0\" class=\"data row1 col0\" >Processed Mass Spectra</td>\n",
       "      <td id=\"T_d261b_row1_col1\" class=\"data row1 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row2_col0\" class=\"data row2 col0\" >Simple Nucleotide Variation</td>\n",
       "      <td id=\"T_d261b_row2_col1\" class=\"data row2 col1\" >11745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row3_col0\" class=\"data row3 col0\" >Copy Number Variation</td>\n",
       "      <td id=\"T_d261b_row3_col1\" class=\"data row3 col1\" >4079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row4_col0\" class=\"data row4 col0\" >Sequencing Reads</td>\n",
       "      <td id=\"T_d261b_row4_col1\" class=\"data row4 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row5_col0\" class=\"data row5 col0\" >Clinical</td>\n",
       "      <td id=\"T_d261b_row5_col1\" class=\"data row5 col1\" >774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row6_col0\" class=\"data row6 col0\" >Somatic Structural Variation</td>\n",
       "      <td id=\"T_d261b_row6_col1\" class=\"data row6 col1\" >370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row7_col0\" class=\"data row7 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_d261b_row7_col1\" class=\"data row7 col1\" >2866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row8_col0\" class=\"data row8 col0\" >Structural Variation</td>\n",
       "      <td id=\"T_d261b_row8_col1\" class=\"data row8 col1\" >2192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row9_col0\" class=\"data row9 col0\" >Raw Mass Spectra</td>\n",
       "      <td id=\"T_d261b_row9_col1\" class=\"data row9 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row10_col0\" class=\"data row10 col0\" >Peptide Spectral Matches</td>\n",
       "      <td id=\"T_d261b_row10_col1\" class=\"data row10 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row11_col0\" class=\"data row11 col0\" >Transcriptome Profiling</td>\n",
       "      <td id=\"T_d261b_row11_col1\" class=\"data row11 col1\" >2820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row12_col0\" class=\"data row12 col0\" >DNA Methylation</td>\n",
       "      <td id=\"T_d261b_row12_col1\" class=\"data row12 col1\" >1947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_d261b_row13_col0\" class=\"data row13 col0\" >Proteome Profiling</td>\n",
       "      <td id=\"T_d261b_row13_col1\" class=\"data row13 col1\" >339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_042a1_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_042a1_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_042a1_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row0_col0\" class=\"data row0 col0\" >DICOM</td>\n",
       "      <td id=\"T_042a1_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row1_col0\" class=\"data row1 col0\" >BAM</td>\n",
       "      <td id=\"T_042a1_row1_col1\" class=\"data row1 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row2_col0\" class=\"data row2 col0\" >VCF</td>\n",
       "      <td id=\"T_042a1_row2_col1\" class=\"data row2 col1\" >6652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row3_col0\" class=\"data row3 col0\" >SVS</td>\n",
       "      <td id=\"T_042a1_row3_col1\" class=\"data row3 col1\" >1111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row4_col0\" class=\"data row4 col0\" >tsv</td>\n",
       "      <td id=\"T_042a1_row4_col1\" class=\"data row4 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row5_col0\" class=\"data row5 col0\" >TSV</td>\n",
       "      <td id=\"T_042a1_row5_col1\" class=\"data row5 col1\" >3980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row6_col0\" class=\"data row6 col0\" >BEDPE</td>\n",
       "      <td id=\"T_042a1_row6_col1\" class=\"data row6 col1\" >1504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row7_col0\" class=\"data row7 col0\" >BCR SSF XML</td>\n",
       "      <td id=\"T_042a1_row7_col1\" class=\"data row7 col1\" >517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row8_col0\" class=\"data row8 col0\" >MAF</td>\n",
       "      <td id=\"T_042a1_row8_col1\" class=\"data row8 col1\" >5235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row9_col0\" class=\"data row9 col0\" >vendor-specific</td>\n",
       "      <td id=\"T_042a1_row9_col1\" class=\"data row9 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row10_col0\" class=\"data row10 col0\" >TXT</td>\n",
       "      <td id=\"T_042a1_row10_col1\" class=\"data row10 col1\" >4827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row11_col0\" class=\"data row11 col0\" >IDAT</td>\n",
       "      <td id=\"T_042a1_row11_col1\" class=\"data row11 col1\" >1298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row12_col0\" class=\"data row12 col0\" >BCR XML</td>\n",
       "      <td id=\"T_042a1_row12_col1\" class=\"data row12 col1\" >1217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row13_col0\" class=\"data row13 col0\" >BCR Auxiliary XML</td>\n",
       "      <td id=\"T_042a1_row13_col1\" class=\"data row13 col1\" >474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row14_col0\" class=\"data row14 col0\" >mzIdentML</td>\n",
       "      <td id=\"T_042a1_row14_col1\" class=\"data row14 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row15_col0\" class=\"data row15 col0\" >mzML</td>\n",
       "      <td id=\"T_042a1_row15_col1\" class=\"data row15 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row16_col0\" class=\"data row16 col0\" >BCR PPS XML</td>\n",
       "      <td id=\"T_042a1_row16_col1\" class=\"data row16 col1\" >193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row17_col0\" class=\"data row17 col0\" >BCR OMF XML</td>\n",
       "      <td id=\"T_042a1_row17_col1\" class=\"data row17 col1\" >39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row18_col0\" class=\"data row18 col0\" >BCR Biotab</td>\n",
       "      <td id=\"T_042a1_row18_col1\" class=\"data row18 col1\" >76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row19_col0\" class=\"data row19 col0\" >XLSX</td>\n",
       "      <td id=\"T_042a1_row19_col1\" class=\"data row19 col1\" >7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row20_col0\" class=\"data row20 col0\" >PDF</td>\n",
       "      <td id=\"T_042a1_row20_col1\" class=\"data row20 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_042a1_row21_col0\" class=\"data row21 col0\" >CDC JSON</td>\n",
       "      <td id=\"T_042a1_row21_col1\" class=\"data row21 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_7ae54_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_7ae54_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_7ae54_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row0_col0\" class=\"data row0 col0\" >Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_7ae54_row0_col1\" class=\"data row0 col1\" >6043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row1_col0\" class=\"data row1 col0\" >None</td>\n",
       "      <td id=\"T_7ae54_row1_col1\" class=\"data row1 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row2_col0\" class=\"data row2 col0\" >Aggregated Somatic Mutation</td>\n",
       "      <td id=\"T_7ae54_row2_col1\" class=\"data row2 col1\" >732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row3_col0\" class=\"data row3 col0\" >Isoform Expression Quantification</td>\n",
       "      <td id=\"T_7ae54_row3_col1\" class=\"data row3 col1\" >700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row4_col0\" class=\"data row4 col0\" >Slide Image</td>\n",
       "      <td id=\"T_7ae54_row4_col1\" class=\"data row4 col1\" >1111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row5_col0\" class=\"data row5 col0\" >Copy Number Segment</td>\n",
       "      <td id=\"T_7ae54_row5_col1\" class=\"data row5 col1\" >1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row6_col0\" class=\"data row6 col0\" >Proprietary</td>\n",
       "      <td id=\"T_7ae54_row6_col1\" class=\"data row6 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row7_col0\" class=\"data row7 col0\" >Masked Intensities</td>\n",
       "      <td id=\"T_7ae54_row7_col1\" class=\"data row7 col1\" >1298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row8_col0\" class=\"data row8 col0\" >Aligned Reads</td>\n",
       "      <td id=\"T_7ae54_row8_col1\" class=\"data row8 col1\" >4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row9_col0\" class=\"data row9 col0\" >Gene Level Copy Number Scores</td>\n",
       "      <td id=\"T_7ae54_row9_col1\" class=\"data row9 col1\" >809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row10_col0\" class=\"data row10 col0\" >miRNA Expression Quantification</td>\n",
       "      <td id=\"T_7ae54_row10_col1\" class=\"data row10 col1\" >700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row11_col0\" class=\"data row11 col0\" >Biospecimen Supplement</td>\n",
       "      <td id=\"T_7ae54_row11_col1\" class=\"data row11 col1\" >1755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row12_col0\" class=\"data row12 col0\" >Transcript Fusion</td>\n",
       "      <td id=\"T_7ae54_row12_col1\" class=\"data row12 col1\" >2278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row13_col0\" class=\"data row13 col0\" >Protein Expression Quantification</td>\n",
       "      <td id=\"T_7ae54_row13_col1\" class=\"data row13 col1\" >339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row14_col0\" class=\"data row14 col0\" >Masked Copy Number Segment</td>\n",
       "      <td id=\"T_7ae54_row14_col1\" class=\"data row14 col1\" >998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row15_col0\" class=\"data row15 col0\" >Raw Simple Somatic Mutation</td>\n",
       "      <td id=\"T_7ae54_row15_col1\" class=\"data row15 col1\" >3397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row16_col0\" class=\"data row16 col0\" >Splice Junction Quantification</td>\n",
       "      <td id=\"T_7ae54_row16_col1\" class=\"data row16 col1\" >710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row17_col0\" class=\"data row17 col0\" >Open Standard</td>\n",
       "      <td id=\"T_7ae54_row17_col1\" class=\"data row17 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row18_col0\" class=\"data row18 col0\" >Gene Expression Quantification</td>\n",
       "      <td id=\"T_7ae54_row18_col1\" class=\"data row18 col1\" >710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row19_col0\" class=\"data row19 col0\" >Structural Rearrangement</td>\n",
       "      <td id=\"T_7ae54_row19_col1\" class=\"data row19 col1\" >284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row20_col0\" class=\"data row20 col0\" >Masked Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_7ae54_row20_col1\" class=\"data row20 col1\" >1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row21_col0\" class=\"data row21 col0\" >Gene Level Copy Number</td>\n",
       "      <td id=\"T_7ae54_row21_col1\" class=\"data row21 col1\" >637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row22_col0\" class=\"data row22 col0\" >Methylation Beta Value</td>\n",
       "      <td id=\"T_7ae54_row22_col1\" class=\"data row22 col1\" >649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row23_col0\" class=\"data row23 col0\" >Masked Somatic Mutation</td>\n",
       "      <td id=\"T_7ae54_row23_col1\" class=\"data row23 col1\" >549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row24_col0\" class=\"data row24 col0\" >Clinical Supplement</td>\n",
       "      <td id=\"T_7ae54_row24_col1\" class=\"data row24 col1\" >773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row25_col0\" class=\"data row25 col0\" >Allele-specific Copy Number Segment</td>\n",
       "      <td id=\"T_7ae54_row25_col1\" class=\"data row25 col1\" >495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row26_col0\" class=\"data row26 col0\" >Text</td>\n",
       "      <td id=\"T_7ae54_row26_col1\" class=\"data row26 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7ae54_row27_col0\" class=\"data row27 col0\" >Pathology Report</td>\n",
       "      <td id=\"T_7ae54_row27_col1\" class=\"data row27 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111f47c0",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Julia decides that a good place to start would be with Slide Images. There's only 1111, so she should be able to quickly scan through them over the next few days and see if they will be useful. So she adds one more filter on her search:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d4a170b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3618 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1111</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m1111\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_6d873_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_6d873_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_6d873_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_6d873_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_6d873_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_245e0_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_245e0_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_245e0_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_245e0_row0_col0\" class=\"data row0 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_245e0_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_98f61_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_98f61_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_98f61_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_98f61_row0_col0\" class=\"data row0 col0\" >SVS</td>\n",
       "      <td id=\"T_98f61_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_15fd6_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_15fd6_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_15fd6_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_15fd6_row0_col0\" class=\"data row0 col0\" >Slide Image</td>\n",
       "      <td id=\"T_15fd6_row0_col1\" class=\"data row0 col1\" >1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JustSlides = Q('file.data_type = \"Slide Image\"')\n",
    "NoadenoJustSlides = NoAdenoData.AND(JustSlides)\n",
    "NoadenoJustSlides.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e385faf",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Finally, Julia uses the pagenation function again to get all the slide files, and merges her metadata file with this file information. This way she will be able to review what phenotypes each slide is associated with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f45f6503",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3522 ms\n"
     ]
    }
   ],
   "source": [
    "slides = NoadenoJustSlides.researchsubject.file.run()\n",
    "slidesdf = pd.DataFrame()\n",
    "for i in slides.paginator(to_df=True):\n",
    "    slidesdf = pd.concat([slidesdf, i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0d1f7a21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>...</th>\n",
       "      <th>identifier_y</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>040229b3-224c-4107-bd33-0854196b6423</td>\n",
       "      <td>[{'system': 'GDC', 'value': '040229b3-224c-410...</td>\n",
       "      <td>TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...</td>\n",
       "      <td>54328093</td>\n",
       "      <td>d4ecae7c6f8f467afbcd060058ffca84</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>040229b3-224c-4107-bd33-0854196b6423</td>\n",
       "      <td>[{'system': 'GDC', 'value': '040229b3-224c-410...</td>\n",
       "      <td>TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...</td>\n",
       "      <td>54328093</td>\n",
       "      <td>d4ecae7c6f8f467afbcd060058ffca84</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a9e316b2-abcf-4e40-870d-3e1d74abf8e4</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...</td>\n",
       "      <td>TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...</td>\n",
       "      <td>192640689</td>\n",
       "      <td>c9526a0e3df583efda8f0dc61bb21d04</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>a9e316b2-abcf-4e40-870d-3e1d74abf8e4</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...</td>\n",
       "      <td>TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...</td>\n",
       "      <td>192640689</td>\n",
       "      <td>c9526a0e3df583efda8f0dc61bb21d04</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1595ab35-ad76-4f27-8b2e-8de0482bf164</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1595ab35-ad76-4f2...</td>\n",
       "      <td>TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...</td>\n",
       "      <td>50173067</td>\n",
       "      <td>79e82fcf3107181102b50ca3e2fde9f5</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23398.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>3349.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145</th>\n",
       "      <td>317ec307-62fd-45d0-b1bd-03aeb094a67e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '317ec307-62fd-45d...</td>\n",
       "      <td>TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...</td>\n",
       "      <td>361909653</td>\n",
       "      <td>df2c6b4dd80890e4f4bc84dd9c1b44fc</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24234.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>916.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2146</th>\n",
       "      <td>0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...</td>\n",
       "      <td>TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...</td>\n",
       "      <td>126509613</td>\n",
       "      <td>3b8d3699fad4ea15f5efc8e60e4977e2</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24234.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>916.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...</td>\n",
       "      <td>TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...</td>\n",
       "      <td>126509613</td>\n",
       "      <td>3b8d3699fad4ea15f5efc8e60e4977e2</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24234.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>916.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2148</th>\n",
       "      <td>358728ce-157b-4002-8377-5391781a3d57</td>\n",
       "      <td>[{'system': 'GDC', 'value': '358728ce-157b-400...</td>\n",
       "      <td>TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...</td>\n",
       "      <td>116519647</td>\n",
       "      <td>592cb544b9ee7331d52d0ab68505d956</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17752.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2149</th>\n",
       "      <td>358728ce-157b-4002-8377-5391781a3d57</td>\n",
       "      <td>[{'system': 'GDC', 'value': '358728ce-157b-400...</td>\n",
       "      <td>TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...</td>\n",
       "      <td>116519647</td>\n",
       "      <td>592cb544b9ee7331d52d0ab68505d956</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17752.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2150 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        id  \\\n",
       "0     040229b3-224c-4107-bd33-0854196b6423   \n",
       "1     040229b3-224c-4107-bd33-0854196b6423   \n",
       "2     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   \n",
       "3     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   \n",
       "4     1595ab35-ad76-4f27-8b2e-8de0482bf164   \n",
       "...                                    ...   \n",
       "2145  317ec307-62fd-45d0-b1bd-03aeb094a67e   \n",
       "2146  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   \n",
       "2147  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   \n",
       "2148  358728ce-157b-4002-8377-5391781a3d57   \n",
       "2149  358728ce-157b-4002-8377-5391781a3d57   \n",
       "\n",
       "                                             identifier  \\\n",
       "0     [{'system': 'GDC', 'value': '040229b3-224c-410...   \n",
       "1     [{'system': 'GDC', 'value': '040229b3-224c-410...   \n",
       "2     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   \n",
       "3     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   \n",
       "4     [{'system': 'GDC', 'value': '1595ab35-ad76-4f2...   \n",
       "...                                                 ...   \n",
       "2145  [{'system': 'GDC', 'value': '317ec307-62fd-45d...   \n",
       "2146  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   \n",
       "2147  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   \n",
       "2148  [{'system': 'GDC', 'value': '358728ce-157b-400...   \n",
       "2149  [{'system': 'GDC', 'value': '358728ce-157b-400...   \n",
       "\n",
       "                                                  label data_category  \\\n",
       "0     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   \n",
       "1     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   \n",
       "2     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   \n",
       "3     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   \n",
       "4     TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...   Biospecimen   \n",
       "...                                                 ...           ...   \n",
       "2145  TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...   Biospecimen   \n",
       "2146  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   \n",
       "2147  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   \n",
       "2148  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   \n",
       "2149  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   \n",
       "\n",
       "        data_type file_format associated_project  \\\n",
       "0     Slide Image         SVS          TCGA-CESC   \n",
       "1     Slide Image         SVS          TCGA-CESC   \n",
       "2     Slide Image         SVS          TCGA-CESC   \n",
       "3     Slide Image         SVS          TCGA-CESC   \n",
       "4     Slide Image         SVS          TCGA-UCEC   \n",
       "...           ...         ...                ...   \n",
       "2145  Slide Image         SVS          TCGA-UCEC   \n",
       "2146  Slide Image         SVS          TCGA-UCEC   \n",
       "2147  Slide Image         SVS          TCGA-UCEC   \n",
       "2148  Slide Image         SVS          TCGA-CESC   \n",
       "2149  Slide Image         SVS          TCGA-CESC   \n",
       "\n",
       "                                                drs_uri  byte_size  \\\n",
       "0     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...   54328093   \n",
       "1     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...   54328093   \n",
       "2     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...  192640689   \n",
       "3     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...  192640689   \n",
       "4     drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...   50173067   \n",
       "...                                                 ...        ...   \n",
       "2145  drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...  361909653   \n",
       "2146  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...  126509613   \n",
       "2147  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...  126509613   \n",
       "2148  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...  116519647   \n",
       "2149  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...  116519647   \n",
       "\n",
       "                              checksum  ...  \\\n",
       "0     d4ecae7c6f8f467afbcd060058ffca84  ...   \n",
       "1     d4ecae7c6f8f467afbcd060058ffca84  ...   \n",
       "2     c9526a0e3df583efda8f0dc61bb21d04  ...   \n",
       "3     c9526a0e3df583efda8f0dc61bb21d04  ...   \n",
       "4     79e82fcf3107181102b50ca3e2fde9f5  ...   \n",
       "...                                ...  ...   \n",
       "2145  df2c6b4dd80890e4f4bc84dd9c1b44fc  ...   \n",
       "2146  3b8d3699fad4ea15f5efc8e60e4977e2  ...   \n",
       "2147  3b8d3699fad4ea15f5efc8e60e4977e2  ...   \n",
       "2148  592cb544b9ee7331d52d0ab68505d956  ...   \n",
       "2149  592cb544b9ee7331d52d0ab68505d956  ...   \n",
       "\n",
       "                                           identifier_y       species     sex  \\\n",
       "0     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "1     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "3     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "4     [{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "2145  [{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...  homo sapiens  female   \n",
       "2146  [{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...  homo sapiens  female   \n",
       "2147  [{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...  homo sapiens  female   \n",
       "2148  [{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...  homo sapiens  female   \n",
       "2149  [{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...  homo sapiens  female   \n",
       "\n",
       "       race               ethnicity days_to_birth subject_associated_project  \\\n",
       "0     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "1     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "2     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "3     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "4     white  not hispanic or latino      -23398.0     [TCGA-UCEC, tcga_ucec]   \n",
       "...     ...                     ...           ...                        ...   \n",
       "2145  white  not hispanic or latino      -24234.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2146  white  not hispanic or latino      -24234.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2147  white  not hispanic or latino      -24234.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2148  white  not hispanic or latino      -17752.0     [TCGA-CESC, tcga_cesc]   \n",
       "2149  white  not hispanic or latino      -17752.0     [TCGA-CESC, tcga_cesc]   \n",
       "\n",
       "     vital_status days_to_death cause_of_death  \n",
       "0           Alive           NaN           None  \n",
       "1           Alive           NaN           None  \n",
       "2           Alive           NaN           None  \n",
       "3           Alive           NaN           None  \n",
       "4            Dead        3349.0           None  \n",
       "...           ...           ...            ...  \n",
       "2145         Dead         916.0           None  \n",
       "2146         Dead         916.0           None  \n",
       "2147         Dead         916.0           None  \n",
       "2148        Alive           NaN           None  \n",
       "2149        Alive           NaN           None  \n",
       "\n",
       "[2150 rows x 32 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata = pd.merge(slidesdf, \n",
    "                         allmetadata, \n",
    "                         on=\"subject_id\")\n",
    "slidemetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abb0f32",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "This merge seems to have created a lot of duplicate lines. So Julia wants to drop the duplicated lines before saving the dataframe to a csv. Since the identifier columns of the data are lists, python will have a hard time telling if they are duplicated. A quick fix is to subset the data to only the columns that should be used for the deduplication, so Julia first gets the list of all the column names, and then pastes everything but \"identifier\", \"identifier_x\", \"identifier_y\" and \"subject_associated_project\"into her drop_duplicates call:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0284925b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'identifier', 'label', 'data_category', 'data_type',\n",
       "       'file_format', 'associated_project', 'drs_uri', 'byte_size', 'checksum',\n",
       "       'data_modality', 'imaging_modality', 'dbgap_accession_number',\n",
       "       'imaging_series', 'subject_id', 'researchsubject_id', 'id_x',\n",
       "       'identifier_x', 'member_of_research_project',\n",
       "       'primary_diagnosis_condition', 'primary_diagnosis_site', 'id_y',\n",
       "       'identifier_y', 'species', 'sex', 'race', 'ethnicity', 'days_to_birth',\n",
       "       'subject_associated_project', 'vital_status', 'days_to_death',\n",
       "       'cause_of_death'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "dd3ea96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "slidemetadata.drop_duplicates(inplace=True, \n",
    "                              subset=['id', 'label', 'data_category', 'data_type',\n",
    "       'file_format', 'associated_project', 'drs_uri', 'byte_size', 'checksum',\n",
    "       'data_modality', 'imaging_modality', 'dbgap_accession_number',\n",
    "       'imaging_series', 'subject_id', 'researchsubject_id', 'id_x',\n",
    "        'member_of_research_project',\n",
    "       'primary_diagnosis_condition', 'primary_diagnosis_site', 'id_y',\n",
    "        'species', 'sex', 'race', 'ethnicity', 'days_to_birth',\n",
    "         'vital_status', 'days_to_death',\n",
    "       'cause_of_death'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b8a74f25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>...</th>\n",
       "      <th>identifier_y</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>040229b3-224c-4107-bd33-0854196b6423</td>\n",
       "      <td>[{'system': 'GDC', 'value': '040229b3-224c-410...</td>\n",
       "      <td>TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...</td>\n",
       "      <td>54328093</td>\n",
       "      <td>d4ecae7c6f8f467afbcd060058ffca84</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a9e316b2-abcf-4e40-870d-3e1d74abf8e4</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...</td>\n",
       "      <td>TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...</td>\n",
       "      <td>192640689</td>\n",
       "      <td>c9526a0e3df583efda8f0dc61bb21d04</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14129.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1595ab35-ad76-4f27-8b2e-8de0482bf164</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1595ab35-ad76-4f2...</td>\n",
       "      <td>TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...</td>\n",
       "      <td>50173067</td>\n",
       "      <td>79e82fcf3107181102b50ca3e2fde9f5</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23398.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>3349.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>e2086d9b-63ba-4e03-966e-a4427b6ade74</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e2086d9b-63ba-4e0...</td>\n",
       "      <td>TCGA-DI-A2QY-01Z-00-DX1.EBBED5CC-3098-4694-BCD...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:e2086d9b-63ba-4e03-966e-a4427b6a...</td>\n",
       "      <td>1283995391</td>\n",
       "      <td>165a9199090fe0ff9ea86fd2168d41f2</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23398.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>3349.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dd8d427d-d63c-41d4-9da7-dab428a3dc3b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'dd8d427d-d63c-41d...</td>\n",
       "      <td>TCGA-DI-A2QY-01A-01-TS1.62316960-36E9-4919-BE3...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:dd8d427d-d63c-41d4-9da7-dab428a3...</td>\n",
       "      <td>255850447</td>\n",
       "      <td>b96eceebd9116cf9567830ce36cd157a</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23398.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>3349.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2140</th>\n",
       "      <td>9bc76014-1b92-4914-8bf1-482b85bae0dd</td>\n",
       "      <td>[{'system': 'GDC', 'value': '9bc76014-1b92-491...</td>\n",
       "      <td>TCGA-AJ-A3BG-01A-01-TS1.FDB1B2A0-9AAD-460C-A87...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:9bc76014-1b92-4914-8bf1-482b85ba...</td>\n",
       "      <td>78523868</td>\n",
       "      <td>6d4195721e8f04d6fbf708fd62f98af3</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AJ-A3BG'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23861.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2142</th>\n",
       "      <td>859f8cfb-116d-44ec-9017-b6f6617d62bf</td>\n",
       "      <td>[{'system': 'GDC', 'value': '859f8cfb-116d-44e...</td>\n",
       "      <td>TCGA-AJ-A3BG-01Z-00-DX1.87799C6D-8229-4DE2-BD6...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:859f8cfb-116d-44ec-9017-b6f6617d...</td>\n",
       "      <td>185346200</td>\n",
       "      <td>67c97665154649a31354b8880aebd713</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AJ-A3BG'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23861.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2144</th>\n",
       "      <td>317ec307-62fd-45d0-b1bd-03aeb094a67e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '317ec307-62fd-45d...</td>\n",
       "      <td>TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...</td>\n",
       "      <td>361909653</td>\n",
       "      <td>df2c6b4dd80890e4f4bc84dd9c1b44fc</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24234.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>916.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2146</th>\n",
       "      <td>0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...</td>\n",
       "      <td>TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...</td>\n",
       "      <td>126509613</td>\n",
       "      <td>3b8d3699fad4ea15f5efc8e60e4977e2</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24234.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>916.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2148</th>\n",
       "      <td>358728ce-157b-4002-8377-5391781a3d57</td>\n",
       "      <td>[{'system': 'GDC', 'value': '358728ce-157b-400...</td>\n",
       "      <td>TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...</td>\n",
       "      <td>116519647</td>\n",
       "      <td>592cb544b9ee7331d52d0ab68505d956</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17752.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1111 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        id  \\\n",
       "0     040229b3-224c-4107-bd33-0854196b6423   \n",
       "2     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   \n",
       "4     1595ab35-ad76-4f27-8b2e-8de0482bf164   \n",
       "6     e2086d9b-63ba-4e03-966e-a4427b6ade74   \n",
       "8     dd8d427d-d63c-41d4-9da7-dab428a3dc3b   \n",
       "...                                    ...   \n",
       "2140  9bc76014-1b92-4914-8bf1-482b85bae0dd   \n",
       "2142  859f8cfb-116d-44ec-9017-b6f6617d62bf   \n",
       "2144  317ec307-62fd-45d0-b1bd-03aeb094a67e   \n",
       "2146  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   \n",
       "2148  358728ce-157b-4002-8377-5391781a3d57   \n",
       "\n",
       "                                             identifier  \\\n",
       "0     [{'system': 'GDC', 'value': '040229b3-224c-410...   \n",
       "2     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   \n",
       "4     [{'system': 'GDC', 'value': '1595ab35-ad76-4f2...   \n",
       "6     [{'system': 'GDC', 'value': 'e2086d9b-63ba-4e0...   \n",
       "8     [{'system': 'GDC', 'value': 'dd8d427d-d63c-41d...   \n",
       "...                                                 ...   \n",
       "2140  [{'system': 'GDC', 'value': '9bc76014-1b92-491...   \n",
       "2142  [{'system': 'GDC', 'value': '859f8cfb-116d-44e...   \n",
       "2144  [{'system': 'GDC', 'value': '317ec307-62fd-45d...   \n",
       "2146  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   \n",
       "2148  [{'system': 'GDC', 'value': '358728ce-157b-400...   \n",
       "\n",
       "                                                  label data_category  \\\n",
       "0     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   \n",
       "2     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   \n",
       "4     TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...   Biospecimen   \n",
       "6     TCGA-DI-A2QY-01Z-00-DX1.EBBED5CC-3098-4694-BCD...   Biospecimen   \n",
       "8     TCGA-DI-A2QY-01A-01-TS1.62316960-36E9-4919-BE3...   Biospecimen   \n",
       "...                                                 ...           ...   \n",
       "2140  TCGA-AJ-A3BG-01A-01-TS1.FDB1B2A0-9AAD-460C-A87...   Biospecimen   \n",
       "2142  TCGA-AJ-A3BG-01Z-00-DX1.87799C6D-8229-4DE2-BD6...   Biospecimen   \n",
       "2144  TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...   Biospecimen   \n",
       "2146  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   \n",
       "2148  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   \n",
       "\n",
       "        data_type file_format associated_project  \\\n",
       "0     Slide Image         SVS          TCGA-CESC   \n",
       "2     Slide Image         SVS          TCGA-CESC   \n",
       "4     Slide Image         SVS          TCGA-UCEC   \n",
       "6     Slide Image         SVS          TCGA-UCEC   \n",
       "8     Slide Image         SVS          TCGA-UCEC   \n",
       "...           ...         ...                ...   \n",
       "2140  Slide Image         SVS          TCGA-UCEC   \n",
       "2142  Slide Image         SVS          TCGA-UCEC   \n",
       "2144  Slide Image         SVS          TCGA-UCEC   \n",
       "2146  Slide Image         SVS          TCGA-UCEC   \n",
       "2148  Slide Image         SVS          TCGA-CESC   \n",
       "\n",
       "                                                drs_uri   byte_size  \\\n",
       "0     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...    54328093   \n",
       "2     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...   192640689   \n",
       "4     drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...    50173067   \n",
       "6     drs://dg.4DFC:e2086d9b-63ba-4e03-966e-a4427b6a...  1283995391   \n",
       "8     drs://dg.4DFC:dd8d427d-d63c-41d4-9da7-dab428a3...   255850447   \n",
       "...                                                 ...         ...   \n",
       "2140  drs://dg.4DFC:9bc76014-1b92-4914-8bf1-482b85ba...    78523868   \n",
       "2142  drs://dg.4DFC:859f8cfb-116d-44ec-9017-b6f6617d...   185346200   \n",
       "2144  drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...   361909653   \n",
       "2146  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...   126509613   \n",
       "2148  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...   116519647   \n",
       "\n",
       "                              checksum  ...  \\\n",
       "0     d4ecae7c6f8f467afbcd060058ffca84  ...   \n",
       "2     c9526a0e3df583efda8f0dc61bb21d04  ...   \n",
       "4     79e82fcf3107181102b50ca3e2fde9f5  ...   \n",
       "6     165a9199090fe0ff9ea86fd2168d41f2  ...   \n",
       "8     b96eceebd9116cf9567830ce36cd157a  ...   \n",
       "...                                ...  ...   \n",
       "2140  6d4195721e8f04d6fbf708fd62f98af3  ...   \n",
       "2142  67c97665154649a31354b8880aebd713  ...   \n",
       "2144  df2c6b4dd80890e4f4bc84dd9c1b44fc  ...   \n",
       "2146  3b8d3699fad4ea15f5efc8e60e4977e2  ...   \n",
       "2148  592cb544b9ee7331d52d0ab68505d956  ...   \n",
       "\n",
       "                                           identifier_y       species     sex  \\\n",
       "0     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'TCGA-VS-A8EI'}, {...  homo sapiens  female   \n",
       "4     [{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...  homo sapiens  female   \n",
       "6     [{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...  homo sapiens  female   \n",
       "8     [{'system': 'GDC', 'value': 'TCGA-DI-A2QY'}, {...  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "2140  [{'system': 'GDC', 'value': 'TCGA-AJ-A3BG'}, {...  homo sapiens  female   \n",
       "2142  [{'system': 'GDC', 'value': 'TCGA-AJ-A3BG'}, {...  homo sapiens  female   \n",
       "2144  [{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...  homo sapiens  female   \n",
       "2146  [{'system': 'GDC', 'value': 'TCGA-AX-A2H4'}, {...  homo sapiens  female   \n",
       "2148  [{'system': 'GDC', 'value': 'TCGA-EK-A2RB'}, {...  homo sapiens  female   \n",
       "\n",
       "       race               ethnicity days_to_birth subject_associated_project  \\\n",
       "0     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "2     white            not reported      -14129.0     [TCGA-CESC, tcga_cesc]   \n",
       "4     white  not hispanic or latino      -23398.0     [TCGA-UCEC, tcga_ucec]   \n",
       "6     white  not hispanic or latino      -23398.0     [TCGA-UCEC, tcga_ucec]   \n",
       "8     white  not hispanic or latino      -23398.0     [TCGA-UCEC, tcga_ucec]   \n",
       "...     ...                     ...           ...                        ...   \n",
       "2140  white  not hispanic or latino      -23861.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2142  white  not hispanic or latino      -23861.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2144  white  not hispanic or latino      -24234.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2146  white  not hispanic or latino      -24234.0     [TCGA-UCEC, tcga_ucec]   \n",
       "2148  white  not hispanic or latino      -17752.0     [TCGA-CESC, tcga_cesc]   \n",
       "\n",
       "     vital_status days_to_death cause_of_death  \n",
       "0           Alive           NaN           None  \n",
       "2           Alive           NaN           None  \n",
       "4            Dead        3349.0           None  \n",
       "6            Dead        3349.0           None  \n",
       "8            Dead        3349.0           None  \n",
       "...           ...           ...            ...  \n",
       "2140        Alive           NaN           None  \n",
       "2142        Alive           NaN           None  \n",
       "2144         Dead         916.0           None  \n",
       "2146         Dead         916.0           None  \n",
       "2148        Alive           NaN           None  \n",
       "\n",
       "[1111 rows x 32 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "369a3068",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "That looks better! Julia saves this dataframe to a csv, and now she has all the information she needs to begin work on her project. She can use the `drs_id` column information to directly download the images she is interested in using a DRS resolver, or she can input the DRS IDs at a cloud workspace such as [Terra](https://terra.bio/) or the [Cancer Genomics Cloud](https://www.cancergenomicscloud.org/) to view the images online. In either case, she has all the metadata she needs to get started, and can save this notebook of her work in case she'd like to come back and modify her search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8ef6e0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "slidemetadata.to_csv(\"slidemetadata.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
