{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c341e94",
   "metadata": {},
   "source": [
    "# Endpoint Chaining\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8387f59",
   "metadata": {},
   "source": [
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "                                                                    \n",
    "Before we do any work, we needs to import these functions cdapython.\n",
    "We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28a10e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "print(Q.get_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da4c0179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()\n",
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8084b3",
   "metadata": {},
   "source": [
    "## Endpoint Chaining\n",
    "\n",
    "We're going to build on our <a href=\"../BasicSearch\">previous basic search</a> to see what information exists about cancers that were first diagnosed in the brain. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d804f764",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "736ad485",
   "metadata": {},
   "source": [
    "Previously we looked at subject, research_subject, specimen and file results seperately, but we can also combine these. \n",
    "\n",
    "Let's say what we're really interested in is finding analysis done on specimens, so we're looking for files that belong to specimens that match our search. If we search the files endpoint directly, we'll get back all files that meet our search criteria, regardless of whether they are specimens specficially. Instead, we can chain our query to the specimen endpoint and then to the files endpoint and get the combined result:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7284df72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3938 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 0a05eb3d-948d-4519-b30e-d94a59a5ed0a\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 405271\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles =  myquery.specimen.file.run()\n",
    "myqueryspecimenfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d9154b",
   "metadata": {},
   "source": [
    "We get back 405031 files that belong to specimens that meet our search criteria. As before, we can preview the results by using the `.to_dataframe()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "410b0e2b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00b5cc1e-e51a-4c5f-a98c-076acf265042</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00b5cc1e-e51a-4c5...</td>\n",
       "      <td>4db50bc3-572f-48bf-a88a-2fd065496409.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:00b5cc1e-e51a-4c5f-a98c-076acf26...</td>\n",
       "      <td>1054</td>\n",
       "      <td>01b29f819d9447e65a1d4b8d80a93b4a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0019286</td>\n",
       "      <td>b1569de4-a705-47f0-842d-4369a3bf43a3</td>\n",
       "      <td>bfae3035-fc8b-5f5e-b2b4-37e817c4f827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>06668371-2079-4471-a06a-7ad6f987a96c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '06668371-2079-447...</td>\n",
       "      <td>5861e7d6-2066-40d7-8868-c14a69255304.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-NKI</td>\n",
       "      <td>drs://dg.4DFC:06668371-2079-4471-a06a-7ad6f987...</td>\n",
       "      <td>1052</td>\n",
       "      <td>e8d1a8534b9fe7a813f1b10a8a706daf</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-NKI-8IN3</td>\n",
       "      <td>54ada9c7-e8f7-4ecc-b829-903be7e3ad82</td>\n",
       "      <td>08b49bbe-09f0-447b-a193-f2c5a923c051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1958cd4c-263d-4a5e-ad69-8bae23a0edc7</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1958cd4c-263d-4a5...</td>\n",
       "      <td>TCGA-HT-7477-01B-21-A29Z-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:1958cd4c-263d-4a5e-ad69-8bae23a0...</td>\n",
       "      <td>22089</td>\n",
       "      <td>ba575069e7c576eb3bf532fd16c164ac</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7477</td>\n",
       "      <td>38f4b9f3-6709-40ff-9a81-5fa2111b1ce6</td>\n",
       "      <td>83cf2e54-8468-4e5c-b427-4134e046009c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19f0c787-654a-4cf6-ba8b-dca9e46bc220</td>\n",
       "      <td>[{'system': 'GDC', 'value': '19f0c787-654a-4cf...</td>\n",
       "      <td>TCGA-QH-A65V-01A-21-A44D-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:19f0c787-654a-4cf6-ba8b-dca9e46b...</td>\n",
       "      <td>21985</td>\n",
       "      <td>92f1aae376db4e1f4e5567cf33069fd3</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-QH-A65V</td>\n",
       "      <td>ad72e61e-51e3-4992-ac77-65e895935736</td>\n",
       "      <td>c76aa96d-53f4-4ec0-a009-a1fbaed96e39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22dbc627-b923-453e-8fe0-a6fbe76c66bc</td>\n",
       "      <td>[{'system': 'GDC', 'value': '22dbc627-b923-453...</td>\n",
       "      <td>18235707-e320-40e6-8628-c17e8e004702.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:22dbc627-b923-453e-8fe0-a6fbe76c...</td>\n",
       "      <td>22932</td>\n",
       "      <td>35562af5c759a3ac265c3db414f4529b</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0003834</td>\n",
       "      <td>d500a0f9-02f9-4eee-b469-3f1cd361d563</td>\n",
       "      <td>d92a9d5d-5391-4913-89ab-09eda609675c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>01f40d93-e2fe-46d9-b0db-5bb3be21d158</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01f40d93-e2fe-46d...</td>\n",
       "      <td>dd13f0a2-db35-4cec-9fa8-12b4830c9403.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-GRCC</td>\n",
       "      <td>drs://dg.4DFC:01f40d93-e2fe-46d9-b0db-5bb3be21...</td>\n",
       "      <td>2753</td>\n",
       "      <td>758d9563a8348e7ba1479de7b5592977</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-GRCC-774ad24a</td>\n",
       "      <td>ebee5ebd-e037-455b-9071-1b0e6f033764</td>\n",
       "      <td>694c18df-47cd-4cf9-ac2e-8cedc3d2feb7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>14b828c9-0eac-4613-9354-64dae9e7acf2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '14b828c9-0eac-461...</td>\n",
       "      <td>TCGA-06-6388-01A-23-1900-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>drs://dg.4DFC:14b828c9-0eac-4613-9354-64dae9e7...</td>\n",
       "      <td>23942</td>\n",
       "      <td>e750e4fc571db02b428ec510527dbf01</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-06-6388</td>\n",
       "      <td>7eb1d526-923d-4b31-b8be-27d83b3dd025</td>\n",
       "      <td>673a0dc8-4bed-43df-9fb0-36fb546a5e58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1825d954-56e3-4faa-991e-56e28d23f31e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1825d954-56e3-4fa...</td>\n",
       "      <td>415b62cf-1ab3-4432-9fd4-c326718d8142.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:1825d954-56e3-4faa-991e-56e28d23...</td>\n",
       "      <td>22932</td>\n",
       "      <td>78fb97b022c9790c89888ba62bfb222e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0002086</td>\n",
       "      <td>864d18aa-dced-4faf-9cc7-dfb01092f760</td>\n",
       "      <td>30b60e4f-9e6b-490e-8b9f-e656a0798402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>311e41c3-017f-4f30-a858-abd99547c1d2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '311e41c3-017f-4f3...</td>\n",
       "      <td>TCGA-HT-7857-01A-21-A29Y-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:311e41c3-017f-4f30-a858-abd99547...</td>\n",
       "      <td>22107</td>\n",
       "      <td>d300e8a5dd261c8fa4181c7f790a805e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7857</td>\n",
       "      <td>4e1606e5-f4c1-41a2-9f2c-ad0d02650070</td>\n",
       "      <td>a7357fa3-3468-4288-99c2-624d885d471c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>329ccbdc-8b9a-4cc2-ac28-007db73dc228</td>\n",
       "      <td>[{'system': 'GDC', 'value': '329ccbdc-8b9a-4cc...</td>\n",
       "      <td>TCGA-DU-7309-01A-21-A29X-20_RPPA_data.tsv</td>\n",
       "      <td>Proteome Profiling</td>\n",
       "      <td>Protein Expression Quantification</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:329ccbdc-8b9a-4cc2-ac28-007db73d...</td>\n",
       "      <td>22009</td>\n",
       "      <td>f686ecd5d6e3c5d793a593f7e03da17d</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-7309</td>\n",
       "      <td>ca427acb-034e-42e2-961a-b9d5dc30b741</td>\n",
       "      <td>21d6c002-65e0-4dcc-b5af-5c3adfd0e75c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   00b5cc1e-e51a-4c5f-a98c-076acf265042   \n",
       "1   06668371-2079-4471-a06a-7ad6f987a96c   \n",
       "2   1958cd4c-263d-4a5e-ad69-8bae23a0edc7   \n",
       "3   19f0c787-654a-4cf6-ba8b-dca9e46bc220   \n",
       "4   22dbc627-b923-453e-8fe0-a6fbe76c66bc   \n",
       "..                                   ...   \n",
       "95  01f40d93-e2fe-46d9-b0db-5bb3be21d158   \n",
       "96  14b828c9-0eac-4613-9354-64dae9e7acf2   \n",
       "97  1825d954-56e3-4faa-991e-56e28d23f31e   \n",
       "98  311e41c3-017f-4f30-a858-abd99547c1d2   \n",
       "99  329ccbdc-8b9a-4cc2-ac28-007db73dc228   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '00b5cc1e-e51a-4c5...   \n",
       "1   [{'system': 'GDC', 'value': '06668371-2079-447...   \n",
       "2   [{'system': 'GDC', 'value': '1958cd4c-263d-4a5...   \n",
       "3   [{'system': 'GDC', 'value': '19f0c787-654a-4cf...   \n",
       "4   [{'system': 'GDC', 'value': '22dbc627-b923-453...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '01f40d93-e2fe-46d...   \n",
       "96  [{'system': 'GDC', 'value': '14b828c9-0eac-461...   \n",
       "97  [{'system': 'GDC', 'value': '1825d954-56e3-4fa...   \n",
       "98  [{'system': 'GDC', 'value': '311e41c3-017f-4f3...   \n",
       "99  [{'system': 'GDC', 'value': '329ccbdc-8b9a-4cc...   \n",
       "\n",
       "                                                label  \\\n",
       "0   4db50bc3-572f-48bf-a88a-2fd065496409.genie.ali...   \n",
       "1   5861e7d6-2066-40d7-8868-c14a69255304.genie.ali...   \n",
       "2           TCGA-HT-7477-01B-21-A29Z-20_RPPA_data.tsv   \n",
       "3           TCGA-QH-A65V-01A-21-A44D-20_RPPA_data.tsv   \n",
       "4   18235707-e320-40e6-8628-c17e8e004702.genie.ali...   \n",
       "..                                                ...   \n",
       "95  dd13f0a2-db35-4cec-9fa8-12b4830c9403.genie.ali...   \n",
       "96          TCGA-06-6388-01A-23-1900-20_RPPA_data.tsv   \n",
       "97  415b62cf-1ab3-4432-9fd4-c326718d8142.genie.ali...   \n",
       "98          TCGA-HT-7857-01A-21-A29Y-20_RPPA_data.tsv   \n",
       "99          TCGA-DU-7309-01A-21-A29X-20_RPPA_data.tsv   \n",
       "\n",
       "                  data_category                          data_type  \\\n",
       "0   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "1   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "2            Proteome Profiling  Protein Expression Quantification   \n",
       "3            Proteome Profiling  Protein Expression Quantification   \n",
       "4         Copy Number Variation      Gene Level Copy Number Scores   \n",
       "..                          ...                                ...   \n",
       "95  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "96           Proteome Profiling  Protein Expression Quantification   \n",
       "97        Copy Number Variation      Gene Level Copy Number Scores   \n",
       "98           Proteome Profiling  Protein Expression Quantification   \n",
       "99           Proteome Profiling  Protein Expression Quantification   \n",
       "\n",
       "   file_format associated_project  \\\n",
       "0          MAF          GENIE-MSK   \n",
       "1          MAF          GENIE-NKI   \n",
       "2          TSV           TCGA-LGG   \n",
       "3          TSV           TCGA-LGG   \n",
       "4          TSV          GENIE-MSK   \n",
       "..         ...                ...   \n",
       "95         MAF         GENIE-GRCC   \n",
       "96         TSV           TCGA-GBM   \n",
       "97         TSV          GENIE-MSK   \n",
       "98         TSV           TCGA-LGG   \n",
       "99         TSV           TCGA-LGG   \n",
       "\n",
       "                                              drs_uri  byte_size  \\\n",
       "0   drs://dg.4DFC:00b5cc1e-e51a-4c5f-a98c-076acf26...       1054   \n",
       "1   drs://dg.4DFC:06668371-2079-4471-a06a-7ad6f987...       1052   \n",
       "2   drs://dg.4DFC:1958cd4c-263d-4a5e-ad69-8bae23a0...      22089   \n",
       "3   drs://dg.4DFC:19f0c787-654a-4cf6-ba8b-dca9e46b...      21985   \n",
       "4   drs://dg.4DFC:22dbc627-b923-453e-8fe0-a6fbe76c...      22932   \n",
       "..                                                ...        ...   \n",
       "95  drs://dg.4DFC:01f40d93-e2fe-46d9-b0db-5bb3be21...       2753   \n",
       "96  drs://dg.4DFC:14b828c9-0eac-4613-9354-64dae9e7...      23942   \n",
       "97  drs://dg.4DFC:1825d954-56e3-4faa-991e-56e28d23...      22932   \n",
       "98  drs://dg.4DFC:311e41c3-017f-4f30-a858-abd99547...      22107   \n",
       "99  drs://dg.4DFC:329ccbdc-8b9a-4cc2-ac28-007db73d...      22009   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   01b29f819d9447e65a1d4b8d80a93b4a       Genomic             None   \n",
       "1   e8d1a8534b9fe7a813f1b10a8a706daf       Genomic             None   \n",
       "2   ba575069e7c576eb3bf532fd16c164ac       Genomic             None   \n",
       "3   92f1aae376db4e1f4e5567cf33069fd3       Genomic             None   \n",
       "4   35562af5c759a3ac265c3db414f4529b       Genomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  758d9563a8348e7ba1479de7b5592977       Genomic             None   \n",
       "96  e750e4fc571db02b428ec510527dbf01       Genomic             None   \n",
       "97  78fb97b022c9790c89888ba62bfb222e       Genomic             None   \n",
       "98  d300e8a5dd261c8fa4181c7f790a805e       Genomic             None   \n",
       "99  f686ecd5d6e3c5d793a593f7e03da17d       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number imaging_series           subject_id  \\\n",
       "0                    None           None  GENIE-MSK-P-0019286   \n",
       "1                    None           None       GENIE-NKI-8IN3   \n",
       "2                    None           None         TCGA-HT-7477   \n",
       "3                    None           None         TCGA-QH-A65V   \n",
       "4                    None           None  GENIE-MSK-P-0003834   \n",
       "..                    ...            ...                  ...   \n",
       "95                   None           None  GENIE-GRCC-774ad24a   \n",
       "96                   None           None         TCGA-06-6388   \n",
       "97                   None           None  GENIE-MSK-P-0002086   \n",
       "98                   None           None         TCGA-HT-7857   \n",
       "99                   None           None         TCGA-DU-7309   \n",
       "\n",
       "                      researchsubject_id           researchsubject_specimen_id  \n",
       "0   b1569de4-a705-47f0-842d-4369a3bf43a3  bfae3035-fc8b-5f5e-b2b4-37e817c4f827  \n",
       "1   54ada9c7-e8f7-4ecc-b829-903be7e3ad82  08b49bbe-09f0-447b-a193-f2c5a923c051  \n",
       "2   38f4b9f3-6709-40ff-9a81-5fa2111b1ce6  83cf2e54-8468-4e5c-b427-4134e046009c  \n",
       "3   ad72e61e-51e3-4992-ac77-65e895935736  c76aa96d-53f4-4ec0-a009-a1fbaed96e39  \n",
       "4   d500a0f9-02f9-4eee-b469-3f1cd361d563  d92a9d5d-5391-4913-89ab-09eda609675c  \n",
       "..                                   ...                                   ...  \n",
       "95  ebee5ebd-e037-455b-9071-1b0e6f033764  694c18df-47cd-4cf9-ac2e-8cedc3d2feb7  \n",
       "96  7eb1d526-923d-4b31-b8be-27d83b3dd025  673a0dc8-4bed-43df-9fb0-36fb546a5e58  \n",
       "97  864d18aa-dced-4faf-9cc7-dfb01092f760  30b60e4f-9e6b-490e-8b9f-e656a0798402  \n",
       "98  4e1606e5-f4c1-41a2-9f2c-ad0d02650070  a7357fa3-3468-4288-99c2-624d885d471c  \n",
       "99  ca427acb-034e-42e2-961a-b9d5dc30b741  21d6c002-65e0-4dcc-b5af-5c3adfd0e75c  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fda2e6",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "<h3>Valid Endpoint Chains</h3>\n",
    "\n",
    "Not all endpoints can be chained together. This is a restriction caused by the data itself. `diagnosis` and `treatment` information does not have files directly attached to it, instead these files are associated with the `researchsubject`, as such both \"myquery.treatment.files.run()\" and \"myquery.diagnosis.files.run()\" will fail, as there are no files to retrieve. Valid chains are:\n",
    "    \n",
    "<ul>\n",
    "<li><b>myquery.subject.file.run:</b> This will return all the files that meet the query and that are directly tied to subject</li>\n",
    "<li><b>myquery.researchsubject.file.run:</b>This will return all the files that meet the query and that are directly tied to researchsubject</li>\n",
    "<li><b>myquery.specimen.file.run:</b> This will return all the files that meet the query and that are directly tied to specimen</li>\n",
    "<li><b>myquery.subject.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to subject</li>\n",
    "<li><b>myquery.researchsubject.file.count.run:</b>This will return the count of files that meet the query and that are directly tied to researchsubject</li>\n",
    "<li><b>myquery.specimen.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to specimen</li>\n",
    "</ul>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
