{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8a16a9fe",
   "metadata": {},
   "source": [
    "# Summarize Search Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f90b531",
   "metadata": {},
   "source": [
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "                                                                    \n",
    "Before we do any work, we needs to import these functions cdapython.\n",
    "We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5245b7b0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "print(Q.get_version())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d586efee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()\n",
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d538da92",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "    \n",
    "And two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n",
    "\n",
    "\n",
    "If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specfic format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.\n",
    "\n",
    "In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.\n",
    "\n",
    "In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76026a3",
   "metadata": {},
   "source": [
    "## Getting simple summary data\n",
    "\n",
    "Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`. This is the same query we ran in the <a href=\"../BasicSearch\">Basic Search</a> notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c92a98ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e855573",
   "metadata": {
    "tags": [
     "Help!"
    ]
   },
   "source": [
    "\n",
    "<div class=\"cdawarn\" style=\"background-color:#f9cfbf;color:black;padding:20px;\">\n",
    "<h3>Where did those terms come from?</h3>\n",
    "    \n",
    "If you aren't sure how we knew what terms to put in our search, please refer back to the <a href=\"../SearchTerms\">What search terms are available?</a> notebook. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e492876",
   "metadata": {},
   "source": [
    "### Overall summary\n",
    "\n",
    "You can get a quick summary of how many unique specimens, treatments, diagnoses, researchsubjects and subjects meet your search criteria by chaining a `count` command into the basic `run` call. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c8e3599f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3735 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">specimen_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">39201</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "specimen_count : \u001b[1;36m39201\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">treatment_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2386</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "treatment_count : \u001b[1;36m2386\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">diagnosis_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1756</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "diagnosis_count : \u001b[1;36m1756\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">researchsubject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3716</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "researchsubject_count : \u001b[1;36m3716\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">subject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2384</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "subject_count : \u001b[1;36m2384\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.count.run()\n",
    "\n",
    "# Total execution time: 3735 ms\n",
    "# specimen_count : 39201\n",
    "# treatment_count : 2386\n",
    "# diagnosis_count : 1756\n",
    "# researchsubject_count : 3716\n",
    "# subject_count : 2384"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d483e944",
   "metadata": {},
   "source": [
    "These numbers are how many total rows of data will come back when querying the various endpoints.\n",
    "\n",
    "\n",
    "\n",
    "### subject summary\n",
    "\n",
    "We can also add `count`to the other run calls we did in the <a href=\"../BasicSearch\">Basic Search</a> notebook to get more detailed summaries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d9137aa",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3565 ms\n"
     ]
    }
   ],
   "source": [
    "subjectresults = myquery.subject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dff2da08",
   "metadata": {},
   "source": [
    "Since we save the output as a variable, we need to look at the variable to see the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "993479db",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2384</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m2384\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">  files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4099497</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "  files : \u001b[1;36m4099497\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a07c6_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a07c6_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a07c6_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a07c6_row0_col0\" class=\"data row0 col0\" >IDC</td>\n",
       "      <td id=\"T_a07c6_row0_col1\" class=\"data row0 col1\" >1955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a07c6_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_a07c6_row1_col1\" class=\"data row1 col1\" >309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a07c6_row2_col0\" class=\"data row2 col0\" >GDC</td>\n",
       "      <td id=\"T_a07c6_row2_col1\" class=\"data row2 col1\" >1454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_7e83b_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_7e83b_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_7e83b_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >sex</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_7e83b_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_7e83b_row0_col1\" class=\"data row0 col1\" >748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7e83b_row1_col0\" class=\"data row1 col0\" >female</td>\n",
       "      <td id=\"T_7e83b_row1_col1\" class=\"data row1 col1\" >653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7e83b_row2_col0\" class=\"data row2 col0\" >male</td>\n",
       "      <td id=\"T_7e83b_row2_col1\" class=\"data row2 col1\" >980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_7e83b_row3_col0\" class=\"data row3 col0\" >not reported</td>\n",
       "      <td id=\"T_7e83b_row3_col1\" class=\"data row3 col1\" >3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_0fb20_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_0fb20_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_0fb20_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >race</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_0fb20_row0_col1\" class=\"data row0 col1\" >748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row1_col0\" class=\"data row1 col0\" >not reported</td>\n",
       "      <td id=\"T_0fb20_row1_col1\" class=\"data row1 col1\" >136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row2_col0\" class=\"data row2 col0\" >white</td>\n",
       "      <td id=\"T_0fb20_row2_col1\" class=\"data row2 col1\" >1311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row3_col0\" class=\"data row3 col0\" >not allowed to collect</td>\n",
       "      <td id=\"T_0fb20_row3_col1\" class=\"data row3 col1\" >25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row4_col0\" class=\"data row4 col0\" >asian</td>\n",
       "      <td id=\"T_0fb20_row4_col1\" class=\"data row4 col1\" >33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row5_col0\" class=\"data row5 col0\" >black or african american</td>\n",
       "      <td id=\"T_0fb20_row5_col1\" class=\"data row5 col1\" >96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row6_col0\" class=\"data row6 col0\" >american indian or alaska native</td>\n",
       "      <td id=\"T_0fb20_row6_col1\" class=\"data row6 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row7_col0\" class=\"data row7 col0\" >other</td>\n",
       "      <td id=\"T_0fb20_row7_col1\" class=\"data row7 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row8_col0\" class=\"data row8 col0\" >Unknown</td>\n",
       "      <td id=\"T_0fb20_row8_col1\" class=\"data row8 col1\" >21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0fb20_row9_col0\" class=\"data row9 col0\" >native hawaiian or other pacific islander</td>\n",
       "      <td id=\"T_0fb20_row9_col1\" class=\"data row9 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_f025e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_f025e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f025e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >ethnicity</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_f025e_row0_col1\" class=\"data row0 col1\" >748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row1_col0\" class=\"data row1 col0\" >hispanic or latino</td>\n",
       "      <td id=\"T_f025e_row1_col1\" class=\"data row1 col1\" >85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row2_col0\" class=\"data row2 col0\" >not reported</td>\n",
       "      <td id=\"T_f025e_row2_col1\" class=\"data row2 col1\" >219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row3_col0\" class=\"data row3 col0\" >not hispanic or latino</td>\n",
       "      <td id=\"T_f025e_row3_col1\" class=\"data row3 col1\" >1285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row4_col0\" class=\"data row4 col0\" >Unknown</td>\n",
       "      <td id=\"T_f025e_row4_col1\" class=\"data row4 col1\" >22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f025e_row5_col0\" class=\"data row5 col0\" >not allowed to collect</td>\n",
       "      <td id=\"T_f025e_row5_col1\" class=\"data row5 col1\" >25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_0865d_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_0865d_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_0865d_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >cause_of_death</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_0865d_row0_col1\" class=\"data row0 col1\" >2098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row1_col0\" class=\"data row1 col0\" >Not Reported</td>\n",
       "      <td id=\"T_0865d_row1_col1\" class=\"data row1 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row2_col0\" class=\"data row2 col0\" >Cancer Related</td>\n",
       "      <td id=\"T_0865d_row2_col1\" class=\"data row2 col1\" >63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row3_col0\" class=\"data row3 col0\" >Infection</td>\n",
       "      <td id=\"T_0865d_row3_col1\" class=\"data row3 col1\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row4_col0\" class=\"data row4 col0\" >Not Cancer Related</td>\n",
       "      <td id=\"T_0865d_row4_col1\" class=\"data row4 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row5_col0\" class=\"data row5 col0\" >Unknown</td>\n",
       "      <td id=\"T_0865d_row5_col1\" class=\"data row5 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0865d_row6_col0\" class=\"data row6 col0\" >Surgical Complications</td>\n",
       "      <td id=\"T_0865d_row6_col1\" class=\"data row6 col1\" >2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults\n",
    "\n",
    "#   total : 2384    \n",
    "#   files : 4099497   \n",
    "# system\tcount\n",
    "# IDC\t1955\n",
    "# PDC\t309\n",
    "# GDC\t1454 \n",
    "# sex\tcount\n",
    "# None\t748\n",
    "# female\t653\n",
    "# male\t980\n",
    "# not reported\t3 \n",
    "# race\tcount\n",
    "# None\t748\n",
    "# not reported\t136\n",
    "# white\t1311\n",
    "# not allowed to collect\t25\n",
    "# asian\t33\n",
    "# black or african american\t96\n",
    "# american indian or alaska native\t4\n",
    "# other\t9\n",
    "# Unknown\t21\n",
    "# native hawaiian or other pacific islander\t1 \n",
    "# ethnicity\tcount\n",
    "# None\t748\n",
    "# hispanic or latino\t85\n",
    "# not reported\t219\n",
    "# not hispanic or latino\t1285\n",
    "# Unknown\t22\n",
    "# not allowed to collect\t25 \n",
    "# cause_of_death\tcount\n",
    "# None\t2098\n",
    "# Not Reported\t200\n",
    "# Cancer Related\t63\n",
    "# Infection\t3\n",
    "# Not Cancer Related\t9\n",
    "# Unknown\t9\n",
    "# Surgical Complications\t2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e6d522",
   "metadata": {},
   "source": [
    "By default, the results are displayed as a table for easy previewing of the data. Since we queried the `subject` endpoint, our default results tell us `subject` level information, that is, information about unique individuals: their sex, race, age, species, etc. Using counts gives us back a nice pivot table type summary of the countable fields for Subjects. Note that above the table it also tells you the total subject count, as well as how many files are associated with those subjects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05e52f3f",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "    \n",
    "    \n",
    "    \n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of subjects returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> files</td>\n",
    "    <td class=\"tg-0lax\"> The number of files that match this search.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">species</td>\n",
    "    <td class=\"tg-0lax\"> The taxonomic group (e.g. species) of the subject.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">sex</td>\n",
    "    <td class=\"tg-0lax\">  The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics. </td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">race</td>\n",
    "    <td class=\"tg-0lax\"> An arbitrary classification of a taxonomic group that is a division of a species.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">ethnicity</td>\n",
    "    <td class=\"tg-0lax\"> An individual's self-described social and cultural grouping.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> cause_of_death</td>\n",
    "    <td class=\"tg-0lax\"> The cause of death, if known</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5356bf41",
   "metadata": {},
   "source": [
    "This gives you a quick way to assess whether the full search results will have the data fields you require. But if you want to get the underlying data for your own downstream applications, you can also get the raw numbers by calling the zeroth value of the variable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "869dfd5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'total': 2384,\n",
       " 'files': 4099497,\n",
       " 'system': [{'system': 'IDC', 'count': 1955},\n",
       "  {'system': 'PDC', 'count': 309},\n",
       "  {'system': 'GDC', 'count': 1454}],\n",
       " 'sex': [{'sex': 'null', 'count': 748},\n",
       "  {'sex': 'female', 'count': 653},\n",
       "  {'sex': 'male', 'count': 980},\n",
       "  {'sex': 'not reported', 'count': 3}],\n",
       " 'race': [{'race': 'null', 'count': 748},\n",
       "  {'race': 'not reported', 'count': 136},\n",
       "  {'race': 'white', 'count': 1311},\n",
       "  {'race': 'not allowed to collect', 'count': 25},\n",
       "  {'race': 'asian', 'count': 33},\n",
       "  {'race': 'black or african american', 'count': 96},\n",
       "  {'race': 'american indian or alaska native', 'count': 4},\n",
       "  {'race': 'other', 'count': 9},\n",
       "  {'race': 'Unknown', 'count': 21},\n",
       "  {'race': 'native hawaiian or other pacific islander', 'count': 1}],\n",
       " 'ethnicity': [{'ethnicity': 'null', 'count': 748},\n",
       "  {'ethnicity': 'hispanic or latino', 'count': 85},\n",
       "  {'ethnicity': 'not reported', 'count': 219},\n",
       "  {'ethnicity': 'not hispanic or latino', 'count': 1285},\n",
       "  {'ethnicity': 'Unknown', 'count': 22},\n",
       "  {'ethnicity': 'not allowed to collect', 'count': 25}],\n",
       " 'cause_of_death': [{'cause_of_death': 'null', 'count': 2098},\n",
       "  {'cause_of_death': 'Not Reported', 'count': 200},\n",
       "  {'cause_of_death': 'Cancer Related', 'count': 63},\n",
       "  {'cause_of_death': 'Infection', 'count': 3},\n",
       "  {'cause_of_death': 'Not Cancer Related', 'count': 9},\n",
       "  {'cause_of_death': 'Unknown', 'count': 9},\n",
       "  {'cause_of_death': 'Surgical Complications', 'count': 2}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults[0]\n",
    "\n",
    "# {'total': 2384,\n",
    "#  'files': 4099497,\n",
    "#  'system': [{'system': 'IDC', 'count': 1955},\n",
    "#   {'system': 'PDC', 'count': 309},\n",
    "#   {'system': 'GDC', 'count': 1454}],\n",
    "#  'sex': [{'sex': 'null', 'count': 748},\n",
    "#   {'sex': 'female', 'count': 653},\n",
    "#   {'sex': 'male', 'count': 980},\n",
    "#   {'sex': 'not reported', 'count': 3}],\n",
    "#  'race': [{'race': 'null', 'count': 748},\n",
    "#   {'race': 'not reported', 'count': 136},\n",
    "#   {'race': 'white', 'count': 1311},\n",
    "#   {'race': 'not allowed to collect', 'count': 25},\n",
    "#   {'race': 'asian', 'count': 33},\n",
    "#   {'race': 'black or african american', 'count': 96},\n",
    "#   {'race': 'american indian or alaska native', 'count': 4},\n",
    "#   {'race': 'other', 'count': 9},\n",
    "#   {'race': 'Unknown', 'count': 21},\n",
    "#   {'race': 'native hawaiian or other pacific islander', 'count': 1}],\n",
    "#  'ethnicity': [{'ethnicity': 'null', 'count': 748},\n",
    "#   {'ethnicity': 'hispanic or latino', 'count': 85},\n",
    "#   {'ethnicity': 'not reported', 'count': 219},\n",
    "#   {'ethnicity': 'not hispanic or latino', 'count': 1285},\n",
    "#   {'ethnicity': 'Unknown', 'count': 22},\n",
    "#   {'ethnicity': 'not allowed to collect', 'count': 25}],\n",
    "#  'cause_of_death': [{'cause_of_death': 'null', 'count': 2098},\n",
    "#   {'cause_of_death': 'Not Reported', 'count': 200},\n",
    "#   {'cause_of_death': 'Cancer Related', 'count': 63},\n",
    "#   {'cause_of_death': 'Infection', 'count': 3},\n",
    "#   {'cause_of_death': 'Not Cancer Related', 'count': 9},\n",
    "#   {'cause_of_death': 'Unknown', 'count': 9},\n",
    "#   {'cause_of_death': 'Surgical Complications', 'count': 2}]}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73745aa6",
   "metadata": {},
   "source": [
    "### researchsubject\n",
    "\n",
    "If we're interested in what researchsubjects meet our critera, we can also run our query against the researchsubject endpoint. Lets run it without saving to a variable this time to make it a bit quicker:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2d8c874",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3568 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3716</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m3716\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">  files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4099477</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "  files : \u001b[1;36m4099477\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a1e4e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a1e4e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a1e4e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a1e4e_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_a1e4e_row0_col1\" class=\"data row0 col1\" >1454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a1e4e_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_a1e4e_row1_col1\" class=\"data row1 col1\" >309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a1e4e_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_a1e4e_row2_col1\" class=\"data row2 col1\" >1953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_25501_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_25501_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_25501_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row0_col0\" class=\"data row0 col0\" >Gliomas</td>\n",
       "      <td id=\"T_25501_row0_col1\" class=\"data row0 col1\" >1246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row1_col0\" class=\"data row1 col0\" >Glioblastoma</td>\n",
       "      <td id=\"T_25501_row1_col1\" class=\"data row1 col1\" >100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row2_col0\" class=\"data row2 col0\" >None</td>\n",
       "      <td id=\"T_25501_row2_col1\" class=\"data row2 col1\" >1953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row3_col0\" class=\"data row3 col0\" >Pediatric/AYA Brain Tumors</td>\n",
       "      <td id=\"T_25501_row3_col1\" class=\"data row3 col1\" >199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row4_col0\" class=\"data row4 col0\" >Other</td>\n",
       "      <td id=\"T_25501_row4_col1\" class=\"data row4 col1\" >10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row5_col0\" class=\"data row5 col0\" >Not Reported</td>\n",
       "      <td id=\"T_25501_row5_col1\" class=\"data row5 col1\" >11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row6_col0\" class=\"data row6 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_25501_row6_col1\" class=\"data row6 col1\" >66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row7_col0\" class=\"data row7 col0\" >Germ Cell Neoplasms</td>\n",
       "      <td id=\"T_25501_row7_col1\" class=\"data row7 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row8_col0\" class=\"data row8 col0\" >Malignant Lymphomas, NOS or Diffuse</td>\n",
       "      <td id=\"T_25501_row8_col1\" class=\"data row8 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row9_col0\" class=\"data row9 col0\" >Not Applicable</td>\n",
       "      <td id=\"T_25501_row9_col1\" class=\"data row9 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row10_col0\" class=\"data row10 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_25501_row10_col1\" class=\"data row10 col1\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_25501_row11_col0\" class=\"data row11 col0\" >Mature B-Cell Lymphomas</td>\n",
       "      <td id=\"T_25501_row11_col1\" class=\"data row11 col1\" >2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_afcb0_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_afcb0_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_afcb0_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_afcb0_row0_col0\" class=\"data row0 col0\" >Brain</td>\n",
       "      <td id=\"T_afcb0_row0_col1\" class=\"data row0 col1\" >3716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.researchsubject.count.run()\n",
    "\n",
    "#    total : 3716    \n",
    "#   files : 4099477   \n",
    "# system\tcount\n",
    "# GDC\t1454\n",
    "# PDC\t309\n",
    "# IDC\t1953 \n",
    "# primary_diagnosis_condition\tcount\n",
    "# Gliomas\t1246\n",
    "# Glioblastoma\t100\n",
    "# None\t1953\n",
    "# Pediatric/AYA Brain Tumors\t199\n",
    "# Other\t10\n",
    "# Not Reported\t11\n",
    "# Neoplasms, NOS\t66\n",
    "# Germ Cell Neoplasms\t104\n",
    "# Malignant Lymphomas, NOS or Diffuse\t14\n",
    "# Not Applicable\t9\n",
    "# Neuroepitheliomatous Neoplasms\t2\n",
    "# Mature B-Cell Lymphomas\t2 \n",
    "# primary_diagnosis_site\tcount\n",
    "# Brain\t3716"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4564beee",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of researchsubjects returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> files</td>\n",
    "    <td class=\"tg-0lax\"> The number of files that match this search.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">primary_diagnosis_condition</td>\n",
    "    <td class=\"tg-0lax\"> The text term used to describe the type of malignant disease.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">primary_diagnosis_site</td>\n",
    "    <td class=\"tg-0lax\"> The text term used to describe the primary site of disease.</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a67cf09",
   "metadata": {},
   "source": [
    "### diagnosis\n",
    "\n",
    "The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria. :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7770d68c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3410 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">1756</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m1756\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_f741d_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_f741d_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f741d_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_f741d_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_f741d_row0_col1\" class=\"data row0 col1\" >1427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f741d_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_f741d_row1_col1\" class=\"data row1 col1\" >329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_f5e20_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_f5e20_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f5e20_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row0_col0\" class=\"data row0 col0\" >Glioblastoma</td>\n",
       "      <td id=\"T_f5e20_row0_col1\" class=\"data row0 col1\" >821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row1_col0\" class=\"data row1 col0\" >Oligodendroglioma, NOS</td>\n",
       "      <td id=\"T_f5e20_row1_col1\" class=\"data row1 col1\" >112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row2_col0\" class=\"data row2 col0\" >Oligodendroglioma, anaplastic</td>\n",
       "      <td id=\"T_f5e20_row2_col1\" class=\"data row2 col1\" >78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row3_col0\" class=\"data row3 col0\" >Ganglioglioma, NOS</td>\n",
       "      <td id=\"T_f5e20_row3_col1\" class=\"data row3 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row4_col0\" class=\"data row4 col0\" >Mixed glioma</td>\n",
       "      <td id=\"T_f5e20_row4_col1\" class=\"data row4 col1\" >131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row5_col0\" class=\"data row5 col0\" >Mixed germ cell tumor</td>\n",
       "      <td id=\"T_f5e20_row5_col1\" class=\"data row5 col1\" >79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row6_col0\" class=\"data row6 col0\" >Astrocytoma, anaplastic</td>\n",
       "      <td id=\"T_f5e20_row6_col1\" class=\"data row6 col1\" >130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row7_col0\" class=\"data row7 col0\" >Glioma, NOS</td>\n",
       "      <td id=\"T_f5e20_row7_col1\" class=\"data row7 col1\" >93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row8_col0\" class=\"data row8 col0\" >Neoplasm, malignant</td>\n",
       "      <td id=\"T_f5e20_row8_col1\" class=\"data row8 col1\" >50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row9_col0\" class=\"data row9 col0\" >Astrocytoma, NOS</td>\n",
       "      <td id=\"T_f5e20_row9_col1\" class=\"data row9 col1\" >64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row10_col0\" class=\"data row10 col0\" >Oligoastrocytoma</td>\n",
       "      <td id=\"T_f5e20_row10_col1\" class=\"data row10 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row11_col0\" class=\"data row11 col0\" >Glioma, malignant</td>\n",
       "      <td id=\"T_f5e20_row11_col1\" class=\"data row11 col1\" >26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row12_col0\" class=\"data row12 col0\" >Glioblastoma multiforme</td>\n",
       "      <td id=\"T_f5e20_row12_col1\" class=\"data row12 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row13_col0\" class=\"data row13 col0\" >Medulloblastoma, NOS</td>\n",
       "      <td id=\"T_f5e20_row13_col1\" class=\"data row13 col1\" >22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row14_col0\" class=\"data row14 col0\" >Ependymoma, NOS</td>\n",
       "      <td id=\"T_f5e20_row14_col1\" class=\"data row14 col1\" >32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row15_col0\" class=\"data row15 col0\" >Atypical teratoid/rhabdoid tumor</td>\n",
       "      <td id=\"T_f5e20_row15_col1\" class=\"data row15 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row16_col0\" class=\"data row16 col0\" >Not Reported</td>\n",
       "      <td id=\"T_f5e20_row16_col1\" class=\"data row16 col1\" >10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row17_col0\" class=\"data row17 col0\" >Teratoma, malignant, NOS</td>\n",
       "      <td id=\"T_f5e20_row17_col1\" class=\"data row17 col1\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row18_col0\" class=\"data row18 col0\" >Craniopharyngioma</td>\n",
       "      <td id=\"T_f5e20_row18_col1\" class=\"data row18 col1\" >16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row19_col0\" class=\"data row19 col0\" >Yolk sac tumor</td>\n",
       "      <td id=\"T_f5e20_row19_col1\" class=\"data row19 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row20_col0\" class=\"data row20 col0\" >Malignant lymphoma, NOS</td>\n",
       "      <td id=\"T_f5e20_row20_col1\" class=\"data row20 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row21_col0\" class=\"data row21 col0\" >Malignant lymphoma, large B-cell, diffuse, NOS</td>\n",
       "      <td id=\"T_f5e20_row21_col1\" class=\"data row21 col1\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row22_col0\" class=\"data row22 col0\" >Gliosarcoma</td>\n",
       "      <td id=\"T_f5e20_row22_col1\" class=\"data row22 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row23_col0\" class=\"data row23 col0\" >Teratoma, benign</td>\n",
       "      <td id=\"T_f5e20_row23_col1\" class=\"data row23 col1\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row24_col0\" class=\"data row24 col0\" >Papillary glioneuronal tumor</td>\n",
       "      <td id=\"T_f5e20_row24_col1\" class=\"data row24 col1\" >2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row25_col0\" class=\"data row25 col0\" >Germinoma</td>\n",
       "      <td id=\"T_f5e20_row25_col1\" class=\"data row25 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row26_col0\" class=\"data row26 col0\" >Embryonal carcinoma, NOS</td>\n",
       "      <td id=\"T_f5e20_row26_col1\" class=\"data row26 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f5e20_row27_col0\" class=\"data row27 col0\" >Neoplasm, uncertain whether benign or malignant</td>\n",
       "      <td id=\"T_f5e20_row27_col1\" class=\"data row27 col1\" >13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_b94b4_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_b94b4_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_b94b4_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >stage</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_b94b4_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_b94b4_row0_col1\" class=\"data row0 col1\" >1427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_b94b4_row1_col0\" class=\"data row1 col0\" >Unknown</td>\n",
       "      <td id=\"T_b94b4_row1_col1\" class=\"data row1 col1\" >219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_b94b4_row2_col0\" class=\"data row2 col0\" >Not Reported</td>\n",
       "      <td id=\"T_b94b4_row2_col1\" class=\"data row2 col1\" >110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_984ec_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_984ec_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_984ec_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >grade</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row0_col0\" class=\"data row0 col0\" >not reported</td>\n",
       "      <td id=\"T_984ec_row0_col1\" class=\"data row0 col1\" >1116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row1_col0\" class=\"data row1 col0\" >G4</td>\n",
       "      <td id=\"T_984ec_row1_col1\" class=\"data row1 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row2_col0\" class=\"data row2 col0\" >G1</td>\n",
       "      <td id=\"T_984ec_row2_col1\" class=\"data row2 col1\" >98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row3_col0\" class=\"data row3 col0\" >High Grade</td>\n",
       "      <td id=\"T_984ec_row3_col1\" class=\"data row3 col1\" >26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row4_col0\" class=\"data row4 col0\" >None</td>\n",
       "      <td id=\"T_984ec_row4_col1\" class=\"data row4 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row5_col0\" class=\"data row5 col0\" >G2</td>\n",
       "      <td id=\"T_984ec_row5_col1\" class=\"data row5 col1\" >52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row6_col0\" class=\"data row6 col0\" >Not Reported</td>\n",
       "      <td id=\"T_984ec_row6_col1\" class=\"data row6 col1\" >392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_984ec_row7_col0\" class=\"data row7 col0\" >Low Grade</td>\n",
       "      <td id=\"T_984ec_row7_col1\" class=\"data row7 col1\" >9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.diagnosis.count.run()\n",
    "\n",
    "#     total : 1756    \n",
    "# system\tcount\n",
    "# GDC\t1427\n",
    "# PDC\t329 \n",
    "# primary_diagnosis\tcount\n",
    "# Glioblastoma\t821\n",
    "# Oligodendroglioma, NOS\t112\n",
    "# Oligodendroglioma, anaplastic\t78\n",
    "# Ganglioglioma, NOS\t18\n",
    "# Mixed glioma\t131\n",
    "# Mixed germ cell tumor\t79\n",
    "# Astrocytoma, anaplastic\t130\n",
    "# Glioma, NOS\t93\n",
    "# Neoplasm, malignant\t50\n",
    "# Astrocytoma, NOS\t64\n",
    "# Oligoastrocytoma\t1\n",
    "# Glioma, malignant\t26\n",
    "# Glioblastoma multiforme\t4\n",
    "# Medulloblastoma, NOS\t22\n",
    "# Ependymoma, NOS\t32\n",
    "# Atypical teratoid/rhabdoid tumor\t12\n",
    "# Not Reported\t10\n",
    "# Teratoma, malignant, NOS\t2\n",
    "# Craniopharyngioma\t16\n",
    "# Yolk sac tumor\t8\n",
    "# Malignant lymphoma, NOS\t14\n",
    "# Malignant lymphoma, large B-cell, diffuse, NOS\t2\n",
    "# Gliosarcoma\t1\n",
    "# Teratoma, benign\t3\n",
    "# Papillary glioneuronal tumor\t2\n",
    "# Germinoma\t4\n",
    "# Embryonal carcinoma, NOS\t8\n",
    "# Neoplasm, uncertain whether benign or malignant\t13 \n",
    "# stage\tcount\n",
    "# None\t1427\n",
    "# Unknown\t219\n",
    "# Not Reported\t110 \n",
    "# grade\tcount\n",
    "# not reported\t1116\n",
    "# G4\t36\n",
    "# G1\t98\n",
    "# High Grade\t26\n",
    "# None\t27\n",
    "# G2\t52\n",
    "# Not Reported\t392\n",
    "# Low Grade\t9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb73357c",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Diagnosis Field Definitions</h3>\n",
    "\n",
    "<i>A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</i>\n",
    "\n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of diagnoses returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-za14\">primary_diagnosis</td>\n",
    "    <td class=\"tg-0pky\"> The diagnosis instance that qualified a subject for inclusion on a ResearchProject.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-za14\">stage</td>\n",
    "    <td class=\"tg-0pky\"> The extent of a cancer in the body.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">grade</td>\n",
    "    <td class=\"tg-0lax\"> The degree of abnormality of cancer cells.</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "\n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97442718",
   "metadata": {},
   "source": [
    "### treatment\n",
    "\n",
    "The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "be1ac64f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3701 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2386</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m2386\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_6c8f6_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_6c8f6_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_6c8f6_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_6c8f6_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_6c8f6_row0_col1\" class=\"data row0 col1\" >2386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_ffaa8_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_ffaa8_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_ffaa8_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >treatment_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row0_col0\" class=\"data row0 col0\" >Targeted Molecular Therapy</td>\n",
       "      <td id=\"T_ffaa8_row0_col1\" class=\"data row0 col1\" >23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row1_col0\" class=\"data row1 col0\" >Pharmaceutical Therapy, NOS</td>\n",
       "      <td id=\"T_ffaa8_row1_col1\" class=\"data row1 col1\" >1117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row2_col0\" class=\"data row2 col0\" >Radiation Therapy, NOS</td>\n",
       "      <td id=\"T_ffaa8_row2_col1\" class=\"data row2 col1\" >1139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row3_col0\" class=\"data row3 col0\" >Chemotherapy</td>\n",
       "      <td id=\"T_ffaa8_row3_col1\" class=\"data row3 col1\" >30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row4_col0\" class=\"data row4 col0\" >Immunotherapy (Including Vaccines)</td>\n",
       "      <td id=\"T_ffaa8_row4_col1\" class=\"data row4 col1\" >23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row5_col0\" class=\"data row5 col0\" >None</td>\n",
       "      <td id=\"T_ffaa8_row5_col1\" class=\"data row5 col1\" >30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row6_col0\" class=\"data row6 col0\" >Surgery</td>\n",
       "      <td id=\"T_ffaa8_row6_col1\" class=\"data row6 col1\" >23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ffaa8_row7_col0\" class=\"data row7 col0\" >Radiation, Proton Beam</td>\n",
       "      <td id=\"T_ffaa8_row7_col1\" class=\"data row7 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_b1d12_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_b1d12_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_b1d12_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >treatment_effect</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_b1d12_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_b1d12_row0_col1\" class=\"data row0 col1\" >2386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.treatment.count.run()\n",
    "\n",
    "#     total : 2386    \n",
    "# system\tcount\n",
    "# GDC\t2386 \n",
    "# treatment_type\tcount\n",
    "# Targeted Molecular Therapy\t23\n",
    "# Pharmaceutical Therapy, NOS\t1117\n",
    "# Radiation Therapy, NOS\t1139\n",
    "# Chemotherapy\t30\n",
    "# Immunotherapy (Including Vaccines)\t23\n",
    "# None\t30\n",
    "# Surgery\t23\n",
    "# Radiation, Proton Beam\t1 \n",
    "# treatment_effect\tcount\n",
    "# None\t2386"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec5dda7f",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Treatment Field Definitions</h3>\n",
    "\n",
    "<i><i> Medication administration or other treatment types. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>\n",
    "    \n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of treatments returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-za14\">treatment_type</td>\n",
    "    <td class=\"tg-0pky\"> The treatment type including medication/therapeutics or other procedures.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">treatment_effect</td>\n",
    "    <td class=\"tg-0lax\">The effect of a treatment on the diagnosis or tumor. </td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    " \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c00bec8",
   "metadata": {},
   "source": [
    "### specimens\n",
    "\n",
    "We can use this same query to see what specimens are available for brain tissue at the CDA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51960eed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3566 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">39201</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   total : \u001b[1;36m39201\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">50546</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m50546\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_40e3a_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_40e3a_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_40e3a_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_40e3a_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_40e3a_row0_col1\" class=\"data row0 col1\" >38543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_40e3a_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_40e3a_row1_col1\" class=\"data row1 col1\" >658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_54a1b_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_54a1b_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_54a1b_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_disease_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row0_col0\" class=\"data row0 col0\" >Gliomas</td>\n",
       "      <td id=\"T_54a1b_row0_col1\" class=\"data row0 col1\" >37567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row1_col0\" class=\"data row1 col0\" >Glioblastoma</td>\n",
       "      <td id=\"T_54a1b_row1_col1\" class=\"data row1 col1\" >200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row2_col0\" class=\"data row2 col0\" >Other</td>\n",
       "      <td id=\"T_54a1b_row2_col1\" class=\"data row2 col1\" >20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row3_col0\" class=\"data row3 col0\" >Not Reported</td>\n",
       "      <td id=\"T_54a1b_row3_col1\" class=\"data row3 col1\" >121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row4_col0\" class=\"data row4 col0\" >Pediatric/AYA Brain Tumors</td>\n",
       "      <td id=\"T_54a1b_row4_col1\" class=\"data row4 col1\" >438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row5_col0\" class=\"data row5 col0\" >Mature B-Cell Lymphomas</td>\n",
       "      <td id=\"T_54a1b_row5_col1\" class=\"data row5 col1\" >54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row6_col0\" class=\"data row6 col0\" >Germ Cell Neoplasms</td>\n",
       "      <td id=\"T_54a1b_row6_col1\" class=\"data row6 col1\" >416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row7_col0\" class=\"data row7 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_54a1b_row7_col1\" class=\"data row7 col1\" >285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row8_col0\" class=\"data row8 col0\" >Not Applicable</td>\n",
       "      <td id=\"T_54a1b_row8_col1\" class=\"data row8 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row9_col0\" class=\"data row9 col0\" >Malignant Lymphomas, NOS or Diffuse</td>\n",
       "      <td id=\"T_54a1b_row9_col1\" class=\"data row9 col1\" >56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_54a1b_row10_col0\" class=\"data row10 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_54a1b_row10_col1\" class=\"data row10 col1\" >8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_5040a_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_5040a_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_5040a_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >source_material_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row0_col0\" class=\"data row0 col0\" >Blood Derived Normal</td>\n",
       "      <td id=\"T_5040a_row0_col1\" class=\"data row0 col1\" >10074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row1_col0\" class=\"data row1 col0\" >Primary Tumor</td>\n",
       "      <td id=\"T_5040a_row1_col1\" class=\"data row1 col1\" >27570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row2_col0\" class=\"data row2 col0\" >Solid Tissue Normal</td>\n",
       "      <td id=\"T_5040a_row2_col1\" class=\"data row2 col1\" >538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row3_col0\" class=\"data row3 col0\" >Next Generation Cancer Model</td>\n",
       "      <td id=\"T_5040a_row3_col1\" class=\"data row3 col1\" >169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row4_col0\" class=\"data row4 col0\" >Recurrent Tumor</td>\n",
       "      <td id=\"T_5040a_row4_col1\" class=\"data row4 col1\" >513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row5_col0\" class=\"data row5 col0\" >Buccal Cell Normal</td>\n",
       "      <td id=\"T_5040a_row5_col1\" class=\"data row5 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row6_col0\" class=\"data row6 col0\" >Expanded Next Generation Cancer Model</td>\n",
       "      <td id=\"T_5040a_row6_col1\" class=\"data row6 col1\" >35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row7_col0\" class=\"data row7 col0\" >Not Reported</td>\n",
       "      <td id=\"T_5040a_row7_col1\" class=\"data row7 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_5040a_row8_col0\" class=\"data row8 col0\" >Metastatic</td>\n",
       "      <td id=\"T_5040a_row8_col1\" class=\"data row8 col1\" >252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_f94cf_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_f94cf_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f94cf_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >specimen_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_f94cf_row0_col0\" class=\"data row0 col0\" >analyte</td>\n",
       "      <td id=\"T_f94cf_row0_col1\" class=\"data row0 col1\" >6671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f94cf_row1_col0\" class=\"data row1 col0\" >aliquot</td>\n",
       "      <td id=\"T_f94cf_row1_col1\" class=\"data row1 col1\" >18696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f94cf_row2_col0\" class=\"data row2 col0\" >portion</td>\n",
       "      <td id=\"T_f94cf_row2_col1\" class=\"data row2 col1\" >5992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f94cf_row3_col0\" class=\"data row3 col0\" >sample</td>\n",
       "      <td id=\"T_f94cf_row3_col1\" class=\"data row3 col1\" >4090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_f94cf_row4_col0\" class=\"data row4 col0\" >slide</td>\n",
       "      <td id=\"T_f94cf_row4_col1\" class=\"data row4 col1\" >3752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.specimen.count.run()\n",
    "\n",
    "#    total : 39201    \n",
    "#    files : 50546    \n",
    "# system\tcount\n",
    "# GDC\t38543\n",
    "# PDC\t658 \n",
    "# primary_disease_type\tcount\n",
    "# Gliomas\t37567\n",
    "# Glioblastoma\t200\n",
    "# Other\t20\n",
    "# Not Reported\t121\n",
    "# Pediatric/AYA Brain Tumors\t438\n",
    "# Mature B-Cell Lymphomas\t54\n",
    "# Germ Cell Neoplasms\t416\n",
    "# Neoplasms, NOS\t285\n",
    "# Not Applicable\t36\n",
    "# Malignant Lymphomas, NOS or Diffuse\t56\n",
    "# Neuroepitheliomatous Neoplasms\t8 \n",
    "# source_material_type\tcount\n",
    "# Blood Derived Normal\t10074\n",
    "# Primary Tumor\t27570\n",
    "# Solid Tissue Normal\t538\n",
    "# Next Generation Cancer Model\t169\n",
    "# Recurrent Tumor\t513\n",
    "# Buccal Cell Normal\t14\n",
    "# Expanded Next Generation Cancer Model\t35\n",
    "# Not Reported\t36\n",
    "# Metastatic\t252 \n",
    "# specimen_type\tcount\n",
    "# analyte\t6671\n",
    "# aliquot\t18696\n",
    "# portion\t5992\n",
    "# sample\t4090\n",
    "# slide\t3752"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f254547",
   "metadata": {},
   "source": [
    "Nearly 40,000 specimens with over 50,000 files meet our search criteria! We would typically expect this number to be much larger than our number of subjects or research_subjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3ed75e5",
   "metadata": {},
   "source": [
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Specimen Field Definitions</h3>\n",
    "\n",
    "<i><i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>\n",
    " A given specimen will have only a single subject ID and a single research subject ID</i>\n",
    "    \n",
    "    \n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "    \n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of specimens returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> files</td>\n",
    "    <td class=\"tg-0lax\"> The number of files that match this search.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-za14\">primary_disease_type</td>\n",
    "    <td class=\"tg-0pky\"> The text term used to describe the type of malignant disease.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">source_material_type</td>\n",
    "    <td class=\"tg-0lax\"> The general kind of material from which the specimen was derived.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">specimen_type</td>\n",
    "    <td class=\"tg-0lax\"> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide.</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccb0f1c",
   "metadata": {},
   "source": [
    "### file\n",
    "\n",
    "The files endpoint returns all files that match our query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d6e0639a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3680 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">  total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4099497</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "  total : \u001b[1;36m4099497\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_fea84_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_fea84_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_fea84_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_fea84_row0_col0\" class=\"data row0 col0\" >IDC</td>\n",
       "      <td id=\"T_fea84_row0_col1\" class=\"data row0 col1\" >4045816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fea84_row1_col0\" class=\"data row1 col0\" >GDC</td>\n",
       "      <td id=\"T_fea84_row1_col1\" class=\"data row1 col1\" >50633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_fea84_row2_col0\" class=\"data row2 col0\" >PDC</td>\n",
       "      <td id=\"T_fea84_row2_col1\" class=\"data row2 col1\" >3048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_9940a_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_9940a_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_9940a_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row0_col0\" class=\"data row0 col0\" >Imaging</td>\n",
       "      <td id=\"T_9940a_row0_col1\" class=\"data row0 col1\" >4045816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row1_col0\" class=\"data row1 col0\" >Simple Nucleotide Variation</td>\n",
       "      <td id=\"T_9940a_row1_col1\" class=\"data row1 col1\" >20481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row2_col0\" class=\"data row2 col0\" >Sequencing Reads</td>\n",
       "      <td id=\"T_9940a_row2_col1\" class=\"data row2 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row3_col0\" class=\"data row3 col0\" >Transcriptome Profiling</td>\n",
       "      <td id=\"T_9940a_row3_col1\" class=\"data row3 col1\" >3112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row4_col0\" class=\"data row4 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_9940a_row4_col1\" class=\"data row4 col1\" >5582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row5_col0\" class=\"data row5 col0\" >Copy Number Variation</td>\n",
       "      <td id=\"T_9940a_row5_col1\" class=\"data row5 col1\" >6909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row6_col0\" class=\"data row6 col0\" >DNA Methylation</td>\n",
       "      <td id=\"T_9940a_row6_col1\" class=\"data row6 col1\" >3339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row7_col0\" class=\"data row7 col0\" >Clinical</td>\n",
       "      <td id=\"T_9940a_row7_col1\" class=\"data row7 col1\" >1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row8_col0\" class=\"data row8 col0\" >Structural Variation</td>\n",
       "      <td id=\"T_9940a_row8_col1\" class=\"data row8 col1\" >3160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row9_col0\" class=\"data row9 col0\" >Peptide Spectral Matches</td>\n",
       "      <td id=\"T_9940a_row9_col1\" class=\"data row9 col1\" >1524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row10_col0\" class=\"data row10 col0\" >Processed Mass Spectra</td>\n",
       "      <td id=\"T_9940a_row10_col1\" class=\"data row10 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row11_col0\" class=\"data row11 col0\" >Proteome Profiling</td>\n",
       "      <td id=\"T_9940a_row11_col1\" class=\"data row11 col1\" >679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row12_col0\" class=\"data row12 col0\" >Raw Mass Spectra</td>\n",
       "      <td id=\"T_9940a_row12_col1\" class=\"data row12 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9940a_row13_col0\" class=\"data row13 col0\" >Somatic Structural Variation</td>\n",
       "      <td id=\"T_9940a_row13_col1\" class=\"data row13 col1\" >299</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_bc240_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_bc240_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_bc240_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row0_col0\" class=\"data row0 col0\" >DICOM</td>\n",
       "      <td id=\"T_bc240_row0_col1\" class=\"data row0 col1\" >4045816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row1_col0\" class=\"data row1 col0\" >BEDPE</td>\n",
       "      <td id=\"T_bc240_row1_col1\" class=\"data row1 col1\" >1886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row2_col0\" class=\"data row2 col0\" >VCF</td>\n",
       "      <td id=\"T_bc240_row2_col1\" class=\"data row2 col1\" >12255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row3_col0\" class=\"data row3 col0\" >IDAT</td>\n",
       "      <td id=\"T_bc240_row3_col1\" class=\"data row3 col1\" >2226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row4_col0\" class=\"data row4 col0\" >TXT</td>\n",
       "      <td id=\"T_bc240_row4_col1\" class=\"data row4 col1\" >8848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row5_col0\" class=\"data row5 col0\" >BAM</td>\n",
       "      <td id=\"T_bc240_row5_col1\" class=\"data row5 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row6_col0\" class=\"data row6 col0\" >vendor-specific</td>\n",
       "      <td id=\"T_bc240_row6_col1\" class=\"data row6 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row7_col0\" class=\"data row7 col0\" >SVS</td>\n",
       "      <td id=\"T_bc240_row7_col1\" class=\"data row7 col1\" >3629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row8_col0\" class=\"data row8 col0\" >MAF</td>\n",
       "      <td id=\"T_bc240_row8_col1\" class=\"data row8 col1\" >8375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row9_col0\" class=\"data row9 col0\" >TSV</td>\n",
       "      <td id=\"T_bc240_row9_col1\" class=\"data row9 col1\" >4335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row10_col0\" class=\"data row10 col0\" >BCR XML</td>\n",
       "      <td id=\"T_bc240_row10_col1\" class=\"data row10 col1\" >2281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row11_col0\" class=\"data row11 col0\" >BCR SSF XML</td>\n",
       "      <td id=\"T_bc240_row11_col1\" class=\"data row11 col1\" >758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row12_col0\" class=\"data row12 col0\" >tsv</td>\n",
       "      <td id=\"T_bc240_row12_col1\" class=\"data row12 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row13_col0\" class=\"data row13 col0\" >mzIdentML</td>\n",
       "      <td id=\"T_bc240_row13_col1\" class=\"data row13 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row14_col0\" class=\"data row14 col0\" >mzML</td>\n",
       "      <td id=\"T_bc240_row14_col1\" class=\"data row14 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row15_col0\" class=\"data row15 col0\" >MEX</td>\n",
       "      <td id=\"T_bc240_row15_col1\" class=\"data row15 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row16_col0\" class=\"data row16 col0\" >BCR Biotab</td>\n",
       "      <td id=\"T_bc240_row16_col1\" class=\"data row16 col1\" >49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row17_col0\" class=\"data row17 col0\" >BCR OMF XML</td>\n",
       "      <td id=\"T_bc240_row17_col1\" class=\"data row17 col1\" >19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row18_col0\" class=\"data row18 col0\" >HDF5</td>\n",
       "      <td id=\"T_bc240_row18_col1\" class=\"data row18 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row19_col0\" class=\"data row19 col0\" >CDC JSON</td>\n",
       "      <td id=\"T_bc240_row19_col1\" class=\"data row19 col1\" >28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_bc240_row20_col0\" class=\"data row20 col0\" >PDF</td>\n",
       "      <td id=\"T_bc240_row20_col1\" class=\"data row20 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_c3334_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_c3334_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_c3334_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_c3334_row0_col1\" class=\"data row0 col1\" >4045816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row1_col0\" class=\"data row1 col0\" >Open Standard</td>\n",
       "      <td id=\"T_c3334_row1_col1\" class=\"data row1 col1\" >1524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row2_col0\" class=\"data row2 col0\" >Allele-specific Copy Number Segment</td>\n",
       "      <td id=\"T_c3334_row2_col1\" class=\"data row2 col1\" >1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row3_col0\" class=\"data row3 col0\" >Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_c3334_row3_col1\" class=\"data row3 col1\" >11808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row4_col0\" class=\"data row4 col0\" >Gene Level Copy Number</td>\n",
       "      <td id=\"T_c3334_row4_col1\" class=\"data row4 col1\" >1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row5_col0\" class=\"data row5 col0\" >Masked Copy Number Segment</td>\n",
       "      <td id=\"T_c3334_row5_col1\" class=\"data row5 col1\" >2185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row6_col0\" class=\"data row6 col0\" >Proprietary</td>\n",
       "      <td id=\"T_c3334_row6_col1\" class=\"data row6 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row7_col0\" class=\"data row7 col0\" >Transcript Fusion</td>\n",
       "      <td id=\"T_c3334_row7_col1\" class=\"data row7 col1\" >3161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row8_col0\" class=\"data row8 col0\" >Slide Image</td>\n",
       "      <td id=\"T_c3334_row8_col1\" class=\"data row8 col1\" >3629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row9_col0\" class=\"data row9 col0\" >Raw Simple Somatic Mutation</td>\n",
       "      <td id=\"T_c3334_row9_col1\" class=\"data row9 col1\" >6202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row10_col0\" class=\"data row10 col0\" >Aligned Reads</td>\n",
       "      <td id=\"T_c3334_row10_col1\" class=\"data row10 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row11_col0\" class=\"data row11 col0\" >Biospecimen Supplement</td>\n",
       "      <td id=\"T_c3334_row11_col1\" class=\"data row11 col1\" >1953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row12_col0\" class=\"data row12 col0\" >Gene Expression Quantification</td>\n",
       "      <td id=\"T_c3334_row12_col1\" class=\"data row12 col1\" >904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row13_col0\" class=\"data row13 col0\" >Copy Number Segment</td>\n",
       "      <td id=\"T_c3334_row13_col1\" class=\"data row13 col1\" >2334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row14_col0\" class=\"data row14 col0\" >Text</td>\n",
       "      <td id=\"T_c3334_row14_col1\" class=\"data row14 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row15_col0\" class=\"data row15 col0\" >Masked Intensities</td>\n",
       "      <td id=\"T_c3334_row15_col1\" class=\"data row15 col1\" >2226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row16_col0\" class=\"data row16 col0\" >Isoform Expression Quantification</td>\n",
       "      <td id=\"T_c3334_row16_col1\" class=\"data row16 col1\" >643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row17_col0\" class=\"data row17 col0\" >Masked Somatic Mutation</td>\n",
       "      <td id=\"T_c3334_row17_col1\" class=\"data row17 col1\" >1144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row18_col0\" class=\"data row18 col0\" >Protein Expression Quantification</td>\n",
       "      <td id=\"T_c3334_row18_col1\" class=\"data row18 col1\" >679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row19_col0\" class=\"data row19 col0\" >miRNA Expression Quantification</td>\n",
       "      <td id=\"T_c3334_row19_col1\" class=\"data row19 col1\" >643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row20_col0\" class=\"data row20 col0\" >Aggregated Somatic Mutation</td>\n",
       "      <td id=\"T_c3334_row20_col1\" class=\"data row20 col1\" >1144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row21_col0\" class=\"data row21 col0\" >Methylation Beta Value</td>\n",
       "      <td id=\"T_c3334_row21_col1\" class=\"data row21 col1\" >1113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row22_col0\" class=\"data row22 col0\" >Splice Junction Quantification</td>\n",
       "      <td id=\"T_c3334_row22_col1\" class=\"data row22 col1\" >868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row23_col0\" class=\"data row23 col0\" >Clinical Supplement</td>\n",
       "      <td id=\"T_c3334_row23_col1\" class=\"data row23 col1\" >1182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row24_col0\" class=\"data row24 col0\" >Differential Gene Expression</td>\n",
       "      <td id=\"T_c3334_row24_col1\" class=\"data row24 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row25_col0\" class=\"data row25 col0\" >Structural Rearrangement</td>\n",
       "      <td id=\"T_c3334_row25_col1\" class=\"data row25 col1\" >298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row26_col0\" class=\"data row26 col0\" >Gene Level Copy Number Scores</td>\n",
       "      <td id=\"T_c3334_row26_col1\" class=\"data row26 col1\" >99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row27_col0\" class=\"data row27 col0\" >Masked Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_c3334_row27_col1\" class=\"data row27 col1\" >183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row28_col0\" class=\"data row28 col0\" >Single Cell Analysis</td>\n",
       "      <td id=\"T_c3334_row28_col1\" class=\"data row28 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_c3334_row29_col0\" class=\"data row29 col0\" >Pathology Report</td>\n",
       "      <td id=\"T_c3334_row29_col1\" class=\"data row29 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.file.count.run()\n",
    "\n",
    "#  total : 4099497   \n",
    "# system\tcount\n",
    "# IDC\t4045816\n",
    "# GDC\t50633\n",
    "# PDC\t3048 \n",
    "# data_category\tcount\n",
    "# Imaging\t4045816\n",
    "# Simple Nucleotide Variation\t20481\n",
    "# Sequencing Reads\t5885\n",
    "# Transcriptome Profiling\t3112\n",
    "# Biospecimen\t5582\n",
    "# Copy Number Variation\t6909\n",
    "# DNA Methylation\t3339\n",
    "# Clinical\t1187\n",
    "# Structural Variation\t3160\n",
    "# Peptide Spectral Matches\t1524\n",
    "# Processed Mass Spectra\t762\n",
    "# Proteome Profiling\t679\n",
    "# Raw Mass Spectra\t762\n",
    "# Somatic Structural Variation\t299 \n",
    "# file_format\tcount\n",
    "# DICOM\t4045816\n",
    "# BEDPE\t1886\n",
    "# VCF\t12255\n",
    "# IDAT\t2226\n",
    "# TXT\t8848\n",
    "# BAM\t5885\n",
    "# vendor-specific\t762\n",
    "# SVS\t3629\n",
    "# MAF\t8375\n",
    "# TSV\t4335\n",
    "# BCR XML\t2281\n",
    "# BCR SSF XML\t758\n",
    "# tsv\t762\n",
    "# mzIdentML\t762\n",
    "# mzML\t762\n",
    "# MEX\t36\n",
    "# BCR Biotab\t49\n",
    "# BCR OMF XML\t19\n",
    "# HDF5\t18\n",
    "# CDC JSON\t28\n",
    "# PDF\t5 \n",
    "# data_type\tcount\n",
    "# None\t4045816\n",
    "# Open Standard\t1524\n",
    "# Allele-specific Copy Number Segment\t1071\n",
    "# Annotated Somatic Mutation\t11808\n",
    "# Gene Level Copy Number\t1220\n",
    "# Masked Copy Number Segment\t2185\n",
    "# Proprietary\t762\n",
    "# Transcript Fusion\t3161\n",
    "# Slide Image\t3629\n",
    "# Raw Simple Somatic Mutation\t6202\n",
    "# Aligned Reads\t5885\n",
    "# Biospecimen Supplement\t1953\n",
    "# Gene Expression Quantification\t904\n",
    "# Copy Number Segment\t2334\n",
    "# Text\t762\n",
    "# Masked Intensities\t2226\n",
    "# Isoform Expression Quantification\t643\n",
    "# Masked Somatic Mutation\t1144\n",
    "# Protein Expression Quantification\t679\n",
    "# miRNA Expression Quantification\t643\n",
    "# Aggregated Somatic Mutation\t1144\n",
    "# Methylation Beta Value\t1113\n",
    "# Splice Junction Quantification\t868\n",
    "# Clinical Supplement\t1182\n",
    "# Differential Gene Expression\t18\n",
    "# Structural Rearrangement\t298\n",
    "# Gene Level Copy Number Scores\t99\n",
    "# Masked Annotated Somatic Mutation\t183\n",
    "# Single Cell Analysis\t36\n",
    "# Pathology Report\t5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34044c9",
   "metadata": {},
   "source": [
    "There are a huge number of files (4099497) that match our search. Likely we would want to additionaly filter the results by file format or data type to get only files we can use. See all the ways you can filter and refine searches  with more search terms in the <a href=\"../Operators\">Operators</a> notebook.\n",
    "\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-7zrl{text-align:left;vertical-align:bottom}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"> id (`total`)</td>\n",
    "    <td class=\"tg-0lax\"> The overall number of files returned.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">identifier.value(`system`)</td>\n",
    "    <td class=\"tg-0lax\"> The identifier for the data provider.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">data_catagory</td>\n",
    "    <td class=\"tg-7zrl\">Broad categorization of the contents of the data file.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">data_type</td>\n",
    "    <td class=\"tg-7zrl\">Specific content type of the data file.</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-7zrl\">file_format</td>\n",
    "    <td class=\"tg-7zrl\">Format of the data files.</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a4fd46e",
   "metadata": {},
   "source": [
    "## Files from a single endpoint (endpoint chaining)\n",
    "\n",
    "If you want all file formats and data types, but only from a specific endpoint, you can also filter the file results by chaining endpoints together. This will return all the files that match our search AND that are specifically from specimens:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2cfb2582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3604 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">50546</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   total : \u001b[1;36m50546\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_33cde_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_33cde_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_33cde_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_33cde_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_33cde_row0_col1\" class=\"data row0 col1\" >47498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_33cde_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_33cde_row1_col1\" class=\"data row1 col1\" >3048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_a862e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a862e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a862e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row0_col0\" class=\"data row0 col0\" >Simple Nucleotide Variation</td>\n",
       "      <td id=\"T_a862e_row0_col1\" class=\"data row0 col1\" >20481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row1_col0\" class=\"data row1 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_a862e_row1_col1\" class=\"data row1 col1\" >3629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row2_col0\" class=\"data row2 col0\" >Transcriptome Profiling</td>\n",
       "      <td id=\"T_a862e_row2_col1\" class=\"data row2 col1\" >3112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row3_col0\" class=\"data row3 col0\" >Copy Number Variation</td>\n",
       "      <td id=\"T_a862e_row3_col1\" class=\"data row3 col1\" >6909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row4_col0\" class=\"data row4 col0\" >Raw Mass Spectra</td>\n",
       "      <td id=\"T_a862e_row4_col1\" class=\"data row4 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row5_col0\" class=\"data row5 col0\" >Sequencing Reads</td>\n",
       "      <td id=\"T_a862e_row5_col1\" class=\"data row5 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row6_col0\" class=\"data row6 col0\" >Proteome Profiling</td>\n",
       "      <td id=\"T_a862e_row6_col1\" class=\"data row6 col1\" >679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row7_col0\" class=\"data row7 col0\" >DNA Methylation</td>\n",
       "      <td id=\"T_a862e_row7_col1\" class=\"data row7 col1\" >3339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row8_col0\" class=\"data row8 col0\" >Peptide Spectral Matches</td>\n",
       "      <td id=\"T_a862e_row8_col1\" class=\"data row8 col1\" >1524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row9_col0\" class=\"data row9 col0\" >Structural Variation</td>\n",
       "      <td id=\"T_a862e_row9_col1\" class=\"data row9 col1\" >3160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row10_col0\" class=\"data row10 col0\" >Processed Mass Spectra</td>\n",
       "      <td id=\"T_a862e_row10_col1\" class=\"data row10 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row11_col0\" class=\"data row11 col0\" >Somatic Structural Variation</td>\n",
       "      <td id=\"T_a862e_row11_col1\" class=\"data row11 col1\" >299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a862e_row12_col0\" class=\"data row12 col0\" >Clinical</td>\n",
       "      <td id=\"T_a862e_row12_col1\" class=\"data row12 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_0c803_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_0c803_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_0c803_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row0_col0\" class=\"data row0 col0\" >VCF</td>\n",
       "      <td id=\"T_0c803_row0_col1\" class=\"data row0 col1\" >12255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row1_col0\" class=\"data row1 col0\" >MEX</td>\n",
       "      <td id=\"T_0c803_row1_col1\" class=\"data row1 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row2_col0\" class=\"data row2 col0\" >IDAT</td>\n",
       "      <td id=\"T_0c803_row2_col1\" class=\"data row2 col1\" >2226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row3_col0\" class=\"data row3 col0\" >TSV</td>\n",
       "      <td id=\"T_0c803_row3_col1\" class=\"data row3 col1\" >4335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row4_col0\" class=\"data row4 col0\" >BAM</td>\n",
       "      <td id=\"T_0c803_row4_col1\" class=\"data row4 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row5_col0\" class=\"data row5 col0\" >mzML</td>\n",
       "      <td id=\"T_0c803_row5_col1\" class=\"data row5 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row6_col0\" class=\"data row6 col0\" >SVS</td>\n",
       "      <td id=\"T_0c803_row6_col1\" class=\"data row6 col1\" >3629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row7_col0\" class=\"data row7 col0\" >MAF</td>\n",
       "      <td id=\"T_0c803_row7_col1\" class=\"data row7 col1\" >8375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row8_col0\" class=\"data row8 col0\" >TXT</td>\n",
       "      <td id=\"T_0c803_row8_col1\" class=\"data row8 col1\" >8848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row9_col0\" class=\"data row9 col0\" >mzIdentML</td>\n",
       "      <td id=\"T_0c803_row9_col1\" class=\"data row9 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row10_col0\" class=\"data row10 col0\" >BEDPE</td>\n",
       "      <td id=\"T_0c803_row10_col1\" class=\"data row10 col1\" >1886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row11_col0\" class=\"data row11 col0\" >tsv</td>\n",
       "      <td id=\"T_0c803_row11_col1\" class=\"data row11 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row12_col0\" class=\"data row12 col0\" >vendor-specific</td>\n",
       "      <td id=\"T_0c803_row12_col1\" class=\"data row12 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row13_col0\" class=\"data row13 col0\" >HDF5</td>\n",
       "      <td id=\"T_0c803_row13_col1\" class=\"data row13 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0c803_row14_col0\" class=\"data row14 col0\" >PDF</td>\n",
       "      <td id=\"T_0c803_row14_col1\" class=\"data row14 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_03438_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_03438_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_03438_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row0_col0\" class=\"data row0 col0\" >Protein Expression Quantification</td>\n",
       "      <td id=\"T_03438_row0_col1\" class=\"data row0 col1\" >679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row1_col0\" class=\"data row1 col0\" >Slide Image</td>\n",
       "      <td id=\"T_03438_row1_col1\" class=\"data row1 col1\" >3629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row2_col0\" class=\"data row2 col0\" >Aligned Reads</td>\n",
       "      <td id=\"T_03438_row2_col1\" class=\"data row2 col1\" >5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row3_col0\" class=\"data row3 col0\" >miRNA Expression Quantification</td>\n",
       "      <td id=\"T_03438_row3_col1\" class=\"data row3 col1\" >643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row4_col0\" class=\"data row4 col0\" >Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_03438_row4_col1\" class=\"data row4 col1\" >11808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row5_col0\" class=\"data row5 col0\" >Proprietary</td>\n",
       "      <td id=\"T_03438_row5_col1\" class=\"data row5 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row6_col0\" class=\"data row6 col0\" >Aggregated Somatic Mutation</td>\n",
       "      <td id=\"T_03438_row6_col1\" class=\"data row6 col1\" >1144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row7_col0\" class=\"data row7 col0\" >Raw Simple Somatic Mutation</td>\n",
       "      <td id=\"T_03438_row7_col1\" class=\"data row7 col1\" >6202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row8_col0\" class=\"data row8 col0\" >Masked Copy Number Segment</td>\n",
       "      <td id=\"T_03438_row8_col1\" class=\"data row8 col1\" >2185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row9_col0\" class=\"data row9 col0\" >Masked Intensities</td>\n",
       "      <td id=\"T_03438_row9_col1\" class=\"data row9 col1\" >2226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row10_col0\" class=\"data row10 col0\" >Structural Rearrangement</td>\n",
       "      <td id=\"T_03438_row10_col1\" class=\"data row10 col1\" >298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row11_col0\" class=\"data row11 col0\" >Open Standard</td>\n",
       "      <td id=\"T_03438_row11_col1\" class=\"data row11 col1\" >1524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row12_col0\" class=\"data row12 col0\" >Transcript Fusion</td>\n",
       "      <td id=\"T_03438_row12_col1\" class=\"data row12 col1\" >3161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row13_col0\" class=\"data row13 col0\" >Gene Expression Quantification</td>\n",
       "      <td id=\"T_03438_row13_col1\" class=\"data row13 col1\" >904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row14_col0\" class=\"data row14 col0\" >Allele-specific Copy Number Segment</td>\n",
       "      <td id=\"T_03438_row14_col1\" class=\"data row14 col1\" >1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row15_col0\" class=\"data row15 col0\" >Masked Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_03438_row15_col1\" class=\"data row15 col1\" >183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row16_col0\" class=\"data row16 col0\" >Copy Number Segment</td>\n",
       "      <td id=\"T_03438_row16_col1\" class=\"data row16 col1\" >2334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row17_col0\" class=\"data row17 col0\" >Gene Level Copy Number</td>\n",
       "      <td id=\"T_03438_row17_col1\" class=\"data row17 col1\" >1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row18_col0\" class=\"data row18 col0\" >Text</td>\n",
       "      <td id=\"T_03438_row18_col1\" class=\"data row18 col1\" >762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row19_col0\" class=\"data row19 col0\" >Methylation Beta Value</td>\n",
       "      <td id=\"T_03438_row19_col1\" class=\"data row19 col1\" >1113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row20_col0\" class=\"data row20 col0\" >Splice Junction Quantification</td>\n",
       "      <td id=\"T_03438_row20_col1\" class=\"data row20 col1\" >868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row21_col0\" class=\"data row21 col0\" >Masked Somatic Mutation</td>\n",
       "      <td id=\"T_03438_row21_col1\" class=\"data row21 col1\" >1144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row22_col0\" class=\"data row22 col0\" >Isoform Expression Quantification</td>\n",
       "      <td id=\"T_03438_row22_col1\" class=\"data row22 col1\" >643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row23_col0\" class=\"data row23 col0\" >Single Cell Analysis</td>\n",
       "      <td id=\"T_03438_row23_col1\" class=\"data row23 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row24_col0\" class=\"data row24 col0\" >Gene Level Copy Number Scores</td>\n",
       "      <td id=\"T_03438_row24_col1\" class=\"data row24 col1\" >99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row25_col0\" class=\"data row25 col0\" >Differential Gene Expression</td>\n",
       "      <td id=\"T_03438_row25_col1\" class=\"data row25 col1\" >18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_03438_row26_col0\" class=\"data row26 col0\" >Pathology Report</td>\n",
       "      <td id=\"T_03438_row26_col1\" class=\"data row26 col1\" >5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.specimen.file.count.run()\n",
    "\n",
    "#   total : 50546    \n",
    "# system\tcount\n",
    "# GDC\t47498\n",
    "# PDC\t3048 \n",
    "# data_category\tcount\n",
    "# Simple Nucleotide Variation\t20481\n",
    "# Biospecimen\t3629\n",
    "# Transcriptome Profiling\t3112\n",
    "# Copy Number Variation\t6909\n",
    "# Raw Mass Spectra\t762\n",
    "# Sequencing Reads\t5885\n",
    "# Proteome Profiling\t679\n",
    "# DNA Methylation\t3339\n",
    "# Peptide Spectral Matches\t1524\n",
    "# Structural Variation\t3160\n",
    "# Processed Mass Spectra\t762\n",
    "# Somatic Structural Variation\t299\n",
    "# Clinical\t5 \n",
    "# file_format\tcount\n",
    "# VCF\t12255\n",
    "# MEX\t36\n",
    "# IDAT\t2226\n",
    "# TSV\t4335\n",
    "# BAM\t5885\n",
    "# mzML\t762\n",
    "# SVS\t3629\n",
    "# MAF\t8375\n",
    "# TXT\t8848\n",
    "# mzIdentML\t762\n",
    "# BEDPE\t1886\n",
    "# tsv\t762\n",
    "# vendor-specific\t762\n",
    "# HDF5\t18\n",
    "# PDF\t5 \n",
    "# data_type\tcount\n",
    "# Protein Expression Quantification\t679\n",
    "# Slide Image\t3629\n",
    "# Aligned Reads\t5885\n",
    "# miRNA Expression Quantification\t643\n",
    "# Annotated Somatic Mutation\t11808\n",
    "# Proprietary\t762\n",
    "# Aggregated Somatic Mutation\t1144\n",
    "# Raw Simple Somatic Mutation\t6202\n",
    "# Masked Copy Number Segment\t2185\n",
    "# Masked Intensities\t2226\n",
    "# Structural Rearrangement\t298\n",
    "# Open Standard\t1524\n",
    "# Transcript Fusion\t3161\n",
    "# Gene Expression Quantification\t904\n",
    "# Allele-specific Copy Number Segment\t1071\n",
    "# Masked Annotated Somatic Mutation\t183\n",
    "# Copy Number Segment\t2334\n",
    "# Gene Level Copy Number\t1220\n",
    "# Text\t762\n",
    "# Methylation Beta Value\t1113\n",
    "# Splice Junction Quantification\t868\n",
    "# Masked Somatic Mutation\t1144\n",
    "# Isoform Expression Quantification\t643\n",
    "# Single Cell Analysis\t36\n",
    "# Gene Level Copy Number Scores\t99\n",
    "# Differential Gene Expression\t18\n",
    "# Pathology Report\t5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6994e312",
   "metadata": {},
   "source": [
    "Learn more about chaining endpoints in the [Chaining endpoints](\"../AdvancedSearch-Chaining\") notebook."
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
