{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "614a478f",
   "metadata": {},
   "source": [
    "# Build a cohort tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8004e87",
   "metadata": {},
   "source": [
    "**Example use case:** \n",
    "\n",
    "<img src=\"../images/julia.png\" alt=\"alt_text\" align=\"left\"\n",
    "\twidth=\"150\" height=\"150\" />\n",
    "Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0cadd7d",
   "metadata": {},
   "source": [
    "Before Julia does any work, she needs to import several functions from cdapython:\n",
    "\n",
    "- `Q` and `query` which power the search\n",
    "- `columns` which lets us view entity field names\n",
    "- `unique_terms` which lets view entity field contents\n",
    "\n",
    "She also asks cdapython to report it's version so she can be sure she's using the one she means to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5265d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.9\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "import pandas as pd \n",
    "print(cdapython.__version__)\n",
    "Q.set_host_url(\"http://35.192.60.10:8080/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75eef23e",
   "metadata": {},
   "source": [
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A specific, unique individual: for e.g. a single human. When consent allows, a given entity will have a single subject ID that can be connected to all their studies and data across all datasets</li>\n",
    "<li><b>researchsubject:</b> a person/plant/animal/microbe within a given study. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> a tissue sample taken from a given subject, or a portion of the original sample. A given specimen will have only a single subject ID and a single research subject ID</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> Information about what medical diagnosis a researchsubject has</li>\n",
    "<li><b>treatment:</b> Information about what medical treatment(s) were performed for a given diagnosis</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "391bc9a7",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef0dd8e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['File.id',\n",
       " 'File.identifier.system',\n",
       " 'File.identifier.value',\n",
       " 'File.label',\n",
       " 'File.data_category',\n",
       " 'File.data_type',\n",
       " 'File.file_format',\n",
       " 'File.associated_project',\n",
       " 'File.drs_uri',\n",
       " 'File.byte_size',\n",
       " 'File.checksum',\n",
       " 'File.data_modality',\n",
       " 'File.imaging_modality',\n",
       " 'File.dbgap_accession_number',\n",
       " 'id',\n",
       " 'identifier.system',\n",
       " 'identifier.value',\n",
       " 'species',\n",
       " 'sex',\n",
       " 'race',\n",
       " 'ethnicity',\n",
       " 'days_to_birth',\n",
       " 'subject_associated_project',\n",
       " 'vital_status',\n",
       " 'age_at_death',\n",
       " 'cause_of_death',\n",
       " 'ResearchSubject.id',\n",
       " 'ResearchSubject.identifier.system',\n",
       " 'ResearchSubject.identifier.value',\n",
       " 'ResearchSubject.member_of_research_project',\n",
       " 'ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles',\n",
       " 'ResearchSubject.Specimen.id',\n",
       " 'ResearchSubject.Specimen.identifier.system',\n",
       " 'ResearchSubject.Specimen.identifier.value',\n",
       " 'ResearchSubject.Specimen.associated_project',\n",
       " 'ResearchSubject.Specimen.age_at_collection',\n",
       " 'ResearchSubject.Specimen.primary_disease_type',\n",
       " 'ResearchSubject.Specimen.anatomical_site',\n",
       " 'ResearchSubject.Specimen.source_material_type',\n",
       " 'ResearchSubject.Specimen.specimen_type',\n",
       " 'ResearchSubject.Specimen.derived_from_specimen',\n",
       " 'ResearchSubject.Specimen.derived_from_subject']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05eba2",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "536970c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list(filters=\"diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63b4cf0",
   "metadata": {},
   "source": [
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "\n",
    "To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retreiving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.\n",
    "    \n",
    "</div>\n",
    "\n",
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Since Julia is interested specificially in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4527dde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Brain',\n",
       " 'Cervix',\n",
       " 'Head - Face Or Neck, Nos',\n",
       " 'Lymph Node(s) Paraaortic',\n",
       " 'Other',\n",
       " 'Pelvis',\n",
       " 'Spine',\n",
       " 'Unknown']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\").to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "740e5955",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Abdomen',\n",
       " 'Abdomen, Mediastinum',\n",
       " 'Adrenal Glands',\n",
       " 'Adrenal gland',\n",
       " 'Anus and anal canal',\n",
       " 'Base of tongue',\n",
       " 'Bile Duct',\n",
       " 'Bladder',\n",
       " 'Bones, joints and articular cartilage of limbs',\n",
       " 'Bones, joints and articular cartilage of other and unspecified sites',\n",
       " 'Brain',\n",
       " 'Breast',\n",
       " 'Bronchus and lung',\n",
       " 'Cervix',\n",
       " 'Cervix uteri',\n",
       " 'Chest',\n",
       " 'Chest-Abdomen-Pelvis, Leg, TSpine',\n",
       " 'Colon',\n",
       " 'Connective, subcutaneous and other soft tissues',\n",
       " 'Corpus uteri',\n",
       " 'Ear',\n",
       " 'Esophagus',\n",
       " 'Extremities',\n",
       " 'Eye and adnexa',\n",
       " 'Floor of mouth',\n",
       " 'Gallbladder',\n",
       " 'Gum',\n",
       " 'Head',\n",
       " 'Head and Neck',\n",
       " 'Head-Neck',\n",
       " 'Heart, mediastinum, and pleura',\n",
       " 'Hematopoietic and reticuloendothelial systems',\n",
       " 'Hypopharynx',\n",
       " 'Intraocular',\n",
       " 'Kidney',\n",
       " 'Larynx',\n",
       " 'Lip',\n",
       " 'Liver',\n",
       " 'Liver and intrahepatic bile ducts',\n",
       " 'Lung',\n",
       " 'Lung Phantom',\n",
       " 'Lymph nodes',\n",
       " 'Marrow, Blood',\n",
       " 'Meninges',\n",
       " 'Mesothelium',\n",
       " 'Nasal cavity and middle ear',\n",
       " 'Nasopharynx',\n",
       " 'Not Reported',\n",
       " 'Oropharynx',\n",
       " 'Other and ill-defined digestive organs',\n",
       " 'Other and ill-defined sites',\n",
       " 'Other and ill-defined sites in lip, oral cavity and pharynx',\n",
       " 'Other and ill-defined sites within respiratory system and intrathoracic organs',\n",
       " 'Other and unspecified female genital organs',\n",
       " 'Other and unspecified major salivary glands',\n",
       " 'Other and unspecified male genital organs',\n",
       " 'Other and unspecified parts of biliary tract',\n",
       " 'Other and unspecified parts of mouth',\n",
       " 'Other and unspecified parts of tongue',\n",
       " 'Other and unspecified urinary organs',\n",
       " 'Other endocrine glands and related structures',\n",
       " 'Ovary',\n",
       " 'Palate',\n",
       " 'Pancreas',\n",
       " 'Pancreas ',\n",
       " 'Pelvis, Prostate, Anus',\n",
       " 'Penis',\n",
       " 'Peripheral nerves and autonomic nervous system',\n",
       " 'Phantom',\n",
       " 'Prostate',\n",
       " 'Prostate gland',\n",
       " 'Rectosigmoid junction',\n",
       " 'Rectum',\n",
       " 'Renal pelvis',\n",
       " 'Retroperitoneum and peritoneum',\n",
       " 'Skin',\n",
       " 'Small intestine',\n",
       " 'Spinal cord, cranial nerves, and other parts of central nervous system',\n",
       " 'Stomach',\n",
       " 'Testicles',\n",
       " 'Testis',\n",
       " 'Thymus',\n",
       " 'Thyroid',\n",
       " 'Thyroid gland',\n",
       " 'Tonsil',\n",
       " 'Trachea',\n",
       " 'Unknown',\n",
       " 'Ureter',\n",
       " 'Uterus',\n",
       " 'Uterus, NOS',\n",
       " 'Vagina',\n",
       " 'Various',\n",
       " 'Various (11 locations)',\n",
       " 'Vulva']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b005036b",
   "metadata": {},
   "source": [
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "    \n",
    "CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e6b8dc",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia sees that \"treatment_anatomic_site\" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the \"primary_diagnosis_site\" results. As she was initially looking for \"uterine\", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the \"ResearchSubject.primary_diagnosis_site\" for 'uter' as that should cover all variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31064125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix uteri', 'Corpus uteri', 'Uterus', 'Uterus, NOS']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"uter\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9311a49e",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Just to be sure, Julia also searches for any other instances of \"cervix\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2038a8cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix', 'Cervix uteri']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"cerv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c4de58",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the \"treatment_anatomic_site\", only one term is of interest, so she uses the `=` operator to get only exact matches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "951fcc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = \"Cervix\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cb5f72",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "However, for \"primary_diagnosis_site\", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36cfd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Dsite = Q('ResearchSubject.primary_diagnosis_site = \"%uter%\" OR ResearchSubject.primary_diagnosis_site = \"%cerv%\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349af6f2",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia adds her two queries together into one large one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9f5f9e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA = Tsite.OR(Dsite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f5cb55",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "355b1706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3452 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">specimen_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">40766</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "specimen_count : \u001b[1;36m40766\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">treatment_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3045</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "treatment_count : \u001b[1;36m3045\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">diagnosis_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3683</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "diagnosis_count : \u001b[1;36m3683\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">researchsubject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4867</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "researchsubject_count : \u001b[1;36m4867\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">subject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3740</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "subject_count : \u001b[1;36m3740\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ce25fc",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "55b0cdeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3613 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4867</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m4867\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">324216</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m324216\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_2c42a_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_2c42a_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_2c42a_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_2c42a_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_2c42a_row0_col1\" class=\"data row0 col1\" >3589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_2c42a_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_2c42a_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_2c42a_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_2c42a_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_a777d_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_a777d_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a777d_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row0_col0\" class=\"data row0 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_a777d_row0_col1\" class=\"data row0 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row1_col0\" class=\"data row1 col0\" >Adenomas and Adenocarcinomas</td>\n",
       "      <td id=\"T_a777d_row1_col1\" class=\"data row1 col1\" >1671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row2_col0\" class=\"data row2 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_a777d_row2_col1\" class=\"data row2 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row3_col0\" class=\"data row3 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_a777d_row3_col1\" class=\"data row3 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row4_col0\" class=\"data row4 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_a777d_row4_col1\" class=\"data row4 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row5_col0\" class=\"data row5 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_a777d_row5_col1\" class=\"data row5 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row6_col0\" class=\"data row6 col0\" >None</td>\n",
       "      <td id=\"T_a777d_row6_col1\" class=\"data row6 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row7_col0\" class=\"data row7 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_a777d_row7_col1\" class=\"data row7 col1\" >187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row8_col0\" class=\"data row8 col0\" >Not Reported</td>\n",
       "      <td id=\"T_a777d_row8_col1\" class=\"data row8 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row9_col0\" class=\"data row9 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_a777d_row9_col1\" class=\"data row9 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row10_col0\" class=\"data row10 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_a777d_row10_col1\" class=\"data row10 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row11_col0\" class=\"data row11 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_a777d_row11_col1\" class=\"data row11 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row12_col0\" class=\"data row12 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_a777d_row12_col1\" class=\"data row12 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row13_col0\" class=\"data row13 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_a777d_row13_col1\" class=\"data row13 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_a777d_row14_col0\" class=\"data row14 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_a777d_row14_col1\" class=\"data row14 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_ca5e1_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_ca5e1_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_ca5e1_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_ca5e1_row0_col0\" class=\"data row0 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_ca5e1_row0_col1\" class=\"data row0 col1\" >780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ca5e1_row1_col0\" class=\"data row1 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_ca5e1_row1_col1\" class=\"data row1 col1\" >1998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ca5e1_row2_col0\" class=\"data row2 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_ca5e1_row2_col1\" class=\"data row2 col1\" >915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ca5e1_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_ca5e1_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_ca5e1_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_ca5e1_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a323e2",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0d526198",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3529 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3196</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m3196\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">297923</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m297923\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_797a3_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_797a3_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_797a3_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_797a3_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_797a3_row0_col1\" class=\"data row0 col1\" >1918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_797a3_row1_col0\" class=\"data row1 col0\" >PDC</td>\n",
       "      <td id=\"T_797a3_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_797a3_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_797a3_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_9e0ec_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_9e0ec_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_9e0ec_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row0_col0\" class=\"data row0 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row0_col1\" class=\"data row0 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row1_col0\" class=\"data row1 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row1_col1\" class=\"data row1 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row2_col0\" class=\"data row2 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row2_col1\" class=\"data row2 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row3_col0\" class=\"data row3 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_9e0ec_row3_col1\" class=\"data row3 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row4_col0\" class=\"data row4 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row4_col1\" class=\"data row4 col1\" >187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row5_col0\" class=\"data row5 col0\" >None</td>\n",
       "      <td id=\"T_9e0ec_row5_col1\" class=\"data row5 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row6_col0\" class=\"data row6 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_9e0ec_row6_col1\" class=\"data row6 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row7_col0\" class=\"data row7 col0\" >Not Reported</td>\n",
       "      <td id=\"T_9e0ec_row7_col1\" class=\"data row7 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row8_col0\" class=\"data row8 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_9e0ec_row8_col1\" class=\"data row8 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row9_col0\" class=\"data row9 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_9e0ec_row9_col1\" class=\"data row9 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row10_col0\" class=\"data row10 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row10_col1\" class=\"data row10 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row11_col0\" class=\"data row11 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row11_col1\" class=\"data row11 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row12_col0\" class=\"data row12 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_9e0ec_row12_col1\" class=\"data row12 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9e0ec_row13_col0\" class=\"data row13 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_9e0ec_row13_col1\" class=\"data row13 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_0b5c8_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_0b5c8_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_0b5c8_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_0b5c8_row0_col0\" class=\"data row0 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_0b5c8_row0_col1\" class=\"data row0 col1\" >688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0b5c8_row1_col0\" class=\"data row1 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_0b5c8_row1_col1\" class=\"data row1 col1\" >373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0b5c8_row2_col0\" class=\"data row2 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_0b5c8_row2_col1\" class=\"data row2 col1\" >961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0b5c8_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_0b5c8_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_0b5c8_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_0b5c8_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Noadeno = Q('ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"')\n",
    "\n",
    "NoAdenoData = ALLDATA.AND(Noadeno)\n",
    "\n",
    "NoAdenoData.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a0191d",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d186b837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3634 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>183e8d87-3476-46bc-a989-ed4b09809227</td>\n",
       "      <td>[{'system': 'GDC', 'value': '183e8d87-3476-46b...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-C5-A8XH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310a3ea9-8c9e-4c52-8bf7-ab53933bb623</td>\n",
       "      <td>[{'system': 'GDC', 'value': '310a3ea9-8c9e-4c5...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0015737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4e0d653d-21f0-468d-91e6-bca55dcddfb0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4e0d653d-21f0-468...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6b74c07a-9932-4a80-8296-9db08d0c6964</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6b74c07a-9932-4a8...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-EX-A1H5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6bcba66e-2c17-4134-afad-c8ee85d4e1b9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6bcba66e-2c17-413...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>C3L-00963__cptac_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-00963'}]</td>\n",
       "      <td>cptac_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>C3L-00963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>C3N-00753__cptac_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-00753'}]</td>\n",
       "      <td>cptac_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>C3N-00753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-AP-A1DO__tcga_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-AP-A1DO'}]</td>\n",
       "      <td>tcga_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-AP-A1DO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-AX-A05W__tcga_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-AX-A05W'}]</td>\n",
       "      <td>tcga_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-AX-A05W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-AX-A2H8__tcga_ucec</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-AX-A2H8'}]</td>\n",
       "      <td>tcga_ucec</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-AX-A2H8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   183e8d87-3476-46bc-a989-ed4b09809227   \n",
       "1   310a3ea9-8c9e-4c52-8bf7-ab53933bb623   \n",
       "2   4e0d653d-21f0-468d-91e6-bca55dcddfb0   \n",
       "3   6b74c07a-9932-4a80-8296-9db08d0c6964   \n",
       "4   6bcba66e-2c17-4134-afad-c8ee85d4e1b9   \n",
       "..                                   ...   \n",
       "95                 C3L-00963__cptac_ucec   \n",
       "96                 C3N-00753__cptac_ucec   \n",
       "97               TCGA-AP-A1DO__tcga_ucec   \n",
       "98               TCGA-AX-A05W__tcga_ucec   \n",
       "99               TCGA-AX-A2H8__tcga_ucec   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '183e8d87-3476-46b...   \n",
       "1   [{'system': 'GDC', 'value': '310a3ea9-8c9e-4c5...   \n",
       "2   [{'system': 'GDC', 'value': '4e0d653d-21f0-468...   \n",
       "3   [{'system': 'GDC', 'value': '6b74c07a-9932-4a8...   \n",
       "4   [{'system': 'GDC', 'value': '6bcba66e-2c17-413...   \n",
       "..                                                ...   \n",
       "95          [{'system': 'IDC', 'value': 'C3L-00963'}]   \n",
       "96          [{'system': 'IDC', 'value': 'C3N-00753'}]   \n",
       "97       [{'system': 'IDC', 'value': 'TCGA-AP-A1DO'}]   \n",
       "98       [{'system': 'IDC', 'value': 'TCGA-AX-A05W'}]   \n",
       "99       [{'system': 'IDC', 'value': 'TCGA-AX-A2H8'}]   \n",
       "\n",
       "   member_of_research_project          primary_diagnosis_condition  \\\n",
       "0                   TCGA-CESC              Squamous Cell Neoplasms   \n",
       "1                   GENIE-MSK  Complex Mixed and Stromal Neoplasms   \n",
       "2               CGCI-HTMCP-CC              Squamous Cell Neoplasms   \n",
       "3                   TCGA-CESC              Squamous Cell Neoplasms   \n",
       "4               CGCI-HTMCP-CC              Squamous Cell Neoplasms   \n",
       "..                        ...                                  ...   \n",
       "95                 cptac_ucec                                 None   \n",
       "96                 cptac_ucec                                 None   \n",
       "97                  tcga_ucec                                 None   \n",
       "98                  tcga_ucec                                 None   \n",
       "99                  tcga_ucec                                 None   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0            Cervix uteri         TCGA-C5-A8XH  \n",
       "1             Uterus, NOS  GENIE-MSK-P-0015737  \n",
       "2            Cervix uteri    HTMCP-03-06-02239  \n",
       "3            Cervix uteri         TCGA-EX-A1H5  \n",
       "4            Cervix uteri    HTMCP-03-06-02409  \n",
       "..                    ...                  ...  \n",
       "95                 Uterus            C3L-00963  \n",
       "96                 Uterus            C3N-00753  \n",
       "97                 Uterus         TCGA-AP-A1DO  \n",
       "98                 Uterus         TCGA-AX-A05W  \n",
       "99                 Uterus         TCGA-AX-A2H8  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.run().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086697b3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a research study, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjectâ€™s privacy. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this researchsubject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the researchsubject had there</li>\n",
    "  <li><b>member_of_research_project:</b> The name of the study/project that the subject particpated in</li>\n",
    "  <li><b>primary_diagnosis_condition:</b> The cancer, disease or other condition under study</li>\n",
    "  <li><b>primary_diagnosis_site:</b> The primary_disease_site that qualifies the researchsubject for the research_project</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8d0f5e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3470 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD1430</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD1430'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD3861</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD3861'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-05848</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-05848'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3N-00323</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'C3N-00323'}, {'sy...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-28627.0</td>\n",
       "      <td>[CPTAC3-Discovery, CPTAC-3, cptac_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-DFCI-006487</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-006487'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22645.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GENIE-VICC-575150</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-VICC-575150'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27028.0</td>\n",
       "      <td>[GENIE-VICC]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>HTMCP-03-06-02222</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'HTMCP-03-06-02222'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CGCI-HTMCP-CC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-AJ-A8CT</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AJ-A8CT'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-21515.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-D1-A175</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-D1-A175'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17802.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-DS-A1OD</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DS-A1OD'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17954.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id                                         identifier  \\\n",
       "0              AD1430             [{'system': 'GDC', 'value': 'AD1430'}]   \n",
       "1              AD3861             [{'system': 'GDC', 'value': 'AD3861'}]   \n",
       "2           C3L-05848          [{'system': 'IDC', 'value': 'C3L-05848'}]   \n",
       "3           C3N-00323  [{'system': 'GDC', 'value': 'C3N-00323'}, {'sy...   \n",
       "4   GENIE-DFCI-006487  [{'system': 'GDC', 'value': 'GENIE-DFCI-006487'}]   \n",
       "..                ...                                                ...   \n",
       "95  GENIE-VICC-575150  [{'system': 'GDC', 'value': 'GENIE-VICC-575150'}]   \n",
       "96  HTMCP-03-06-02222  [{'system': 'GDC', 'value': 'HTMCP-03-06-02222'}]   \n",
       "97       TCGA-AJ-A8CT  [{'system': 'GDC', 'value': 'TCGA-AJ-A8CT'}, {...   \n",
       "98       TCGA-D1-A175  [{'system': 'GDC', 'value': 'TCGA-D1-A175'}, {...   \n",
       "99       TCGA-DS-A1OD  [{'system': 'GDC', 'value': 'TCGA-DS-A1OD'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "0   Homo sapiens  female               not reported            not reported   \n",
       "1   Homo sapiens  female               not reported            not reported   \n",
       "2   Homo sapiens    None                       None                    None   \n",
       "3   Homo sapiens  female               not reported            not reported   \n",
       "4   Homo sapiens  female  black or african american  not hispanic or latino   \n",
       "..           ...     ...                        ...                     ...   \n",
       "95  Homo sapiens  female                      white  not hispanic or latino   \n",
       "96  Homo sapiens  female  black or african american  not hispanic or latino   \n",
       "97  Homo sapiens  female  black or african american            not reported   \n",
       "98  Homo sapiens  female                      white  not hispanic or latino   \n",
       "99  Homo sapiens  female                      white  not hispanic or latino   \n",
       "\n",
       "    days_to_birth               subject_associated_project  vital_status  \\\n",
       "0             NaN                                  [FM-AD]  Not Reported   \n",
       "1             NaN                                  [FM-AD]  Not Reported   \n",
       "2             NaN                             [cptac_ucec]          None   \n",
       "3        -28627.0  [CPTAC3-Discovery, CPTAC-3, cptac_ucec]         Alive   \n",
       "4        -22645.0                             [GENIE-DFCI]  Not Reported   \n",
       "..            ...                                      ...           ...   \n",
       "95       -27028.0                             [GENIE-VICC]  Not Reported   \n",
       "96            NaN                          [CGCI-HTMCP-CC]         Alive   \n",
       "97       -21515.0                   [TCGA-UCEC, tcga_ucec]         Alive   \n",
       "98       -17802.0                   [TCGA-UCEC, tcga_ucec]         Alive   \n",
       "99       -17954.0                   [TCGA-CESC, tcga_cesc]         Alive   \n",
       "\n",
       "    age_at_death cause_of_death  \n",
       "0            NaN           None  \n",
       "1            NaN           None  \n",
       "2            NaN           None  \n",
       "3            NaN   Not Reported  \n",
       "4            NaN           None  \n",
       "..           ...            ...  \n",
       "95           NaN           None  \n",
       "96           NaN           None  \n",
       "97           NaN           None  \n",
       "98           NaN           None  \n",
       "99           NaN           None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.subject.run().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec76132",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A subject is a specific, unique individual: for e.g. a single human. When consent allows, a given entity will have a single subject ID that can be connected to all their studies and data across all datasets</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this subject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the subject had there</li>\n",
    "  <li><b>species:</b> The species of the subject</li>\n",
    "  <li><b>sex:</b> A reference to the biological sex of the donor organism. </li>\n",
    "  <li><b>race:</b> The race of the subject</li>\n",
    "  <li><b>ethnicity:</b> The ethnicity of the subject</li>\n",
    "  <li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days</li>\n",
    "  <li><b>subject_associated_project:</b> An embedded array of the names of projects (studies) the subject was part of</li>\n",
    "  <li><b>vital_status:</b> Whether the subject is alive</li>\n",
    "  <li><b>age_at_death:</b> The number of days after first enrollment that the subject died</li>\n",
    "  <li><b>cause_of_death:</b> The cause of death, if known</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "04e04136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3742 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>07734405-fc4b-4fa8-b6f9-ffc94124d330</td>\n",
       "      <td>[{'system': 'GDC', 'value': '07734405-fc4b-4fa...</td>\n",
       "      <td>979d6ef9-860d-4a0e-9e42-8eb2f2ddd34c.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:07734405-fc4b-4fa8-b6f9-ffc94124...</td>\n",
       "      <td>80149</td>\n",
       "      <td>51a07c367fac962816220c6ba724c948</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>1673c9c4-0adc-4b78-b1c0-4a0d7fee8e6f</td>\n",
       "      <td>f07f2302-2276-4d5c-b52b-0b900464d39c</td>\n",
       "      <td>HTMCP-03-06-02340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>415706a1-c58d-485c-aa3d-244a6760ed18</td>\n",
       "      <td>[{'system': 'GDC', 'value': '415706a1-c58d-485...</td>\n",
       "      <td>cfd92d9c-f6ed-4a01-89a4-83acfa32b794.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:415706a1-c58d-485c-aa3d-244a6760...</td>\n",
       "      <td>43688</td>\n",
       "      <td>6bd747ef331eed0a5c2b59892435865a</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>2c7cdfb5-07a9-403a-892e-1100dc1e595f</td>\n",
       "      <td>4b477751-4768-4d87-bd18-cd9a801acae3</td>\n",
       "      <td>HTMCP-03-06-02417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>513f94b7-ef01-4405-8b2d-7cd814403761</td>\n",
       "      <td>[{'system': 'GDC', 'value': '513f94b7-ef01-440...</td>\n",
       "      <td>b22b6c29-57ad-4497-ac67-bdc545d02afc.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:513f94b7-ef01-4405-8b2d-7cd81440...</td>\n",
       "      <td>121533</td>\n",
       "      <td>95e1d8e6dada42bfb2312894c15e019e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>b3467f65-dc04-44a3-a5ac-fa5dad498431</td>\n",
       "      <td>d86bf73c-b964-4fb6-8e49-c487ce2251a7</td>\n",
       "      <td>HTMCP-03-06-02068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5e65e42f-c1be-4f1e-a8aa-d7d8798aee26</td>\n",
       "      <td>[{'system': 'GDC', 'value': '5e65e42f-c1be-4f1...</td>\n",
       "      <td>66c6778c-0553-483c-a395-db3aad8fe31a.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:5e65e42f-c1be-4f1e-a8aa-d7d8798a...</td>\n",
       "      <td>54335</td>\n",
       "      <td>9e1badc4c154ef61b9ce658640b7ad23</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>fb2563c5-650f-4cd1-9f73-b7d89ad79a45</td>\n",
       "      <td>d02a8fb3-0f3e-475f-b2f7-27732908a284</td>\n",
       "      <td>HTMCP-03-06-02217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>707c1b5d-d343-48e0-afae-ee4913a03f7d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '707c1b5d-d343-48e...</td>\n",
       "      <td>46c46275-6659-44fb-814b-9738601495df.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:707c1b5d-d343-48e0-afae-ee4913a0...</td>\n",
       "      <td>205680</td>\n",
       "      <td>484700cf4e5e687cdf3a4583ae948fe7</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>a6e3ab1e-171e-4d7d-9d19-a6952bff7dcd</td>\n",
       "      <td>89c3d38f-8eda-4f62-bb9d-8d3b818c969f</td>\n",
       "      <td>HTMCP-03-06-02427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>513f94b7-ef01-4405-8b2d-7cd814403761</td>\n",
       "      <td>[{'system': 'GDC', 'value': '513f94b7-ef01-440...</td>\n",
       "      <td>b22b6c29-57ad-4497-ac67-bdc545d02afc.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:513f94b7-ef01-4405-8b2d-7cd81440...</td>\n",
       "      <td>121533</td>\n",
       "      <td>95e1d8e6dada42bfb2312894c15e019e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>0baf0fe2-e569-4d68-ad4a-507c5d813c4c</td>\n",
       "      <td>d86bf73c-b964-4fb6-8e49-c487ce2251a7</td>\n",
       "      <td>HTMCP-03-06-02068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>542a5563-4db3-42f5-bd07-97e94490d02f</td>\n",
       "      <td>[{'system': 'GDC', 'value': '542a5563-4db3-42f...</td>\n",
       "      <td>4410dfdd-a824-49c8-945f-6d4e2165fbab.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:542a5563-4db3-42f5-bd07-97e94490...</td>\n",
       "      <td>30249</td>\n",
       "      <td>3134d728acb4deb20681beb25a8efbe7</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>80044043-11ff-4fe6-a9eb-2a8f20efdb15</td>\n",
       "      <td>18bce2c3-5060-4282-b7a0-11ada9a3eebf</td>\n",
       "      <td>HTMCP-03-06-02333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>62374de1-b93f-4fa5-b248-d34bd09f82d8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '62374de1-b93f-4fa...</td>\n",
       "      <td>2cb4c39b-e508-47a6-8b58-3f6bf2033927.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:62374de1-b93f-4fa5-b248-d34bd09f...</td>\n",
       "      <td>109310</td>\n",
       "      <td>29671513576f3bcfd10ae24021294bca</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>3e4e9e49-35af-4fbf-a386-ed9908f98144</td>\n",
       "      <td>cef5dedc-fd37-4653-b4b8-6c9d325c990d</td>\n",
       "      <td>HTMCP-03-06-02054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>74e27c67-8e5f-4f63-b143-250424456bba</td>\n",
       "      <td>[{'system': 'GDC', 'value': '74e27c67-8e5f-4f6...</td>\n",
       "      <td>58fd73fa-8129-4777-b424-58d553d8f3da.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>BEDPE</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:74e27c67-8e5f-4f63-b143-25042445...</td>\n",
       "      <td>83204</td>\n",
       "      <td>41dc96b20a61d2d7c89323c3b5a3e99d</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>48a6ea79-e75e-429b-ba14-352019e6d120</td>\n",
       "      <td>a7814814-6a54-425f-a7cd-a58b3e510ea1</td>\n",
       "      <td>HTMCP-03-06-02063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>77550c39-97f8-4e49-9dac-93e679d9e3c5</td>\n",
       "      <td>[{'system': 'GDC', 'value': '77550c39-97f8-4e4...</td>\n",
       "      <td>3529837f-ea43-44b2-b3aa-6f83fed6f16d.wgs.BRASS...</td>\n",
       "      <td>Somatic Structural Variation</td>\n",
       "      <td>Structural Rearrangement</td>\n",
       "      <td>VCF</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>drs://dg.4DFC:77550c39-97f8-4e49-9dac-93e679d9...</td>\n",
       "      <td>44018</td>\n",
       "      <td>2c3a62b775c337129ceaeb57bbd6152c</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>3b6c0add-e026-457f-9c9a-14a72138b37c</td>\n",
       "      <td>1ea7047a-7ba0-487b-8aad-b0bdc5fe65d1</td>\n",
       "      <td>HTMCP-03-06-02262</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   07734405-fc4b-4fa8-b6f9-ffc94124d330   \n",
       "1   415706a1-c58d-485c-aa3d-244a6760ed18   \n",
       "2   513f94b7-ef01-4405-8b2d-7cd814403761   \n",
       "3   5e65e42f-c1be-4f1e-a8aa-d7d8798aee26   \n",
       "4   707c1b5d-d343-48e0-afae-ee4913a03f7d   \n",
       "..                                   ...   \n",
       "95  513f94b7-ef01-4405-8b2d-7cd814403761   \n",
       "96  542a5563-4db3-42f5-bd07-97e94490d02f   \n",
       "97  62374de1-b93f-4fa5-b248-d34bd09f82d8   \n",
       "98  74e27c67-8e5f-4f63-b143-250424456bba   \n",
       "99  77550c39-97f8-4e49-9dac-93e679d9e3c5   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '07734405-fc4b-4fa...   \n",
       "1   [{'system': 'GDC', 'value': '415706a1-c58d-485...   \n",
       "2   [{'system': 'GDC', 'value': '513f94b7-ef01-440...   \n",
       "3   [{'system': 'GDC', 'value': '5e65e42f-c1be-4f1...   \n",
       "4   [{'system': 'GDC', 'value': '707c1b5d-d343-48e...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '513f94b7-ef01-440...   \n",
       "96  [{'system': 'GDC', 'value': '542a5563-4db3-42f...   \n",
       "97  [{'system': 'GDC', 'value': '62374de1-b93f-4fa...   \n",
       "98  [{'system': 'GDC', 'value': '74e27c67-8e5f-4f6...   \n",
       "99  [{'system': 'GDC', 'value': '77550c39-97f8-4e4...   \n",
       "\n",
       "                                                label  \\\n",
       "0   979d6ef9-860d-4a0e-9e42-8eb2f2ddd34c.wgs.BRASS...   \n",
       "1   cfd92d9c-f6ed-4a01-89a4-83acfa32b794.wgs.BRASS...   \n",
       "2   b22b6c29-57ad-4497-ac67-bdc545d02afc.wgs.BRASS...   \n",
       "3   66c6778c-0553-483c-a395-db3aad8fe31a.wgs.BRASS...   \n",
       "4   46c46275-6659-44fb-814b-9738601495df.wgs.BRASS...   \n",
       "..                                                ...   \n",
       "95  b22b6c29-57ad-4497-ac67-bdc545d02afc.wgs.BRASS...   \n",
       "96  4410dfdd-a824-49c8-945f-6d4e2165fbab.wgs.BRASS...   \n",
       "97  2cb4c39b-e508-47a6-8b58-3f6bf2033927.wgs.BRASS...   \n",
       "98  58fd73fa-8129-4777-b424-58d553d8f3da.wgs.BRASS...   \n",
       "99  3529837f-ea43-44b2-b3aa-6f83fed6f16d.wgs.BRASS...   \n",
       "\n",
       "                   data_category                 data_type file_format  \\\n",
       "0   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "1   Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "2   Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "3   Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "4   Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "..                           ...                       ...         ...   \n",
       "95  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "96  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "97  Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "98  Somatic Structural Variation  Structural Rearrangement       BEDPE   \n",
       "99  Somatic Structural Variation  Structural Rearrangement         VCF   \n",
       "\n",
       "   associated_project                                            drs_uri  \\\n",
       "0       CGCI-HTMCP-CC  drs://dg.4DFC:07734405-fc4b-4fa8-b6f9-ffc94124...   \n",
       "1       CGCI-HTMCP-CC  drs://dg.4DFC:415706a1-c58d-485c-aa3d-244a6760...   \n",
       "2       CGCI-HTMCP-CC  drs://dg.4DFC:513f94b7-ef01-4405-8b2d-7cd81440...   \n",
       "3       CGCI-HTMCP-CC  drs://dg.4DFC:5e65e42f-c1be-4f1e-a8aa-d7d8798a...   \n",
       "4       CGCI-HTMCP-CC  drs://dg.4DFC:707c1b5d-d343-48e0-afae-ee4913a0...   \n",
       "..                ...                                                ...   \n",
       "95      CGCI-HTMCP-CC  drs://dg.4DFC:513f94b7-ef01-4405-8b2d-7cd81440...   \n",
       "96      CGCI-HTMCP-CC  drs://dg.4DFC:542a5563-4db3-42f5-bd07-97e94490...   \n",
       "97      CGCI-HTMCP-CC  drs://dg.4DFC:62374de1-b93f-4fa5-b248-d34bd09f...   \n",
       "98      CGCI-HTMCP-CC  drs://dg.4DFC:74e27c67-8e5f-4f63-b143-25042445...   \n",
       "99      CGCI-HTMCP-CC  drs://dg.4DFC:77550c39-97f8-4e49-9dac-93e679d9...   \n",
       "\n",
       "    byte_size                          checksum data_modality  \\\n",
       "0       80149  51a07c367fac962816220c6ba724c948       Genomic   \n",
       "1       43688  6bd747ef331eed0a5c2b59892435865a       Genomic   \n",
       "2      121533  95e1d8e6dada42bfb2312894c15e019e       Genomic   \n",
       "3       54335  9e1badc4c154ef61b9ce658640b7ad23       Genomic   \n",
       "4      205680  484700cf4e5e687cdf3a4583ae948fe7       Genomic   \n",
       "..        ...                               ...           ...   \n",
       "95     121533  95e1d8e6dada42bfb2312894c15e019e       Genomic   \n",
       "96      30249  3134d728acb4deb20681beb25a8efbe7       Genomic   \n",
       "97     109310  29671513576f3bcfd10ae24021294bca       Genomic   \n",
       "98      83204  41dc96b20a61d2d7c89323c3b5a3e99d       Genomic   \n",
       "99      44018  2c3a62b775c337129ceaeb57bbd6152c       Genomic   \n",
       "\n",
       "   imaging_modality dbgap_accession_number  \\\n",
       "0              None              phs000528   \n",
       "1              None              phs000528   \n",
       "2              None              phs000528   \n",
       "3              None              phs000528   \n",
       "4              None              phs000528   \n",
       "..              ...                    ...   \n",
       "95             None              phs000528   \n",
       "96             None              phs000528   \n",
       "97             None              phs000528   \n",
       "98             None              phs000528   \n",
       "99             None              phs000528   \n",
       "\n",
       "             researchsubject_specimen_id  \\\n",
       "0   1673c9c4-0adc-4b78-b1c0-4a0d7fee8e6f   \n",
       "1   2c7cdfb5-07a9-403a-892e-1100dc1e595f   \n",
       "2   b3467f65-dc04-44a3-a5ac-fa5dad498431   \n",
       "3   fb2563c5-650f-4cd1-9f73-b7d89ad79a45   \n",
       "4   a6e3ab1e-171e-4d7d-9d19-a6952bff7dcd   \n",
       "..                                   ...   \n",
       "95  0baf0fe2-e569-4d68-ad4a-507c5d813c4c   \n",
       "96  80044043-11ff-4fe6-a9eb-2a8f20efdb15   \n",
       "97  3e4e9e49-35af-4fbf-a386-ed9908f98144   \n",
       "98  48a6ea79-e75e-429b-ba14-352019e6d120   \n",
       "99  3b6c0add-e026-457f-9c9a-14a72138b37c   \n",
       "\n",
       "                      researchsubject_id         subject_id  \n",
       "0   f07f2302-2276-4d5c-b52b-0b900464d39c  HTMCP-03-06-02340  \n",
       "1   4b477751-4768-4d87-bd18-cd9a801acae3  HTMCP-03-06-02417  \n",
       "2   d86bf73c-b964-4fb6-8e49-c487ce2251a7  HTMCP-03-06-02068  \n",
       "3   d02a8fb3-0f3e-475f-b2f7-27732908a284  HTMCP-03-06-02217  \n",
       "4   89c3d38f-8eda-4f62-bb9d-8d3b818c969f  HTMCP-03-06-02427  \n",
       "..                                   ...                ...  \n",
       "95  d86bf73c-b964-4fb6-8e49-c487ce2251a7  HTMCP-03-06-02068  \n",
       "96  18bce2c3-5060-4282-b7a0-11ada9a3eebf  HTMCP-03-06-02333  \n",
       "97  cef5dedc-fd37-4653-b4b8-6c9d325c990d  HTMCP-03-06-02054  \n",
       "98  a7814814-6a54-425f-a7cd-a58b3e510ea1  HTMCP-03-06-02063  \n",
       "99  1ea7047a-7ba0-487b-8aad-b0bdc5fe65d1  HTMCP-03-06-02262  \n",
       "\n",
       "[100 rows x 16 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.file.run().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf9f2d3",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#c1f5ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this file</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the file had there</li>\n",
    "  <li><b>label:</b> The full name of the file</li>\n",
    "  <li><b>data_catagory:</b> A desecription of the kind of general kind data the file holds</li>\n",
    "  <li><b>data_type:</b> A more specific descripton of the data type</li>\n",
    "  <li><b>file_format:</b> String to identify the full file extension including compression extensions</li>\n",
    "  <li><b>associated_project:</b> The name the data center uses for the study this file was generated for</li>\n",
    "  <li><b>drs_uri:</b> A unique identifier that can be used to retreive this specific file from a server</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file</li>\n",
    "  <li><b>data_modality:</b> Describes the biological nature of the information gathered as the result of an activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\"</li>\n",
    "  <li><b>imaging_modality:</b> For files with the `data_modality` of \"Imaging\", a descriptor for the image type</li>\n",
    "  <li><b>dbgap_accession_number:</b> The project id number for this data on dbGaP</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6aadbe",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c2cec2bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3481 ms\n"
     ]
    }
   ],
   "source": [
    "researchsubs = NoAdenoData.researchsubject.run()\n",
    "rsdf = pd.DataFrame()\n",
    "for i in researchsubs.paginator(to_df=True):\n",
    "    rsdf = pd.concat([rsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a1258057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3367 ms\n"
     ]
    }
   ],
   "source": [
    "subs = NoAdenoData.subject.run()\n",
    "subsdf = pd.DataFrame()\n",
    "for i in subs.paginator(to_df=True):\n",
    "    subsdf = pd.concat([subsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "04cd73df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>183e8d87-3476-46bc-a989-ed4b09809227</td>\n",
       "      <td>[{'system': 'GDC', 'value': '183e8d87-3476-46b...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-C5-A8XH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310a3ea9-8c9e-4c52-8bf7-ab53933bb623</td>\n",
       "      <td>[{'system': 'GDC', 'value': '310a3ea9-8c9e-4c5...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0015737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4e0d653d-21f0-468d-91e6-bca55dcddfb0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4e0d653d-21f0-468...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6b74c07a-9932-4a80-8296-9db08d0c6964</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6b74c07a-9932-4a8...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-EX-A1H5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6bcba66e-2c17-4134-afad-c8ee85d4e1b9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6bcba66e-2c17-413...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>db6db2c4-210b-4f7a-a80c-2603957b7bc5</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'db6db2c4-210b-4f7...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>GENIE-DFCI-028657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>e87136c1-29a2-4e4c-aae8-05ce8b4bb090</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e87136c1-29a2-4e4...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Complex Epithelial Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>GENIE-DFCI-008895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>e9fa66f5-282c-416a-9dcf-46714ea12fcf</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e9fa66f5-282c-416...</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>TCGA-DI-A2QU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>f93ccd2d-3c77-4bd3-8cac-5203bd57735f</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f93ccd2d-3c77-4bd...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-Q1-A5R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>fca99890-34ff-49c1-a98f-931b0596aeab</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'fca99890-34ff-49c...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>GENIE-UHN-209072</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3196 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   183e8d87-3476-46bc-a989-ed4b09809227   \n",
       "1   310a3ea9-8c9e-4c52-8bf7-ab53933bb623   \n",
       "2   4e0d653d-21f0-468d-91e6-bca55dcddfb0   \n",
       "3   6b74c07a-9932-4a80-8296-9db08d0c6964   \n",
       "4   6bcba66e-2c17-4134-afad-c8ee85d4e1b9   \n",
       "..                                   ...   \n",
       "91  db6db2c4-210b-4f7a-a80c-2603957b7bc5   \n",
       "92  e87136c1-29a2-4e4c-aae8-05ce8b4bb090   \n",
       "93  e9fa66f5-282c-416a-9dcf-46714ea12fcf   \n",
       "94  f93ccd2d-3c77-4bd3-8cac-5203bd57735f   \n",
       "95  fca99890-34ff-49c1-a98f-931b0596aeab   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '183e8d87-3476-46b...   \n",
       "1   [{'system': 'GDC', 'value': '310a3ea9-8c9e-4c5...   \n",
       "2   [{'system': 'GDC', 'value': '4e0d653d-21f0-468...   \n",
       "3   [{'system': 'GDC', 'value': '6b74c07a-9932-4a8...   \n",
       "4   [{'system': 'GDC', 'value': '6bcba66e-2c17-413...   \n",
       "..                                                ...   \n",
       "91  [{'system': 'GDC', 'value': 'db6db2c4-210b-4f7...   \n",
       "92  [{'system': 'GDC', 'value': 'e87136c1-29a2-4e4...   \n",
       "93  [{'system': 'GDC', 'value': 'e9fa66f5-282c-416...   \n",
       "94  [{'system': 'GDC', 'value': 'f93ccd2d-3c77-4bd...   \n",
       "95  [{'system': 'GDC', 'value': 'fca99890-34ff-49c...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                   TCGA-CESC                Squamous Cell Neoplasms   \n",
       "1                   GENIE-MSK    Complex Mixed and Stromal Neoplasms   \n",
       "2               CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "3                   TCGA-CESC                Squamous Cell Neoplasms   \n",
       "4               CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "91                 GENIE-DFCI                Squamous Cell Neoplasms   \n",
       "92                 GENIE-DFCI           Complex Epithelial Neoplasms   \n",
       "93                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   \n",
       "94                  TCGA-CESC  Cystic, Mucinous and Serous Neoplasms   \n",
       "95                  GENIE-UHN              Epithelial Neoplasms, NOS   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0            Cervix uteri         TCGA-C5-A8XH  \n",
       "1             Uterus, NOS  GENIE-MSK-P-0015737  \n",
       "2            Cervix uteri    HTMCP-03-06-02239  \n",
       "3            Cervix uteri         TCGA-EX-A1H5  \n",
       "4            Cervix uteri    HTMCP-03-06-02409  \n",
       "..                    ...                  ...  \n",
       "91           Cervix uteri    GENIE-DFCI-028657  \n",
       "92           Cervix uteri    GENIE-DFCI-008895  \n",
       "93           Corpus uteri         TCGA-DI-A2QU  \n",
       "94           Cervix uteri         TCGA-Q1-A5R1  \n",
       "95           Corpus uteri     GENIE-UHN-209072  \n",
       "\n",
       "[3196 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rsdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "92a6f811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD1430</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD1430'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD3861</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD3861'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-05848</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-05848'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3N-00323</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'C3N-00323'}, {'sy...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-28627.0</td>\n",
       "      <td>[CPTAC3-Discovery, CPTAC-3, cptac_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-DFCI-006487</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-006487'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22645.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-EA-A410</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EA-A410'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-18720.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-EO-A2CH</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EO-A2CH'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-26966.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>590.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-FL-A3WE</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-FL-A3WE'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-N7-A59B</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N7-A59B'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-25360.0</td>\n",
       "      <td>[tcga_ucs, TCGA-UCS]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-VS-A9UM</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A9UM'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-14255.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>829.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2608 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id                                         identifier  \\\n",
       "0              AD1430             [{'system': 'GDC', 'value': 'AD1430'}]   \n",
       "1              AD3861             [{'system': 'GDC', 'value': 'AD3861'}]   \n",
       "2           C3L-05848          [{'system': 'IDC', 'value': 'C3L-05848'}]   \n",
       "3           C3N-00323  [{'system': 'GDC', 'value': 'C3N-00323'}, {'sy...   \n",
       "4   GENIE-DFCI-006487  [{'system': 'GDC', 'value': 'GENIE-DFCI-006487'}]   \n",
       "..                ...                                                ...   \n",
       "3        TCGA-EA-A410  [{'system': 'GDC', 'value': 'TCGA-EA-A410'}, {...   \n",
       "4        TCGA-EO-A2CH  [{'system': 'GDC', 'value': 'TCGA-EO-A2CH'}, {...   \n",
       "5        TCGA-FL-A3WE  [{'system': 'GDC', 'value': 'TCGA-FL-A3WE'}, {...   \n",
       "6        TCGA-N7-A59B  [{'system': 'GDC', 'value': 'TCGA-N7-A59B'}, {...   \n",
       "7        TCGA-VS-A9UM  [{'system': 'GDC', 'value': 'TCGA-VS-A9UM'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "0   Homo sapiens  female               not reported            not reported   \n",
       "1   Homo sapiens  female               not reported            not reported   \n",
       "2   Homo sapiens    None                       None                    None   \n",
       "3   Homo sapiens  female               not reported            not reported   \n",
       "4   Homo sapiens  female  black or african american  not hispanic or latino   \n",
       "..           ...     ...                        ...                     ...   \n",
       "3   Homo sapiens  female                      white  not hispanic or latino   \n",
       "4   Homo sapiens  female               not reported            not reported   \n",
       "5   Homo sapiens    None                       None                    None   \n",
       "6   Homo sapiens  female                      white  not hispanic or latino   \n",
       "7   Homo sapiens  female  black or african american            not reported   \n",
       "\n",
       "    days_to_birth               subject_associated_project  vital_status  \\\n",
       "0             NaN                                  [FM-AD]  Not Reported   \n",
       "1             NaN                                  [FM-AD]  Not Reported   \n",
       "2             NaN                             [cptac_ucec]          None   \n",
       "3        -28627.0  [CPTAC3-Discovery, CPTAC-3, cptac_ucec]         Alive   \n",
       "4        -22645.0                             [GENIE-DFCI]  Not Reported   \n",
       "..            ...                                      ...           ...   \n",
       "3        -18720.0                   [TCGA-CESC, tcga_cesc]         Alive   \n",
       "4        -26966.0                   [TCGA-UCEC, tcga_ucec]          Dead   \n",
       "5             NaN                   [TCGA-UCEC, tcga_ucec]          None   \n",
       "6        -25360.0                     [tcga_ucs, TCGA-UCS]         Alive   \n",
       "7        -14255.0                   [TCGA-CESC, tcga_cesc]          Dead   \n",
       "\n",
       "    age_at_death cause_of_death  \n",
       "0            NaN           None  \n",
       "1            NaN           None  \n",
       "2            NaN           None  \n",
       "3            NaN   Not Reported  \n",
       "4            NaN           None  \n",
       "..           ...            ...  \n",
       "3            NaN           None  \n",
       "4          590.0           None  \n",
       "5            NaN           None  \n",
       "6            NaN           None  \n",
       "7          829.0           None  \n",
       "\n",
       "[2608 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bcbe86",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Then Julia uses the `id` fields in each result to join them together into one big dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9b7a3383",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata = rsdf.set_index(\"subject_id\").join(subsdf.set_index(\"id\"), lsuffix='resub', rsuffix=\"subject\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a01f8c5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifierresub</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>identifiersubject</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AD100</th>\n",
       "      <td>0f08e2e9-9983-4204-972f-a630b7ab2c25</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0f08e2e9-9983-420...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD100'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AD1026</th>\n",
       "      <td>6d9d6cb9-652f-4749-b4c5-aa9e6b80de69</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6d9d6cb9-652f-474...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD1026'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AD10328</th>\n",
       "      <td>514fc104-1ee5-4701-8f45-9a011143f1e2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '514fc104-1ee5-470...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD10328'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AD10460</th>\n",
       "      <td>8c36611d-be2f-432a-afde-e684ab4333ea</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8c36611d-be2f-432...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD10460'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AD10485</th>\n",
       "      <td>0ad0fdda-dd96-48df-8edd-e5e471e9f680</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0ad0fdda-dd96-48d...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD10485'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZJ-AB0H</th>\n",
       "      <td>TCGA-ZJ-AB0H__tcga_cesc</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-ZJ-AB0H'}]</td>\n",
       "      <td>tcga_cesc</td>\n",
       "      <td>None</td>\n",
       "      <td>Cervix</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-ZJ-AB0H'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-17869.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZJ-AB0I</th>\n",
       "      <td>a4f13656-a941-498a-9ac9-f020ed559b35</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a4f13656-a941-498...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-ZJ-AB0I'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-9486.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZJ-AB0I</th>\n",
       "      <td>TCGA-ZJ-AB0I__tcga_cesc</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-ZJ-AB0I'}]</td>\n",
       "      <td>tcga_cesc</td>\n",
       "      <td>None</td>\n",
       "      <td>Cervix</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-ZJ-AB0I'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-9486.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZX-AA5X</th>\n",
       "      <td>4756acc0-4e96-44d4-b359-04d64dc7eb84</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4756acc0-4e96-44d...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-ZX-AA5X'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23440.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZX-AA5X</th>\n",
       "      <td>TCGA-ZX-AA5X__tcga_cesc</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-ZX-AA5X'}]</td>\n",
       "      <td>tcga_cesc</td>\n",
       "      <td>None</td>\n",
       "      <td>Cervix</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-ZX-AA5X'}, {...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23440.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3196 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                id  \\\n",
       "AD100         0f08e2e9-9983-4204-972f-a630b7ab2c25   \n",
       "AD1026        6d9d6cb9-652f-4749-b4c5-aa9e6b80de69   \n",
       "AD10328       514fc104-1ee5-4701-8f45-9a011143f1e2   \n",
       "AD10460       8c36611d-be2f-432a-afde-e684ab4333ea   \n",
       "AD10485       0ad0fdda-dd96-48df-8edd-e5e471e9f680   \n",
       "...                                            ...   \n",
       "TCGA-ZJ-AB0H               TCGA-ZJ-AB0H__tcga_cesc   \n",
       "TCGA-ZJ-AB0I  a4f13656-a941-498a-9ac9-f020ed559b35   \n",
       "TCGA-ZJ-AB0I               TCGA-ZJ-AB0I__tcga_cesc   \n",
       "TCGA-ZX-AA5X  4756acc0-4e96-44d4-b359-04d64dc7eb84   \n",
       "TCGA-ZX-AA5X               TCGA-ZX-AA5X__tcga_cesc   \n",
       "\n",
       "                                                identifierresub  \\\n",
       "AD100         [{'system': 'GDC', 'value': '0f08e2e9-9983-420...   \n",
       "AD1026        [{'system': 'GDC', 'value': '6d9d6cb9-652f-474...   \n",
       "AD10328       [{'system': 'GDC', 'value': '514fc104-1ee5-470...   \n",
       "AD10460       [{'system': 'GDC', 'value': '8c36611d-be2f-432...   \n",
       "AD10485       [{'system': 'GDC', 'value': '0ad0fdda-dd96-48d...   \n",
       "...                                                         ...   \n",
       "TCGA-ZJ-AB0H       [{'system': 'IDC', 'value': 'TCGA-ZJ-AB0H'}]   \n",
       "TCGA-ZJ-AB0I  [{'system': 'GDC', 'value': 'a4f13656-a941-498...   \n",
       "TCGA-ZJ-AB0I       [{'system': 'IDC', 'value': 'TCGA-ZJ-AB0I'}]   \n",
       "TCGA-ZX-AA5X  [{'system': 'GDC', 'value': '4756acc0-4e96-44d...   \n",
       "TCGA-ZX-AA5X       [{'system': 'IDC', 'value': 'TCGA-ZX-AA5X'}]   \n",
       "\n",
       "             member_of_research_project  \\\n",
       "AD100                             FM-AD   \n",
       "AD1026                            FM-AD   \n",
       "AD10328                           FM-AD   \n",
       "AD10460                           FM-AD   \n",
       "AD10485                           FM-AD   \n",
       "...                                 ...   \n",
       "TCGA-ZJ-AB0H                  tcga_cesc   \n",
       "TCGA-ZJ-AB0I                  TCGA-CESC   \n",
       "TCGA-ZJ-AB0I                  tcga_cesc   \n",
       "TCGA-ZX-AA5X                  TCGA-CESC   \n",
       "TCGA-ZX-AA5X                  tcga_cesc   \n",
       "\n",
       "                        primary_diagnosis_condition primary_diagnosis_site  \\\n",
       "AD100                       Squamous Cell Neoplasms           Cervix uteri   \n",
       "AD1026          Complex Mixed and Stromal Neoplasms            Uterus, NOS   \n",
       "AD10328                     Squamous Cell Neoplasms           Cervix uteri   \n",
       "AD10460       Cystic, Mucinous and Serous Neoplasms            Uterus, NOS   \n",
       "AD10485       Cystic, Mucinous and Serous Neoplasms            Uterus, NOS   \n",
       "...                                             ...                    ...   \n",
       "TCGA-ZJ-AB0H                                   None                 Cervix   \n",
       "TCGA-ZJ-AB0I                Squamous Cell Neoplasms           Cervix uteri   \n",
       "TCGA-ZJ-AB0I                                   None                 Cervix   \n",
       "TCGA-ZX-AA5X                Squamous Cell Neoplasms           Cervix uteri   \n",
       "TCGA-ZX-AA5X                                   None                 Cervix   \n",
       "\n",
       "                                              identifiersubject       species  \\\n",
       "AD100                     [{'system': 'GDC', 'value': 'AD100'}]  Homo sapiens   \n",
       "AD1026                   [{'system': 'GDC', 'value': 'AD1026'}]  Homo sapiens   \n",
       "AD10328                 [{'system': 'GDC', 'value': 'AD10328'}]  Homo sapiens   \n",
       "AD10460                 [{'system': 'GDC', 'value': 'AD10460'}]  Homo sapiens   \n",
       "AD10485                 [{'system': 'GDC', 'value': 'AD10485'}]  Homo sapiens   \n",
       "...                                                         ...           ...   \n",
       "TCGA-ZJ-AB0H  [{'system': 'GDC', 'value': 'TCGA-ZJ-AB0H'}, {...  Homo sapiens   \n",
       "TCGA-ZJ-AB0I  [{'system': 'GDC', 'value': 'TCGA-ZJ-AB0I'}, {...  Homo sapiens   \n",
       "TCGA-ZJ-AB0I  [{'system': 'GDC', 'value': 'TCGA-ZJ-AB0I'}, {...  Homo sapiens   \n",
       "TCGA-ZX-AA5X  [{'system': 'GDC', 'value': 'TCGA-ZX-AA5X'}, {...  Homo sapiens   \n",
       "TCGA-ZX-AA5X  [{'system': 'GDC', 'value': 'TCGA-ZX-AA5X'}, {...  Homo sapiens   \n",
       "\n",
       "                 sex          race               ethnicity  days_to_birth  \\\n",
       "AD100         female  not reported            not reported            NaN   \n",
       "AD1026        female  not reported            not reported            NaN   \n",
       "AD10328       female  not reported            not reported            NaN   \n",
       "AD10460       female  not reported            not reported            NaN   \n",
       "AD10485       female  not reported            not reported            NaN   \n",
       "...              ...           ...                     ...            ...   \n",
       "TCGA-ZJ-AB0H  female  not reported            not reported       -17869.0   \n",
       "TCGA-ZJ-AB0I  female         white  not hispanic or latino        -9486.0   \n",
       "TCGA-ZJ-AB0I  female         white  not hispanic or latino        -9486.0   \n",
       "TCGA-ZX-AA5X  female         white  not hispanic or latino       -23440.0   \n",
       "TCGA-ZX-AA5X  female         white  not hispanic or latino       -23440.0   \n",
       "\n",
       "             subject_associated_project  vital_status  age_at_death  \\\n",
       "AD100                           [FM-AD]  Not Reported           NaN   \n",
       "AD1026                          [FM-AD]  Not Reported           NaN   \n",
       "AD10328                         [FM-AD]  Not Reported           NaN   \n",
       "AD10460                         [FM-AD]  Not Reported           NaN   \n",
       "AD10485                         [FM-AD]  Not Reported           NaN   \n",
       "...                                 ...           ...           ...   \n",
       "TCGA-ZJ-AB0H     [TCGA-CESC, tcga_cesc]         Alive           NaN   \n",
       "TCGA-ZJ-AB0I     [TCGA-CESC, tcga_cesc]         Alive           NaN   \n",
       "TCGA-ZJ-AB0I     [TCGA-CESC, tcga_cesc]         Alive           NaN   \n",
       "TCGA-ZX-AA5X     [TCGA-CESC, tcga_cesc]         Alive           NaN   \n",
       "TCGA-ZX-AA5X     [TCGA-CESC, tcga_cesc]         Alive           NaN   \n",
       "\n",
       "             cause_of_death  \n",
       "AD100                  None  \n",
       "AD1026                 None  \n",
       "AD10328                None  \n",
       "AD10460                None  \n",
       "AD10485                None  \n",
       "...                     ...  \n",
       "TCGA-ZJ-AB0H           None  \n",
       "TCGA-ZJ-AB0I           None  \n",
       "TCGA-ZJ-AB0I           None  \n",
       "TCGA-ZX-AA5X           None  \n",
       "TCGA-ZX-AA5X           None  \n",
       "\n",
       "[3196 rows x 15 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024da831",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "And saves it out to a csv so she can browse it with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6628de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata.to_csv(\"allmetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246644d3",
   "metadata": {},
   "source": [
    "<img src=\"../images/julia.png\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia knows from her subject count summary that there are 33480 files associated with her subjects, which is likely far more than she needs. In the future, she would like to be able to run something like `file.count.run()` to see a summary of file types so she can filter these down based on data type. Then she could make one data set with histological data, another with MRIs, etc.\n",
    "\n",
    "She would then take her smaller file list and import the files to somewhere that fits her analysis needs like Terra or the Seven Bridges CGC platform."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ae1ae079",
   "metadata": {},
   "outputs": [],
   "source": [
    "#files = NoAdenoData.file.run()\n",
    "#filedf = pd.DataFrame()\n",
    "#for i in files.paginator(to_df=True):\n",
    "#    filedf = pd.concat([subsdf, i])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
